// @weapp-vite/miniprogram-automator/dist/index.mjs
import { randomUUID } from "node:crypto";
import net from "node:net";
import process from "node:process";
import { spawn } from "node:child_process";
import path from "node:path";
import { EventEmitter } from "node:events";
import { createDebug } from "obug";
import WebSocket from "ws";
import fs from "node:fs/promises";
import os from "node:os";
import { decodeQrCodeFromBase64, renderTerminalQrCode } from "@weapp-vite/qr";
import { Buffer } from "node:buffer";
//#region src/internal/compat/async.ts
/**
* @file 异步控制工具。
*/
/** sleep 的方法封装。 */
function sleep$1(timeout) {
	return new Promise((resolve) => setTimeout(resolve, timeout));
}
/** waitUntil 的方法封装。 */
async function waitUntil(condition, timeout = 0, interval = 250) {
	const startTime = Date.now();
	while (true) {
		const value = await condition();
		if (value) return value;
		if (timeout && Date.now() - startTime >= timeout) throw new Error(`Wait timed out after ${timeout} ms`);
		await sleep$1(interval);
	}
}
//#endregion
//#region src/internal/compat/format.ts
/**
* @file 标识与格式化工具。
*/
function padZero(value, width = 2) {
	return String(value).padStart(width, "0");
}
/** dateFormat 的方法封装。 */
function dateFormat(mask, value = /* @__PURE__ */ new Date()) {
	const date = value instanceof Date ? value : new Date(value);
	return mask.replace("yyyy", String(date.getFullYear())).replace("mm", padZero(date.getMonth() + 1)).replace("dd", padZero(date.getDate())).replace("HH", padZero(date.getHours())).replace("MM", padZero(date.getMinutes())).replace("ss", padZero(date.getSeconds())).replace("l", padZero(date.getMilliseconds(), 3));
}
/** uuid 的方法封装。 */
function uuid() {
	return randomUUID();
}
//#endregion
//#region src/internal/compat/network.ts
/**
* @file 网络与环境工具。
*/
const absolutePathPattern = /^(?:[a-z]+:)?[\\/]/i;
/** getPort 的方法封装。 */
function getPort(ports, host) {
	const queue = Array.isArray(ports) ? [...ports] : [ports];
	queue.push(0);
	const checkPort = (port) => {
		return new Promise((resolve, reject) => {
			const server = net.createServer();
			server.unref();
			server.on("error", reject);
			const options = { port };
			if (host) options.host = host;
			server.listen(options, () => {
				const address = server.address();
				const resolvedPort = typeof address === "object" && address ? address.port : port;
				server.close(() => {
					resolve(resolvedPort);
				});
			});
		});
	};
	return queue.reduce((sequence, port) => {
		return sequence.catch(() => checkPort(port));
	}, Promise.reject(/* @__PURE__ */ new Error("No available port")));
}
/** isRelative 的方法封装。 */
function isRelative(value) {
	return !absolutePathPattern.test(value);
}
const isWindows = process.platform === "win32";
//#endregion
//#region src/internal/compat/typeGuards.ts
/**
* @file 基础类型守卫工具。
*/
const objectToString = Object.prototype.toString;
/** isFn 的方法封装。 */
function isFn(value) {
	const type = objectToString.call(value);
	return type === "[object Function]" || type === "[object AsyncFunction]" || type === "[object GeneratorFunction]";
}
/** isNum 的方法封装。 */
function isNum(value) {
	return objectToString.call(value) === "[object Number]";
}
/** isStr 的方法封装。 */
function isStr(value) {
	return objectToString.call(value) === "[object String]";
}
/** isUndef 的方法封装。 */
function isUndef(value) {
	return typeof value === "undefined";
}
/** isPlainObject 的方法封装。 */
function isPlainObject(value) {
	if (objectToString.call(value) !== "[object Object]") return false;
	const prototype = Object.getPrototypeOf(value);
	return prototype === null || prototype === Object.prototype;
}
//#endregion
//#region src/internal/compat/object.ts
/**
* @file 对象与序列化工具。
*/
function cloneDeep(value) {
	if (Array.isArray(value)) return value.map((item) => cloneDeep(item));
	if (isPlainObject(value)) {
		const result = {};
		for (const [key, item] of Object.entries(value)) {
			if (key === "__proto__" || key === "constructor" || key === "prototype") continue;
			result[key] = cloneDeep(item);
		}
		return result;
	}
	return value;
}
function stableSerialize(value) {
	const stack = [];
	const keys = [];
	return JSON.stringify(value, function serialize(key, currentValue) {
		if (stack.length > 0) {
			const currentIndex = stack.indexOf(this);
			if (currentIndex >= 0) {
				stack.splice(currentIndex + 1);
				keys.splice(currentIndex, Infinity, key);
			} else {
				stack.push(this);
				keys.push(key);
			}
			const valueIndex = stack.indexOf(currentValue);
			if (valueIndex >= 0) {
				if (valueIndex === 0) return "[Circular ~]";
				return `[Circular ~.${keys.slice(0, valueIndex).join(".")}]`;
			}
		} else stack.push(currentValue);
		if (typeof currentValue === "function") return `[Function ${currentValue.toString()}]`;
		if (currentValue instanceof RegExp) return `[RegExp ${currentValue.toString()}]`;
		if (typeof currentValue === "undefined") return null;
		return currentValue;
	});
}
/** extendDeep 的方法封装。 */
function extendDeep(target, ...sources) {
	let result = target;
	for (const source of sources) {
		if (isPlainObject(result) && isPlainObject(source)) {
			for (const [key, value] of Object.entries(source)) {
				if (key === "__proto__" || key === "constructor" || key === "prototype") continue;
				result[key] = extendDeep(result[key], value);
			}
			continue;
		}
		result = cloneDeep(source);
	}
	return result;
}
/** isEmpty 的方法封装。 */
function isEmpty(value) {
	if (value == null) return true;
	if (Array.isArray(value) || typeof value === "string") return value.length === 0;
	if (typeof value === "object") return Object.keys(value).length === 0;
	return false;
}
/** stringify 的方法封装。 */
function stringify(value) {
	return stableSerialize(value);
}
//#endregion
//#region src/internal/compat/strings.ts
/**
* @file 字符串与版本处理工具。
*/
const trimPatternCache = /* @__PURE__ */ new Map();
function getTrimPattern(chars) {
	const existing = trimPatternCache.get(chars);
	if (existing) return existing;
	const escaped = chars.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
	const pattern = new RegExp(`^[${escaped}]+|[${escaped}]+$`, "g");
	trimPatternCache.set(chars, pattern);
	return pattern;
}
/** startWith 的方法封装。 */
function startWith(value, prefix) {
	return value.startsWith(prefix);
}
/** endWith 的方法封装。 */
function endWith(value, suffix) {
	return value.endsWith(suffix);
}
/** trim 的方法封装。 */
function trim(value, chars) {
	if (!chars) return value.trim();
	return value.replace(getTrimPattern(chars), "");
}
/** cmpVersion 的方法封装。 */
function cmpVersion(versionA, versionB) {
	const left = versionA.split(".");
	const right = versionB.split(".");
	const length = Math.max(left.length, right.length);
	for (let index = 0; index < length; index += 1) {
		const currentLeft = Number.parseInt(left[index] || "0", 10);
		const currentRight = Number.parseInt(right[index] || "0", 10);
		if (currentLeft > currentRight) return 1;
		if (currentLeft < currentRight) return -1;
	}
	return 0;
}
/** toStr 的方法封装。 */
function toStr(value) {
	return value == null ? "" : String(value);
}
//#endregion
//#region src/Element.ts
/** Element 的实现。 */
var Element = class Element {
	connection;
	elementMap;
	tagName = "";
	nodeId;
	videoId;
	id;
	pageId;
	publicProps;
	constructor(connection, options, elementMap) {
		this.connection = connection;
		this.elementMap = elementMap;
		this.id = options.elementId;
		this.pageId = options.pageId;
		this.nodeId = options.nodeId || null;
		this.videoId = options.videoId || null;
		this.tagName = options.tagName;
	}
	async $(selector) {
		try {
			const element = await this.send("Element.getElement", { selector });
			return Element.create(this.connection, {
				...element,
				pageId: this.pageId
			}, this.elementMap);
		} catch {
			return null;
		}
	}
	async $$(selector) {
		const { elements } = await this.send("Element.getElements", { selector });
		return elements.map((element) => {
			return Element.create(this.connection, {
				...element,
				pageId: this.pageId
			}, this.elementMap);
		});
	}
	async size() {
		const [width, height] = await this.domProperty(["offsetWidth", "offsetHeight"]);
		return {
			width,
			height
		};
	}
	async offset() {
		return await this.send("Element.getOffset");
	}
	async text() {
		return await this.domProperty("innerText");
	}
	async attribute(name) {
		if (!isStr(name)) throw new Error("name must be a string");
		return await this.getter(name, "getAttributes", "attributes");
	}
	async value() {
		return await this.property("value");
	}
	async property(name) {
		if (!isStr(name)) throw new Error("name must be a string");
		let publicProps = this.publicProps;
		if (!publicProps) {
			publicProps = await this._property("__propPublic");
			this.publicProps = publicProps;
		}
		if (!publicProps?.[name]) throw new Error(`${this.tagName}.${name} not exists`);
		return await this._property(name);
	}
	async wxml() {
		return (await this.send("Element.getWXML", { type: "inner" })).wxml;
	}
	async outerWxml() {
		return (await this.send("Element.getWXML", { type: "outer" })).wxml;
	}
	async style(name) {
		if (!isStr(name)) throw new Error("name must be a string");
		return await this.getter(name, "getStyles", "styles");
	}
	async tap() {
		await this.send("Element.tap");
	}
	async longpress() {
		await this.touchstart();
		await sleep$1(350);
		await this.touchend();
	}
	async trigger(type, detail) {
		const payload = { type };
		if (!isUndef(detail)) payload.detail = detail;
		await this.send("Element.triggerEvent", payload);
	}
	async touchstart(options = {}) {
		await this.send("Element.touchstart", options);
	}
	async touchmove(options = {}) {
		await this.send("Element.touchmove", options);
	}
	async touchend(options = {}) {
		await this.send("Element.touchend", options);
	}
	async dispatchEvent(options) {
		await this.send("Element.dispatchEvent", options);
	}
	async _property(name) {
		return await this.getter(name, "getProperties", "properties");
	}
	async send(method, params = {}) {
		params.elementId = this.id;
		params.pageId = this.pageId;
		if (this.nodeId) params.nodeId = this.nodeId;
		if (this.videoId) params.videoId = this.videoId;
		return await this.connection.send(method, params);
	}
	async callFunction(functionName, ...args) {
		const { result } = await this.send("Element.callFunction", {
			functionName,
			args
		});
		return result;
	}
	async domProperty(name) {
		return await this.getter(name, "getDOMProperties", "properties");
	}
	async getter(name, method, field) {
		const names = isStr(name) ? [name] : name;
		const values = (await this.send(`Element.${method}`, { names }))[field];
		return isStr(name) ? values[0] : values;
	}
	static create(connection, options, elementMap) {
		const existing = elementMap.get(options.elementId);
		if (existing) return existing;
		let ElementCtor = Element;
		if (options.nodeId) ElementCtor = CustomElement;
		else switch (options.tagName) {
			case "input":
				ElementCtor = InputElement;
				break;
			case "textarea":
				ElementCtor = TextareaElement;
				break;
			case "scroll-view":
				ElementCtor = ScrollViewElement;
				break;
			case "swiper":
				ElementCtor = SwiperElement;
				break;
			case "movable-view":
				ElementCtor = MovableViewElement;
				break;
			case "switch":
				ElementCtor = SwitchElement;
				break;
			case "slider":
				ElementCtor = SliderElement;
				break;
			case "video":
				ElementCtor = ContextElement;
				break;
			default: ElementCtor = Element;
		}
		const element = new ElementCtor(connection, options, elementMap);
		elementMap.set(options.elementId, element);
		return element;
	}
};
/** CustomElement 的实现。 */
var CustomElement = class extends Element {
	async setData(data) {
		await this.send("Element.setData", { data });
	}
	async data(path) {
		const payload = {};
		if (path) payload.path = path;
		return (await this.send("Element.getData", payload)).data;
	}
	async callMethod(method, ...args) {
		return (await this.send("Element.callMethod", {
			method,
			args
		})).result;
	}
};
/** InputElement 的实现。 */
var InputElement = class extends Element {
	async input(value) {
		await this.callFunction("input.input", value);
	}
};
/** TextareaElement 的实现。 */
var TextareaElement = class extends Element {
	async input(value) {
		await this.callFunction("textarea.input", value);
	}
};
/** ScrollViewElement 的实现。 */
var ScrollViewElement = class extends Element {
	async scrollTo(x, y) {
		await this.callFunction("scroll-view.scrollTo", x, y);
	}
	async property(name) {
		if (name === "scrollTop") return await this.callFunction("scroll-view.scrollTop");
		if (name === "scrollLeft") return await this.callFunction("scroll-view.scrollLeft");
		return await super.property(name);
	}
	async scrollWidth() {
		return await this.callFunction("scroll-view.scrollWidth");
	}
	async scrollHeight() {
		return await this.callFunction("scroll-view.scrollHeight");
	}
};
/** SwiperElement 的实现。 */
var SwiperElement = class extends Element {
	async swipeTo(index) {
		await this.callFunction("swiper.swipeTo", index);
	}
};
/** MovableViewElement 的实现。 */
var MovableViewElement = class extends Element {
	async moveTo(x, y) {
		await this.callFunction("movable-view.moveTo", x, y);
	}
	async property(name) {
		if (name === "x") return await this._property("_translateX");
		if (name === "y") return await this._property("_translateY");
		return await super.property(name);
	}
};
/** SwitchElement 的实现。 */
var SwitchElement = class extends Element {
	async tap() {
		await this.callFunction("switch.tap");
	}
};
/** SliderElement 的实现。 */
var SliderElement = class extends Element {
	async slideTo(value) {
		await this.callFunction("slider.slideTo", value);
	}
};
/** ContextElement 的实现。 */
var ContextElement = class extends Element {
	async callContextMethod(method, ...args) {
		return (await this.send("Element.callContextMethod", {
			method,
			args
		})).result;
	}
};
//#endregion
//#region src/Transport.ts
/** Transport 的实现。 */
var Transport = class extends EventEmitter {
	ws;
	constructor(ws) {
		super();
		this.ws = ws;
		ws.on("message", (data) => {
			this.emit("message", typeof data === "string" ? data : data.toString());
		});
		ws.on("close", () => {
			this.emit("close");
		});
	}
	send(message) {
		this.ws.send(message);
	}
	close() {
		this.ws.close();
	}
};
//#endregion
//#region src/Connection.ts
/**
* @file 自动化协议连接实现。
*/
const debugProtocol = createDebug("automator:protocol");
const closeErrTip = "Connection closed, check if wechat web devTools is still running";
const REQUEST_TIMEOUT = 3e4;
/** Connection 的实现。 */
var Connection = class Connection extends EventEmitter {
	transport;
	callbacks = /* @__PURE__ */ new Map();
	constructor(transport) {
		super();
		this.transport = transport;
		transport.on("message", this.onMessage);
		transport.on("close", this.onClose);
	}
	send(method, params = {}, options = {}) {
		const id = uuid();
		const payload = stringify({
			id,
			method,
			params
		});
		const requestTimeout = options.timeout ?? REQUEST_TIMEOUT;
		debugProtocol(`${dateFormat("yyyy-mm-dd HH:MM:ss:l")} SEND ► ${payload}`);
		return new Promise((resolve, reject) => {
			const timeout = setTimeout(() => {
				this.callbacks.delete(id);
				const error = /* @__PURE__ */ new Error(`DevTools did not respond to protocol method ${method} within ${requestTimeout}ms`);
				error.code = "DEVTOOLS_PROTOCOL_TIMEOUT";
				error.method = method;
				reject(error);
			}, requestTimeout);
			this.callbacks.set(id, {
				resolve,
				reject,
				timeout
			});
			try {
				this.transport.send(payload);
			} catch {
				clearTimeout(timeout);
				this.callbacks.delete(id);
				reject(/* @__PURE__ */ new Error(closeErrTip));
			}
		});
	}
	dispose() {
		this.transport.close();
	}
	onMessage = (message) => {
		debugProtocol(`${dateFormat("yyyy-mm-dd HH:MM:ss:l")} ◀ RECV ${message}`);
		const { id, method, error, result, params } = JSON.parse(message);
		if (!id) {
			this.emit(method, params);
			return;
		}
		const callback = this.callbacks.get(id);
		if (!callback) return;
		this.callbacks.delete(id);
		clearTimeout(callback.timeout);
		if (error) {
			callback.reject(new Error(error.message || closeErrTip));
			return;
		}
		callback.resolve(result);
	};
	onClose = () => {
		for (const callback of this.callbacks.values()) {
			clearTimeout(callback.timeout);
			callback.reject(/* @__PURE__ */ new Error(closeErrTip));
		}
		this.callbacks.clear();
	};
	static create(url) {
		return new Promise((resolve, reject) => {
			const ws = new WebSocket(url);
			ws.on("open", () => {
				resolve(new Connection(new Transport(ws)));
			});
			ws.on("error", reject);
		});
	}
};
//#endregion
//#region src/headless.ts
async function launchHeadlessAutomator(options) {
	return await (await import("./launch-2X75zDRY.mjs")).launch({ projectPath: options.projectPath });
}
//#endregion
//#region src/launcher/portLease.ts
/**
* @file DevTools 自动化端口租约。
*/
const DEFAULT_PORT = 9420;
const PORT_RANGE_SIZE = 200;
const PORT_LEASE_DIR = path.join(os.tmpdir(), "weapp-vite-automator-port-leases");
function createCandidatePorts(preferredPort = DEFAULT_PORT) {
	return Array.from({ length: PORT_RANGE_SIZE }, (_, index) => preferredPort + index);
}
function isPathExistsError(error) {
	return error && typeof error === "object" && "code" in error && error.code === "EEXIST";
}
function isProcessAlive(pid) {
	try {
		process.kill(pid, 0);
		return true;
	} catch (error) {
		if (error && typeof error === "object" && "code" in error && error.code === "ESRCH") return false;
		return true;
	}
}
async function removePortLock(leasePath) {
	await fs.rm(leasePath, { force: true }).catch(() => {});
}
async function removeStalePortLock(leasePath) {
	try {
		const raw = await fs.readFile(leasePath, "utf8");
		const payload = JSON.parse(raw);
		if (typeof payload.pid === "number" && isProcessAlive(payload.pid)) return false;
	} catch {}
	await removePortLock(leasePath);
	return true;
}
async function writePortLock(port, leasePath) {
	const handle = await fs.open(leasePath, "wx");
	await handle.writeFile(JSON.stringify({
		pid: process.pid,
		port,
		updatedAt: (/* @__PURE__ */ new Date()).toISOString()
	}));
	await handle.close();
}
async function createPortLock(port) {
	await fs.mkdir(PORT_LEASE_DIR, { recursive: true });
	const leasePath = path.join(PORT_LEASE_DIR, `${port}.lock`);
	try {
		await writePortLock(port, leasePath);
	} catch (error) {
		if (!isPathExistsError(error) || !await removeStalePortLock(leasePath)) throw error;
		await writePortLock(port, leasePath);
	}
	return leasePath;
}
function createLease(port, leasePath) {
	let released = false;
	return {
		path: leasePath,
		port,
		release: async () => {
			if (released) return;
			released = true;
			if (leasePath) await removePortLock(leasePath);
		}
	};
}
/**
* @description 选择并短暂租约一个自动化端口，避免并发启动选择到同一端口。
*/
async function acquireAutomatorPortLease(preferredPort) {
	if (preferredPort) {
		const port = await getPort(preferredPort, "127.0.0.1");
		if (port !== preferredPort) throw new Error(`Port ${preferredPort} is in use, please specify another port`);
		let leasePath;
		try {
			leasePath = await createPortLock(port);
		} catch (error) {
			if (isPathExistsError(error)) throw new Error(`Port ${preferredPort} is in use, please specify another port`);
			throw error;
		}
		return createLease(port, leasePath);
	}
	for (const candidate of createCandidatePorts()) {
		const port = await getPort(candidate, "127.0.0.1");
		if (port !== candidate) continue;
		try {
			return createLease(port, await createPortLock(port));
		} catch (error) {
			if (isPathExistsError(error)) continue;
			throw error;
		}
	}
	return createLease(await getPort(0, "127.0.0.1"));
}
//#endregion
//#region package.json
var version = "1.2.2";
//#endregion
//#region src/Native.ts
/** Native 的实现。 */
var Native = class extends EventEmitter {
	connection;
	constructor(connection) {
		super();
		this.connection = connection;
	}
	async goHome() {
		return await this.sendNative("goHome");
	}
	async navigateLeft() {
		return await this.sendNative("navigateLeft");
	}
	async confirmModal() {
		return await this.sendNative("confirmModal");
	}
	async cancelModal() {
		return await this.sendNative("cancelModal");
	}
	async switchTab(params) {
		return await this.sendNative("switchTab", params);
	}
	async authorizeCancel() {
		return await this.sendNative("authorizeCancel");
	}
	async authorizeAllow() {
		return await this.sendNative("authorizeAllow");
	}
	async closePaymentDialog() {
		return await this.sendNative("closePaymentDialog");
	}
	async shareCancel() {
		return await this.sendNative("shareCancel");
	}
	async shareConfirm() {
		return await this.sendNative("shareConfirm");
	}
	async send(method, params = {}) {
		return await this.connection.send(method, params);
	}
	async sendNative(method, data) {
		return await this.send("Tool.native", {
			method,
			data
		});
	}
};
//#endregion
//#region src/Page.ts
/** Page 的实现。 */
var Page = class Page {
	connection;
	path = "";
	query = {};
	id;
	elementMap = /* @__PURE__ */ new Map();
	constructor(connection, options) {
		this.connection = connection;
		this.id = options.id;
		this.path = options.path;
		this.query = options.query;
	}
	async waitFor(condition) {
		if (isNum(condition)) {
			await sleep$1(condition);
			return;
		}
		if (isFn(condition)) {
			await waitUntil(condition);
			return;
		}
		if (isStr(condition)) await waitUntil(async () => (await this.$$(condition)).length > 0);
	}
	async $(selector) {
		try {
			const element = await this.send("Page.getElement", { selector });
			return Element.create(this.connection, {
				...element,
				pageId: this.id
			}, this.elementMap);
		} catch {
			return null;
		}
	}
	async $$(selector) {
		const { elements } = await this.send("Page.getElements", { selector });
		return elements.map((element) => {
			return Element.create(this.connection, {
				...element,
				pageId: this.id
			}, this.elementMap);
		});
	}
	async getElementByXpath(selector) {
		try {
			const element = await this.send("Page.getElementByXpath", { selector });
			return Element.create(this.connection, {
				...element,
				pageId: this.id
			}, this.elementMap);
		} catch {
			return null;
		}
	}
	async getElementsByXpath(selector) {
		const { elements } = await this.send("Page.getElementsByXpath", { selector });
		return elements.map((element) => {
			return Element.create(this.connection, {
				...element,
				pageId: this.id
			}, this.elementMap);
		});
	}
	async xpath(selector) {
		return await this.getElementByXpath(selector);
	}
	async data(path) {
		const payload = {};
		if (path) payload.path = path;
		return (await this.send("Page.getData", payload)).data;
	}
	async setData(data) {
		await this.send("Page.setData", { data });
	}
	async size() {
		const [width, height] = await this.windowProperty(["document.documentElement.scrollWidth", "document.documentElement.scrollHeight"]);
		return {
			width,
			height
		};
	}
	async callMethod(method, ...args) {
		return (await this.send("Page.callMethod", {
			method,
			args
		})).result;
	}
	async scrollTop() {
		const bodyScrollTop = await this.windowProperty("document.body.scrollTop");
		const documentScrollTop = await this.windowProperty("document.documentElement.scrollTop");
		return bodyScrollTop || documentScrollTop;
	}
	async windowProperty(name) {
		const names = isStr(name) ? [name] : name;
		const properties = (await this.send("Page.getWindowProperties", { names })).properties;
		return isStr(name) ? properties[0] : properties;
	}
	async send(method, params = {}) {
		params.pageId = this.id;
		return await this.connection.send(method, params);
	}
	static create(connection, options, pageMap) {
		const existing = pageMap.get(options.id);
		if (existing) return existing;
		const page = new Page(connection, options);
		pageMap.set(options.id, page);
		return page;
	}
};
//#endregion
//#region src/util.ts
/**
* @file 二维码与插件路径工具。
*/
const regPluginId = /^plugin-private:\/\/([0-9a-zA-Z]+)\//;
/** decodeQrCode 的方法封装。 */
function decodeQrCode(qrCode) {
	return decodeQrCodeFromBase64(qrCode);
}
/** printQrCode 的方法封装。 */
function printQrCode(content) {
	return new Promise((resolve) => {
		process.stdout.write(renderTerminalQrCode(content, { small: true }));
		resolve();
	});
}
/** isPluginPath 的方法封装。 */
function isPluginPath(p) {
	return startWith(p, "plugin-private://");
}
/** extractPluginId 的方法封装。 */
function extractPluginId(p) {
	const match = p.match(regPluginId);
	return match ? match[1] : "";
}
//#endregion
//#region src/MiniProgram.ts
const CLOSE_STEP_TIMEOUT = 2e3;
const CURRENT_PAGE_RETRIES = 3;
const CURRENT_PAGE_RETRY_DELAY = 400;
const CHANGE_ROUTE_CONTEXT_TIMEOUT = 12e3;
const CHANGE_ROUTE_CALL_TIMEOUT = 12e3;
const CHANGE_ROUTE_READY_TIMEOUT = 15e3;
const CHANGE_ROUTE_POLL_DELAY = 500;
const APP_READY_TIMEOUT = 2e4;
const APP_READY_PROBE_TIMEOUT = 3e3;
const APP_READY_POLL_DELAY = 500;
const CHANGE_ROUTE_DEBUG_ENABLED = process.env.WEAPP_VITE_E2E_CHANGE_ROUTE_DEBUG === "1";
function sleep(ms) {
	return new Promise((resolve) => setTimeout(resolve, ms));
}
function withTimeout(task, timeoutMs) {
	return new Promise((resolve, reject) => {
		const timeout = setTimeout(() => {
			reject(/* @__PURE__ */ new Error(`Operation timed out after ${timeoutMs}ms`));
		}, timeoutMs);
		task.then((value) => {
			clearTimeout(timeout);
			resolve(value);
		}).catch((error) => {
			clearTimeout(timeout);
			reject(error);
		});
	});
}
function isOperationTimeoutError(error, timeoutMs) {
	return error instanceof Error && error.message.includes(`Operation timed out after ${timeoutMs}ms`);
}
function isFnStr(value) {
	if (!isStr(value)) return false;
	const trimmed = trim(value);
	return startWith(trimmed, "function") || startWith(trimmed, "() =>");
}
function isCurrentPageProtocolTimeout(error) {
	return error instanceof Error && "code" in error && error.code === "DEVTOOLS_PROTOCOL_TIMEOUT" && "method" in error && error.method === "App.getCurrentPage";
}
function isPageStackProtocolTimeout(error) {
	return error instanceof Error && "code" in error && error.code === "DEVTOOLS_PROTOCOL_TIMEOUT" && "method" in error && error.method === "App.getPageStack";
}
function isCallWxMethodProtocolTimeout(error) {
	return error instanceof Error && "code" in error && error.code === "DEVTOOLS_PROTOCOL_TIMEOUT" && "method" in error && error.method === "App.callWxMethod";
}
function isRouteContextProbeError(error) {
	return isCurrentPageProtocolTimeout(error) || isPageStackProtocolTimeout(error);
}
function normalizeRoutePath(value) {
	return String(value ?? "").split("?", 1)[0].replace(/^\/+/, "").replace(/\/+$/, "");
}
function logChangeRouteDebug(message) {
	if (!CHANGE_ROUTE_DEBUG_ENABLED) return;
	process.stdout.write(`[automator:changeRoute] ${message}\n`);
}
/** MiniProgram 的实现。 */
var MiniProgram = class extends EventEmitter {
	connection;
	appBindings = /* @__PURE__ */ new Map();
	pageMap = /* @__PURE__ */ new Map();
	nativeIns;
	constructor(connection) {
		super();
		this.connection = connection;
		connection.on("App.logAdded", this.onLogAdded);
		connection.on("App.bindingCalled", this.onBindingCalled);
		connection.on("App.exceptionThrown", this.onExceptionThrown);
	}
	async pageStack() {
		const { pageStack } = await this.send("App.getPageStack");
		return pageStack.map((page) => {
			return Page.create(this.connection, {
				id: page.pageId,
				path: page.path,
				query: page.query
			}, this.pageMap);
		});
	}
	async navigateTo(url) {
		return await this.changeRoute("navigateTo", url);
	}
	async redirectTo(url) {
		return await this.changeRoute("redirectTo", url);
	}
	async navigateBack() {
		return await this.changeRoute("navigateBack");
	}
	async reLaunch(url) {
		return await this.changeRoute("reLaunch", url);
	}
	async switchTab(url) {
		return await this.changeRoute("switchTab", url);
	}
	async currentPage(options = {}) {
		let lastError;
		const retries = options.retries ?? CURRENT_PAGE_RETRIES;
		const sendOptions = options.timeout ? { timeout: options.timeout } : void 0;
		for (let attempt = 1; attempt <= retries; attempt += 1) try {
			const { pageId, path, query } = await this.send("App.getCurrentPage", {}, sendOptions);
			return Page.create(this.connection, {
				id: pageId,
				path,
				query
			}, this.pageMap);
		} catch (error) {
			lastError = error;
			if (!isCurrentPageProtocolTimeout(error)) throw error;
			if (attempt < retries) {
				await sleep(CURRENT_PAGE_RETRY_DELAY);
				continue;
			}
		}
		if (isCurrentPageProtocolTimeout(lastError)) try {
			const { pageStack } = await this.send("App.getPageStack", {}, sendOptions);
			const page = pageStack[pageStack.length - 1];
			if (page) return Page.create(this.connection, {
				id: page.pageId,
				path: page.path,
				query: page.query
			}, this.pageMap);
		} catch (error) {
			if (!isPageStackProtocolTimeout(error)) throw error;
		}
		throw lastError;
	}
	async systemInfo() {
		return await this.callWxMethod("getSystemInfoSync");
	}
	async callWxMethod(method, ...args) {
		return (await this.send("App.callWxMethod", {
			method,
			args
		})).result;
	}
	async callWxMethodWithTimeout(method, timeout, ...args) {
		return (await this.send("App.callWxMethod", {
			method,
			args
		}, { timeout })).result;
	}
	async mockWxMethod(method, result, ...args) {
		if (isFn(result) || isFnStr(result)) {
			await this.send("App.mockWxMethod", {
				method,
				functionDeclaration: result.toString(),
				args
			});
			return;
		}
		await this.send("App.mockWxMethod", {
			method,
			result
		});
	}
	async restoreWxMethod(method) {
		await this.send("App.mockWxMethod", { method });
	}
	async callPluginWxMethod(pluginId, method, ...args) {
		return (await this.send("App.callWxMethod", {
			method,
			pluginId,
			args
		})).result;
	}
	async mockPluginWxMethod(pluginId, method, result, ...args) {
		if (isFn(result) || isFnStr(result)) {
			await this.send("App.mockWxMethod", {
				method,
				pluginId,
				functionDeclaration: result.toString(),
				args
			});
			return;
		}
		await this.send("App.mockWxMethod", {
			method,
			pluginId,
			result
		});
	}
	async restorePluginWxMethod(pluginId, method) {
		await this.send("App.mockWxMethod", {
			method,
			pluginId
		});
	}
	async evaluate(appFunction, ...args) {
		const { result } = await this.send("App.callFunction", {
			functionDeclaration: appFunction.toString(),
			args
		});
		return result;
	}
	async pageScrollTo(scrollTop) {
		await this.callWxMethod("pageScrollTo", {
			scrollTop,
			duration: 0
		});
	}
	async close() {
		try {
			await withTimeout(this.send("App.exit"), CLOSE_STEP_TIMEOUT);
		} catch {}
		await sleep(1e3);
		try {
			await withTimeout(this.send("Tool.close"), CLOSE_STEP_TIMEOUT);
		} catch {} finally {
			this.disconnect();
		}
	}
	async remote(auto = false) {
		const { qrCode } = await this.send("Tool.enableRemoteDebug", { auto });
		if (qrCode) await printQrCode(await decodeQrCode(qrCode));
		return new Promise((resolve) => {
			this.connection.once("Tool.onRemoteDebugConnected", async () => {
				await sleep(1e3);
				resolve();
			});
		});
	}
	disconnect() {
		this.connection.dispose();
	}
	on(event, listener) {
		if (event === "console") this.send("App.enableLog").catch(() => {});
		super.on(event, listener);
		return this;
	}
	async exposeFunction(name, bindingFunction) {
		if (this.appBindings.has(name)) throw new Error(`Failed to expose function with name ${name}: already exists!`);
		this.appBindings.set(name, bindingFunction);
		await this.send("App.addBinding", { name });
	}
	async checkVersion() {
		const sdkVersion = (await this.send("Tool.getInfo")).SDKVersion;
		if (sdkVersion !== "dev" && cmpVersion(sdkVersion, "2.7.3") < 0) throw new Error(`SDKVersion is currently ${sdkVersion}, while automator(${version}) requires at least version 2.7.3`);
	}
	/**
	* @description 等待小程序 App 域协议可用，避免返回半就绪自动化会话。
	*/
	async waitForAppReady(timeout = APP_READY_TIMEOUT) {
		const startedAt = Date.now();
		let lastError;
		while (Date.now() - startedAt <= timeout) try {
			await this.send("App.captureScreenshot", {}, { timeout: APP_READY_PROBE_TIMEOUT });
			return;
		} catch (error) {
			lastError = error;
			await sleep(APP_READY_POLL_DELAY);
		}
		throw lastError instanceof Error ? lastError : /* @__PURE__ */ new Error(`Timed out waiting App domain readiness after ${timeout}ms`);
	}
	async screenshot(options = {}) {
		const { data } = await this.send("App.captureScreenshot");
		if (!options.path) return data;
		await fs.writeFile(options.path, data, "base64");
	}
	async testAccounts() {
		return (await this.send("Tool.getTestAccounts")).accounts;
	}
	/**
	* @description 获取开发者工具 Tool 域基础信息。
	*/
	async toolInfo() {
		return await this.send("Tool.getInfo");
	}
	/**
	* @description 触发开发者工具执行一次项目编译。
	*/
	async compile(options = {}) {
		return await this.send("Tool.compile", options);
	}
	/**
	* @description 清理开发者工具缓存。
	*/
	async clearCache(options) {
		return await this.send("Tool.clearCache", options);
	}
	/**
	* @description 调用开发者工具 Tool 域协议方法。
	*/
	async tool(method, params = {}) {
		return await this.send(`Tool.${method}`, params);
	}
	async stopAudits(options = {}) {
		const result = await this.send("Tool.stopAudits");
		if (options.path) await fs.writeFile(options.path, result.report, "utf8");
		return JSON.parse(result.data);
	}
	async getTicket() {
		return await this.send("Tool.getTicket");
	}
	async setTicket(ticket) {
		await this.send("Tool.setTicket", { ticket });
	}
	async refreshTicket() {
		await this.send("Tool.refreshTicket");
	}
	native() {
		if (!this.nativeIns) this.nativeIns = new Native(this.connection);
		return this.nativeIns;
	}
	async changeRoute(method, url) {
		const currentPage = await this.resolveRouteContextPage();
		logChangeRouteDebug(`start method=${method} url=${url ?? "<none>"} current=${currentPage?.path ?? "<none>"}`);
		try {
			if (currentPage && isPluginPath(currentPage.path)) await this.callPluginWxMethod(extractPluginId(currentPage.path), method, { url });
			else await this.callWxMethodWithTimeout(method, CHANGE_ROUTE_CALL_TIMEOUT, { url });
		} catch (error) {
			if (isCallWxMethodProtocolTimeout(error) || isOperationTimeoutError(error, CHANGE_ROUTE_CALL_TIMEOUT)) logChangeRouteDebug(`call-timeout method=${method} url=${url ?? "<none>"}`);
			else throw error;
		}
		const expectedRoute = normalizeRoutePath(url);
		const startedAt = Date.now();
		let lastPage;
		let lastError;
		while (Date.now() - startedAt <= CHANGE_ROUTE_READY_TIMEOUT) {
			try {
				const page = await this.readRoutePollingCurrentPage();
				lastPage = page;
				logChangeRouteDebug(`poll method=${method} url=${url ?? "<none>"} current=${page?.path ?? "<none>"}`);
				if (!expectedRoute || normalizeRoutePath(page?.path) === expectedRoute) {
					logChangeRouteDebug(`ready method=${method} url=${url ?? "<none>"} current=${page?.path ?? "<none>"}`);
					return page;
				}
			} catch (error) {
				lastError = error;
				logChangeRouteDebug(`poll-error method=${method} url=${url ?? "<none>"} error=${error instanceof Error ? error.message : String(error)}`);
			}
			try {
				const stack = await this.readRoutePollingPageStack();
				const stackTop = stack[stack.length - 1];
				if (stackTop) {
					lastPage = stackTop;
					logChangeRouteDebug(`stack method=${method} url=${url ?? "<none>"} current=${stackTop.path}`);
					if (!expectedRoute || normalizeRoutePath(stackTop.path) === expectedRoute) {
						logChangeRouteDebug(`stack-ready method=${method} url=${url ?? "<none>"} current=${stackTop.path}`);
						return stackTop;
					}
				}
			} catch (error) {
				lastError = error;
				logChangeRouteDebug(`stack-error method=${method} url=${url ?? "<none>"} error=${error instanceof Error ? error.message : String(error)}`);
				if (isOperationTimeoutError(error, CHANGE_ROUTE_CONTEXT_TIMEOUT)) {
					await sleep(CHANGE_ROUTE_POLL_DELAY);
					continue;
				}
			}
			await sleep(CHANGE_ROUTE_POLL_DELAY);
		}
		if (lastPage) {
			logChangeRouteDebug(`timeout method=${method} url=${url ?? "<none>"} current=${lastPage.path}`);
			throw new Error(`Timed out waiting route ${expectedRoute || "<current>"} after ${method}; current page: ${lastPage.path}`);
		}
		logChangeRouteDebug(`timeout method=${method} url=${url ?? "<none>"} current=<none> error=${lastError instanceof Error ? lastError.message : String(lastError)}`);
		throw lastError instanceof Error ? lastError : /* @__PURE__ */ new Error(`Timed out waiting route ${expectedRoute || "<current>"} after ${method}`);
	}
	async resolveRouteContextPage() {
		const currentPageTask = this.currentPage().catch((error) => {
			if (isRouteContextProbeError(error)) return;
			throw error;
		});
		const timeoutTask = sleep(CHANGE_ROUTE_CONTEXT_TIMEOUT).then(() => void 0);
		try {
			return await Promise.race([currentPageTask, timeoutTask]);
		} finally {
			currentPageTask.catch(() => {});
		}
	}
	async readRoutePollingCurrentPage() {
		const { pageId, path, query } = await this.send("App.getCurrentPage", {}, { timeout: CHANGE_ROUTE_CONTEXT_TIMEOUT });
		return Page.create(this.connection, {
			id: pageId,
			path,
			query
		}, this.pageMap);
	}
	async readRoutePollingPageStack() {
		const { pageStack } = await this.send("App.getPageStack", {}, { timeout: CHANGE_ROUTE_CONTEXT_TIMEOUT });
		return pageStack.map((page) => {
			return Page.create(this.connection, {
				id: page.pageId,
				path: page.path,
				query: page.query
			}, this.pageMap);
		});
	}
	async send(method, params = {}, options) {
		return options ? await this.connection.send(method, params, options) : await this.connection.send(method, params);
	}
	onLogAdded = (payload) => {
		this.emit("console", payload);
	};
	onBindingCalled = (payload) => {
		const binding = this.appBindings.get(payload.name);
		if (!binding) return;
		try {
			binding(...payload.args);
		} catch {}
	};
	onExceptionThrown = (payload) => {
		this.emit("exception", payload);
	};
};
//#endregion
//#region src/platform.ts
function normalizePlatform(platform = "wechat") {
	return platform === "baidu" ? "swan" : platform;
}
//#endregion
//#region src/smartapp/errors.ts
/**
* @file 百度智能小程序自动化错误定义。
*/
var SmartappUnsupportedError = class extends Error {
	constructor(feature) {
		super(`${feature} is not available in the lightweight TypeScript smartapp runtime.`);
		this.name = "SmartappUnsupportedError";
	}
};
function unsupported(feature) {
	throw new SmartappUnsupportedError(feature);
}
//#endregion
//#region src/smartapp/time.ts
/**
* @file 百度智能小程序自动化时间工具。
*/
function currentTimestamp() {
	return Date.now();
}
async function delay(ms) {
	await new Promise((resolve) => setTimeout(resolve, ms));
}
const time = {
	currentTimestamp,
	delay
};
//#endregion
//#region src/smartapp/Element.ts
function getCoreElementId(element, selectors, index) {
	if (!element) return `${selectors}:${index}`;
	const elementId = Reflect.get(element, "id");
	return typeof elementId === "string" ? elementId : `${selectors}:${index}`;
}
function isInputCapableElement(element) {
	return typeof Reflect.get(element, "input") === "function";
}
var SmartappElement = class {
	element;
	selectors;
	index;
	id;
	tagName;
	constructor(element, selectors = "", index = 0) {
		this.element = element;
		this.selectors = selectors;
		this.index = index;
		this.id = getCoreElementId(element, selectors, index);
		this.tagName = element?.tagName || "";
	}
	async scrollIntoView(options = {}) {
		await delay(options.wait ?? 1e3);
	}
	async boundingBox() {
		if (!this.element) return null;
		return await this.element.offset();
	}
	async tap(options = {}) {
		const count = options.count ?? 1;
		for (let index = 0; index < count; index++) await this.element?.tap();
		await delay(options.wait ?? 1e3);
	}
	async input(text, options = {}) {
		if (this.element) {
			if (!isInputCapableElement(this.element)) throw new SmartappUnsupportedError(`element.input(${this.tagName || "unknown"})`);
			await this.element.input(text);
		}
		await delay(options.wait ?? 1e3);
	}
	async text() {
		return await this.element?.text();
	}
	async value() {
		return await this.property("value");
	}
	async attribute(name) {
		return await this.element?.attribute(name);
	}
	async property(name) {
		return await this.element?.property(name);
	}
	async style(style) {
		return await this.element?.property(style);
	}
	async longpress(options = {}) {
		await this.element?.longpress();
		await delay(options.duration ?? 0);
	}
	async move(options = {}) {
		await delay(options.wait ?? 1e3);
	}
	async screenshot(options = {}) {
		return Buffer.alloc(0);
	}
};
//#endregion
//#region src/smartapp/Page.ts
var SmartappPage = class {
	page;
	path;
	query;
	uri;
	constructor(page, path = "", query = {}) {
		this.page = page;
		this.path = page?.path || path;
		this.query = page?.query || query;
		const queryText = new URLSearchParams(this.query).toString();
		this.uri = queryText ? `${this.path}?${queryText}` : this.path;
	}
	async data(path) {
		return await this.page?.data(path);
	}
	async setData(data) {
		await this.page?.setData(data);
	}
	async callMethod(method, ...args) {
		return await this.page?.callMethod(method, ...args);
	}
};
//#endregion
//#region src/smartapp/App.ts
var SmartappApp = class {
	miniProgram;
	appKey;
	constructor(miniProgram, appKey) {
		this.miniProgram = miniProgram;
		this.appKey = appKey;
	}
	async goto(path = "", options = {}) {
		const retry = options.retry ?? 0;
		let lastError;
		for (let count = 0; count <= retry; count++) try {
			return new SmartappPage(await this.miniProgram?.reLaunch(path) || null, path);
		} catch (error) {
			lastError = error;
		}
		throw lastError;
	}
	async screenshot(options = {}) {
		return await this.miniProgram?.screenshot(options);
	}
	async currentPage() {
		return new SmartappPage(await this.miniProgram?.currentPage() || null);
	}
	async data() {
		return await this.miniProgram?.callWxMethod("getSystemInfo", {});
	}
	async tabs() {
		return [];
	}
	async pages() {
		return await this.miniProgram?.pageStack();
	}
	async getSystemInfo() {
		return await this.miniProgram?.callWxMethod("getSystemInfo", {});
	}
	async source() {
		const page = await this.miniProgram?.currentPage();
		return JSON.stringify(await page?.data());
	}
	async swan(method, ...args) {
		return await this.miniProgram?.callWxMethod(method, ...args);
	}
	async getAbTestInfo() {
		return {};
	}
	async $(selectors, options = {}) {
		await delay(options.duration ?? 0);
		const element = await (await this.miniProgram?.currentPage())?.$(selectors);
		return element ? new SmartappElement(element, selectors, 0) : null;
	}
	async $$(selectors, options = {}) {
		await delay(options.duration ?? 0);
		return (await (await this.miniProgram?.currentPage())?.$$(selectors) || []).map((element, index) => new SmartappElement(element, selectors, index));
	}
	async $x(_expression) {
		return [];
	}
	async close() {
		this.miniProgram?.disconnect();
	}
};
//#endregion
//#region src/smartapp/Device.ts
var SmartappDevice = class {
	type;
	connectType;
	id;
	driver;
	constructor(type, connectType, id, driver = {}) {
		this.type = type;
		this.connectType = connectType;
		this.id = id;
		this.driver = driver;
	}
	async source(options = {}) {
		if (this.driver.source) return await this.driver.source();
		return "";
	}
	async screenshot(options = {}) {
		if (this.driver.screenshot) return await this.driver.screenshot();
		return Buffer.alloc(0);
	}
	async launchSmartapp(_container, appKey, path = "") {
		const app = await this.newSmartapp(_container, appKey);
		if (path) await app.goto(path);
		return app;
	}
	async getScreenSize() {
		return {
			height: 0,
			width: 0
		};
	}
	async terminateApp() {
		unsupported("device.terminateApp");
	}
	async launchApp() {
		unsupported("device.launchApp");
	}
	async reLaunchApp() {
		unsupported("device.reLaunchApp");
	}
	async install() {
		unsupported("device.install");
	}
	async uninstall() {
		unsupported("device.uninstall");
	}
	async newSmartapp(_container, appKey) {
		return new SmartappApp(this.driver.miniProgram || null, appKey);
	}
	async tap() {
		unsupported("device.tap");
	}
	async swipe() {
		unsupported("device.swipe");
	}
	async longpress() {
		unsupported("device.longpress");
	}
	async $x() {
		return [];
	}
	async home() {
		unsupported("device.home");
	}
	async enter() {
		unsupported("device.enter");
	}
	async back() {
		unsupported("device.back");
	}
	async isInstall() {
		return false;
	}
	async close() {
		await this.driver.close?.();
		this.driver.miniProgram?.disconnect();
	}
};
//#endregion
//#region src/smartapp/file.ts
/**
* @file 百度智能小程序自动化文件工具。
*/
async function mkdir(dir) {
	await fs.mkdir(dir, { recursive: true });
}
async function rmdir(dir) {
	await fs.rm(dir, {
		recursive: true,
		force: true
	});
}
async function wget(url, targetPath) {
	const response = await fetch(url);
	if (!response.ok) throw new Error(`Failed downloading ${url}: ${response.status} ${response.statusText}`);
	await mkdir(path.dirname(targetPath));
	const buffer = Buffer.from(await response.arrayBuffer());
	await fs.writeFile(targetPath, buffer);
}
async function changeProjectSwanJson(projectPath, values) {
	const swanJsonPath = path.resolve(projectPath, "project.swan.json");
	const raw = await fs.readFile(swanJsonPath, "utf8");
	const current = JSON.parse(raw);
	await fs.writeFile(swanJsonPath, JSON.stringify({
		...current,
		...values
	}, null, 2), "utf8");
}
const file = {
	changeProjectSwanJson,
	mkdir,
	rmdir,
	wget
};
//#endregion
//#region src/smartapp/image.ts
async function drawBorder(imageBuffer) {
	return imageBuffer;
}
async function extract(imageBuffer) {
	return imageBuffer;
}
async function border(imageBuffer) {
	return imageBuffer;
}
const image = {
	border,
	drawBorder,
	extract
};
//#endregion
//#region src/smartapp/log.ts
const rawConsole = console;
const loggers = /* @__PURE__ */ new Map();
function createLogger(label) {
	return {
		error: (message) => rawConsole.error(`[${label}] ${message}`),
		info: (message) => rawConsole.info(`[${label}] ${message}`),
		warn: (message) => rawConsole.warn(`[${label}] ${message}`)
	};
}
function set(_dir, label = "smartapp-automator") {
	const logger = createLogger(label);
	loggers.set(label, logger);
	return logger;
}
function get(label = "smartapp-automator") {
	const logger = loggers.get(label);
	if (!logger) return set("", label);
	return logger;
}
const log = {
	get,
	set
};
//#endregion
//#region src/smartapp/index.ts
const DEFAULT_CONNECT_TYPE = "usb";
function resolveDeviceType(deviceType) {
	switch (deviceType) {
		case "android":
		case "iOS":
		case "web":
		case "devtools":
		case "simulator": return deviceType;
		default: return "simulator";
	}
}
function resolveConnectType(connectType) {
	switch (connectType) {
		case "usb":
		case "wifi":
		case "newBaiduCloud":
		case "mtpaas": return connectType;
		default: return DEFAULT_CONNECT_TYPE;
	}
}
async function connectMiniProgram(wsEndpoint) {
	return new MiniProgram(await Connection.create(wsEndpoint));
}
async function launch(options = {}) {
	const deviceType = resolveDeviceType(options.deviceType);
	const connectType = resolveConnectType(options.connectType);
	const deviceId = options.deviceId || `${deviceType}-${connectType}`;
	if (deviceType === "simulator" || deviceType === "devtools") return new SmartappDevice("devtools", connectType, deviceId, { miniProgram: options.wsEndpoint ? await connectMiniProgram(options.wsEndpoint) : void 0 });
	return new SmartappDevice(deviceType, connectType, deviceId);
}
async function connect(options) {
	if (typeof options.wsEndpoint !== "string") throw new TypeError("wsEndpoint is required for smartapp runtime connection.");
	return new SmartappDevice("devtools", DEFAULT_CONNECT_TYPE, "devtools-usb", { miniProgram: await connectMiniProgram(options.wsEndpoint) });
}
async function devices(deviceType, _connectType = DEFAULT_CONNECT_TYPE) {
	const resolvedDeviceType = resolveDeviceType(deviceType);
	if (resolvedDeviceType === "simulator" || resolvedDeviceType === "devtools") return { devtools: { status: 0 } };
	return {};
}
const SmartappAutomator$1 = {
	connect,
	devices,
	file,
	image,
	launch,
	log,
	time
};
//#endregion
//#region src/SwanLauncher.ts
const smartappAutomatorRuntime = SmartappAutomator$1;
function toSmartappLaunchOptions(options) {
	const { cliPath, platform: _platform, runtimeProvider: _runtimeProvider, ...rest } = options;
	return {
		deviceType: options.deviceType || "simulator",
		...rest,
		...cliPath ? { devtoolsPath: cliPath } : {}
	};
}
var SwanLauncher = class {
	automator;
	constructor(automator = SmartappAutomator$1) {
		this.automator = automator;
	}
	async launch(options) {
		return await this.automator.launch(toSmartappLaunchOptions(options));
	}
	async connect(options) {
		if (!this.automator.connect) throw new Error("smartapp-automator does not expose connect(). Please use launch() for Baidu smart program automation.");
		return await this.automator.connect({ ...options });
	}
};
//#endregion
//#region src/Launcher.ts
/**
* @file 开发者工具启动与连接流程。
*/
const DEFAULT_TIMEOUT = 3e4;
const AUTOMATOR_LAUNCH_RETRIES = 3;
const DEFAULT_RUNTIME_PROVIDER_ENV = "WEAPP_VITE_AUTOMATOR_RUNTIME_PROVIDER";
const LEGACY_RUNTIME_PROVIDER_ENV = "WEAPP_VITE_E2E_RUNTIME_PROVIDER";
const EXTENSION_CONTEXT_INVALIDATED_RE = /Extension context invalidated/i;
const RETRYABLE_LAUNCH_PORT_RE = /Wait timed out after \d+ ms|Failed connecting to ws:\/\/127\.0\.0\.1:\d+|Failed connecting to devtools websocket endpoint|Failed to launch wechat web devTools, please make sure cliPath is correctly specified/i;
const WINDOWS_BATCH_CLI_RE = /\.(?:bat|cmd)$/i;
let localhostListenPatched = false;
function isExtensionContextInvalidatedError(error) {
	return error instanceof Error && EXTENSION_CONTEXT_INVALIDATED_RE.test(error.message);
}
function isRetryableAutomatorPortLaunchError(error) {
	return error instanceof Error && RETRYABLE_LAUNCH_PORT_RE.test(error.message);
}
function retainPortLeaseUntilSessionClose(miniProgram, portLease) {
	let released = false;
	const release = async () => {
		if (released) return;
		released = true;
		await portLease.release();
	};
	const target = miniProgram;
	const rawClose = target.close;
	const rawDisconnect = target.disconnect;
	if (typeof rawClose !== "function" && typeof rawDisconnect !== "function") return false;
	if (typeof rawDisconnect === "function") target.disconnect = function disconnectWithPortLeaseRelease() {
		try {
			return rawDisconnect.call(this);
		} finally {
			release();
		}
	};
	if (typeof rawClose === "function") target.close = async function closeWithPortLeaseRelease() {
		try {
			return await rawClose.call(this);
		} finally {
			await release();
		}
	};
	return true;
}
function patchNetListenToLoopback() {
	if (localhostListenPatched) return;
	localhostListenPatched = true;
	const rawListen = net.Server.prototype.listen;
	net.Server.prototype.listen = function patchedListen(...args) {
		const firstArg = args[0];
		if (firstArg && typeof firstArg === "object" && !Array.isArray(firstArg)) {
			if (!("host" in firstArg) || !firstArg.host) args[0] = {
				...firstArg,
				host: "127.0.0.1"
			};
			return rawListen.apply(this, args);
		}
		if ((typeof firstArg === "number" || typeof firstArg === "string") && typeof args[1] !== "string") args.splice(1, 0, "127.0.0.1");
		return rawListen.apply(this, args);
	};
}
/** IConnectOptions 的类型定义。 */
function shouldUseWindowsCommandShell(cliPath) {
	return isWindows && WINDOWS_BATCH_CLI_RE.test(cliPath);
}
function escapeWindowsCmdArg(arg) {
	const escaped = arg.replace(/"/g, "\"\"").replace(/%/g, "%%");
	return /[\s"&<>^|()]/.test(arg) ? `"${escaped}"` : escaped;
}
function resolveWindowsBatchSpawn(cliPath, args) {
	return {
		file: process.env.ComSpec || "cmd.exe",
		args: [
			"/d",
			"/s",
			"/c",
			`"${[cliPath, ...args].map(escapeWindowsCmdArg).join(" ")}"`
		]
	};
}
function resolveRuntimeProvider(options) {
	return options.runtimeProvider || process.env[DEFAULT_RUNTIME_PROVIDER_ENV] || process.env[LEGACY_RUNTIME_PROVIDER_ENV] || "devtools";
}
/** Launcher 的实现。 */
var Launcher = class {
	async launch(options) {
		if (normalizePlatform(options.platform) === "swan") return await new SwanLauncher().launch(options);
		if (resolveRuntimeProvider(options) === "headless") {
			if (!options.projectPath) throw new Error("projectPath is not provided");
			return await launchHeadlessAutomator({ projectPath: options.projectPath });
		}
		patchNetListenToLoopback();
		if (options.port) return await this.launchWechatDevtools(options);
		let lastError = null;
		for (let attempt = 1; attempt <= AUTOMATOR_LAUNCH_RETRIES; attempt += 1) try {
			return await this.launchWechatDevtools(options);
		} catch (error) {
			lastError = error;
			if (!isRetryableAutomatorPortLaunchError(error) || attempt === AUTOMATOR_LAUNCH_RETRIES) throw error;
		}
		throw lastError instanceof Error ? lastError : new Error(String(lastError));
	}
	async launchWechatDevtools(options) {
		const { cliPath = await this.resolveCliPath(), timeout = DEFAULT_TIMEOUT, projectConfig = {}, ticket = "", cwd = "", account = "", trustProject = false } = options;
		let { args = [], projectPath } = options;
		const portLease = await acquireAutomatorPortLease(options.port);
		let releasePortLeaseOnExit = true;
		try {
			const port = portLease.port;
			if (!cliPath) throw new Error("Wechat web devTools not found, please specify cliPath option");
			if (isWindows && endWith(cliPath, ".exe")) throw new Error("cliPath is not correct, it's usually named as 'cli' or 'cli.bat'");
			if (!projectPath) throw new Error("projectPath is not provided");
			const resolvedProjectPath = isRelative(projectPath) ? path.resolve(projectPath) : projectPath;
			if (isRelative(projectPath)) projectPath = resolvedProjectPath;
			if (!await import("node:fs/promises").then((fs) => fs.access(resolvedProjectPath).then(() => true).catch(() => false))) throw new Error(`Project path ${resolvedProjectPath} doesn't exist`);
			if (!isEmpty(projectConfig)) await this.extendProjectConfig(projectConfig, resolvedProjectPath);
			let processError = null;
			let processExitCode = null;
			let processSignal = null;
			args = [
				...args,
				"auto",
				"--project",
				resolvedProjectPath,
				"--auto-port",
				toStr(port)
			];
			if (account) args.push("--auto-account", account);
			else if (ticket) args.push("--ticket", ticket);
			if (trustProject) args.push("--trust-project");
			try {
				const spawnTarget = shouldUseWindowsCommandShell(cliPath) ? resolveWindowsBatchSpawn(cliPath, args) : {
					file: cliPath,
					args
				};
				const child = spawn(spawnTarget.file, spawnTarget.args, {
					stdio: "ignore",
					cwd: cwd || void 0,
					...shouldUseWindowsCommandShell(cliPath) ? {
						windowsHide: true,
						windowsVerbatimArguments: true
					} : {}
				});
				child.on("error", (error) => {
					processError = error;
				});
				child.on("exit", (code, signal) => {
					processExitCode = code;
					processSignal = signal;
					if (code !== 0 || signal) processError = /* @__PURE__ */ new Error(`DevTools cli exited unexpectedly with code ${code ?? "null"}${signal ? ` and signal ${signal}` : ""}`);
				});
				child.unref();
			} catch (error) {
				processError = error;
			}
			let miniProgram = null;
			let lastConnectError = null;
			await waitUntil(async () => {
				try {
					if (processError) return true;
					const candidate = await this.connectTool({ wsEndpoint: `ws://127.0.0.1:${port}` });
					try {
						await candidate.checkVersion();
						if (typeof candidate.waitForAppReady === "function") await candidate.waitForAppReady(timeout);
					} catch (error) {
						candidate.disconnect();
						lastConnectError = error;
						if (isExtensionContextInvalidatedError(error)) return false;
						throw error;
					}
					miniProgram = candidate;
					return true;
				} catch (error) {
					lastConnectError = error;
					return false;
				}
			}, timeout, 1e3);
			if (!miniProgram) {
				if (processError) throw new Error("Failed to launch wechat web devTools, please make sure cliPath is correctly specified");
				if (lastConnectError) throw lastConnectError;
				if (processExitCode !== null || processSignal) throw new Error("Failed to launch wechat web devTools, please make sure http port is open");
				throw new Error("Failed connecting to devtools websocket endpoint");
			}
			const resolvedMiniProgram = miniProgram;
			Reflect.set(resolvedMiniProgram, "__WEAPP_VITE_SESSION_METADATA", {
				port,
				projectPath: resolvedProjectPath,
				wsEndpoint: `ws://127.0.0.1:${port}`
			});
			releasePortLeaseOnExit = !retainPortLeaseUntilSessionClose(resolvedMiniProgram, portLease);
			await sleep$1(5e3);
			return resolvedMiniProgram;
		} finally {
			if (releasePortLeaseOnExit) await portLease.release();
		}
	}
	async connect(options) {
		if (normalizePlatform(options.platform) === "swan") return await new SwanLauncher().connect(options);
		const miniProgram = await this.connectTool(options);
		await miniProgram.checkVersion();
		return miniProgram;
	}
	async extendProjectConfig(projectConfig, projectPath) {
		const projectConfigPath = path.resolve(projectPath, "project.config.json");
		const fs = await import("node:fs/promises");
		const raw = await fs.readFile(projectConfigPath, "utf8");
		const current = JSON.parse(raw);
		extendDeep(current, projectConfig);
		await fs.writeFile(projectConfigPath, JSON.stringify(current, null, 2), "utf8");
	}
	async connectTool(options) {
		try {
			return new MiniProgram(await Connection.create(options.wsEndpoint));
		} catch {
			throw new Error(`Failed connecting to ${options.wsEndpoint}, check if target project window is opened with automation enabled`);
		}
	}
	async resolveCliPath() {
		const fs = await import("node:fs/promises");
		const cliPath = isWindows ? "C:/Program Files (x86)/Tencent/微信web开发者工具/cli.bat" : "/Applications/wechatwebdevtools.app/Contents/MacOS/cli";
		try {
			await fs.access(cliPath);
			return cliPath;
		} catch {
			return "";
		}
	}
};
//#endregion
//#region src/Automator.ts
/** Automator 的实现。 */
var Automator = class {
	Element = Element;
	CustomElement = CustomElement;
	InputElement = InputElement;
	ScrollViewElement = ScrollViewElement;
	SwiperElement = SwiperElement;
	MovableViewElement = MovableViewElement;
	SwitchElement = SwitchElement;
	SliderElement = SliderElement;
	ContextElement = ContextElement;
	launcher = new Launcher();
	async connect(options) {
		return await this.launcher.connect(options);
	}
	async launch(options) {
		return await this.launcher.launch(options);
	}
};
//#endregion
//#region src/index.ts
/**
* @file miniprogram-automator 对外导出入口。
*/
const SmartappAutomator = smartappAutomatorRuntime;
//#endregion
export { Automator, Connection, ContextElement, CustomElement, Element, InputElement, Launcher, MiniProgram, MovableViewElement, Native, Page, ScrollViewElement, SliderElement, SmartappAutomator, SwiperElement, SwitchElement, TextareaElement, Transport, decodeQrCode, extractPluginId, isPluginPath, normalizePlatform, printQrCode };
