// @weapp-vite/qr/dist/index.mjs
import { Buffer } from "node:buffer";
import { readFile } from "node:fs/promises";
import sharp from "sharp";
//#region src/reader/core/result.ts
/** normalizeQrCodeReaderResult 的方法封装。 */
function normalizeQrCodeReaderResult(value) {
	if (value && typeof value === "object" && "result" in value && typeof value.result === "string") return value;
	throw new TypeError("Invalid QR code reader result");
}
//#endregion
//#region src/reader/parse/bitmat.ts
/**
* @file 二维码解析内部模块：bitmat。
*/
var BitMatrix = class {
	width;
	height;
	rowSize;
	bits;
	constructor(width, height) {
		const resolvedHeight = height ?? width;
		if (width < 1 || resolvedHeight < 1) throw new Error("Both dimensions must be greater than 0");
		this.width = width;
		this.height = resolvedHeight;
		let rowSize = width >> 5;
		if ((width & 31) !== 0) rowSize++;
		this.rowSize = rowSize;
		this.bits = Array.from({ length: rowSize * resolvedHeight }, () => 0);
	}
	get Dimension() {
		if (this.width !== this.height) throw new Error("Can't call getDimension() on a non-square matrix");
		return this.width;
	}
	get_Renamed(x, y) {
		const offset = y * this.rowSize + (x >> 5);
		return (URShift(this.bits[offset], x & 31) & 1) !== 0;
	}
	set_Renamed(x, y) {
		const offset = y * this.rowSize + (x >> 5);
		this.bits[offset] |= 1 << (x & 31);
	}
	flip(x, y) {
		const offset = y * this.rowSize + (x >> 5);
		this.bits[offset] ^= 1 << (x & 31);
	}
	clear() {
		const max = this.bits.length;
		for (let i = 0; i < max; i++) this.bits[i] = 0;
	}
	setRegion(left, top, width, height) {
		if (top < 0 || left < 0) throw new Error("Left and top must be nonnegative");
		if (height < 1 || width < 1) throw new Error("Height and width must be at least 1");
		const right = left + width;
		const bottom = top + height;
		if (bottom > this.height || right > this.width) throw new Error("The region must fit inside the matrix");
		for (let y = top; y < bottom; y++) {
			const offset = y * this.rowSize;
			for (let x = left; x < right; x++) this.bits[offset + (x >> 5)] |= 1 << (x & 31);
		}
	}
};
//#endregion
//#region src/reader/parse/errorlevel.ts
/**
* @file 二维码解析内部模块：errorlevel。
*/
const FOR_BITS = [];
var ErrorCorrectionLevel = class {
	static forBits(bits) {
		if (bits < 0 || bits >= FOR_BITS.length) throw new Error("ArgumentException");
		return FOR_BITS[bits];
	}
	ordinal_Renamed_Field;
	bits;
	name;
	constructor(ordinal, bits, name) {
		this.ordinal_Renamed_Field = ordinal;
		this.bits = bits;
		this.name = name;
	}
	ordinal() {
		return this.ordinal_Renamed_Field;
	}
};
FOR_BITS.push(new ErrorCorrectionLevel(1, 0, "M"), new ErrorCorrectionLevel(0, 1, "L"), new ErrorCorrectionLevel(3, 2, "H"), new ErrorCorrectionLevel(2, 3, "Q"));
//#endregion
//#region src/reader/parse/formatinf.ts
/**
* @file 二维码解析内部模块：formatinf。
*/
const FORMAT_INFO_MASK_QR = 21522;
const FORMAT_INFO_DECODE_LOOKUP = [
	[21522, 0],
	[20773, 1],
	[24188, 2],
	[23371, 3],
	[17913, 4],
	[16590, 5],
	[20375, 6],
	[19104, 7],
	[30660, 8],
	[29427, 9],
	[32170, 10],
	[30877, 11],
	[26159, 12],
	[25368, 13],
	[27713, 14],
	[26998, 15],
	[5769, 16],
	[5054, 17],
	[7399, 18],
	[6608, 19],
	[1890, 20],
	[597, 21],
	[3340, 22],
	[2107, 23],
	[13663, 24],
	[12392, 25],
	[16177, 26],
	[14854, 27],
	[9396, 28],
	[8579, 29],
	[11994, 30],
	[11245, 31]
];
const BITS_SET_IN_HALF_BYTE = [
	0,
	1,
	1,
	2,
	1,
	2,
	2,
	3,
	1,
	2,
	2,
	3,
	2,
	3,
	3,
	4
];
var FormatInformation = class FormatInformation {
	static numBitsDiffering(a, b) {
		a ^= b;
		return BITS_SET_IN_HALF_BYTE[a & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 4) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 8) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 12) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 16) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 20) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 24) & 15] + BITS_SET_IN_HALF_BYTE[URShift(a, 28) & 15];
	}
	static decodeFormatInformation(maskedFormatInfo) {
		const formatInfo = FormatInformation.doDecodeFormatInformation(maskedFormatInfo);
		if (formatInfo != null) return formatInfo;
		return FormatInformation.doDecodeFormatInformation(maskedFormatInfo ^ FORMAT_INFO_MASK_QR);
	}
	static doDecodeFormatInformation(maskedFormatInfo) {
		let bestDifference = 4294967295;
		let bestFormatInfo = 0;
		for (let i = 0; i < FORMAT_INFO_DECODE_LOOKUP.length; i++) {
			const decodeInfo = FORMAT_INFO_DECODE_LOOKUP[i];
			const targetInfo = decodeInfo[0];
			if (targetInfo === maskedFormatInfo) return new FormatInformation(decodeInfo[1]);
			const bitsDifference = this.numBitsDiffering(maskedFormatInfo, targetInfo);
			if (bitsDifference < bestDifference) {
				bestFormatInfo = decodeInfo[1];
				bestDifference = bitsDifference;
			}
		}
		if (bestDifference <= 3) return new FormatInformation(bestFormatInfo);
		return null;
	}
	errorCorrectionLevel;
	dataMask;
	constructor(formatInfo) {
		this.errorCorrectionLevel = ErrorCorrectionLevel.forBits(formatInfo >> 3 & 3);
		this.dataMask = formatInfo & 7;
	}
	GetHashCode() {
		return this.errorCorrectionLevel.ordinal() << 3 | this.dataMask;
	}
	Equals(o) {
		return this.errorCorrectionLevel === o.errorCorrectionLevel && this.dataMask === o.dataMask;
	}
};
//#endregion
//#region src/reader/parse/version.ts
/**
* @file 二维码解析内部模块：version。
*/
var ECB = class {
	count;
	dataCodewords;
	constructor(count, dataCodewords) {
		this.count = count;
		this.dataCodewords = dataCodewords;
	}
};
var ECBlocks = class {
	ecCodewordsPerBlock;
	ecBlocks;
	constructor(ecCodewordsPerBlock, ecBlocks1, ecBlocks2) {
		this.ecCodewordsPerBlock = ecCodewordsPerBlock;
		this.ecBlocks = ecBlocks2 ? [ecBlocks1, ecBlocks2] : [ecBlocks1];
	}
	get TotalECCodewords() {
		return this.ecCodewordsPerBlock * this.NumBlocks;
	}
	get NumBlocks() {
		let total = 0;
		for (let i = 0; i < this.ecBlocks.length; i++) total += this.ecBlocks[i].count;
		return total;
	}
	getECBlocks() {
		return this.ecBlocks;
	}
};
var Version = class Version {
	static VERSION_DECODE_INFO = [
		31892,
		34236,
		39577,
		42195,
		48118,
		51042,
		55367,
		58893,
		63784,
		68472,
		70749,
		76311,
		79154,
		84390,
		87683,
		92361,
		96236,
		102084,
		102881,
		110507,
		110734,
		117786,
		119615,
		126325,
		127568,
		133589,
		136944,
		141498,
		145311,
		150283,
		152622,
		158308,
		161089,
		167017
	];
	static VERSIONS;
	static getVersionForNumber(versionNumber) {
		if (versionNumber < 1 || versionNumber > 40) throw new Error("ArgumentException");
		return Version.VERSIONS[versionNumber - 1];
	}
	static getProvisionalVersionForDimension(dimension) {
		if (dimension % 4 !== 1) throw new Error("Error getProvisionalVersionForDimension");
		try {
			return Version.getVersionForNumber(dimension - 17 >> 2);
		} catch {
			throw new Error("Error getVersionForNumber");
		}
	}
	static decodeVersionInformation(versionBits) {
		let bestDifference = 4294967295;
		let bestVersion = 0;
		for (let i = 0; i < Version.VERSION_DECODE_INFO.length; i++) {
			const targetVersion = Version.VERSION_DECODE_INFO[i];
			if (targetVersion === versionBits) return Version.getVersionForNumber(i + 7);
			const bitsDifference = FormatInformation.numBitsDiffering(versionBits, targetVersion);
			if (bitsDifference < bestDifference) {
				bestVersion = i + 7;
				bestDifference = bitsDifference;
			}
		}
		if (bestDifference <= 3) return Version.getVersionForNumber(bestVersion);
		return null;
	}
	versionNumber;
	alignmentPatternCenters;
	ecBlocks;
	totalCodewords;
	constructor(versionNumber, alignmentPatternCenters, ecBlocks1, ecBlocks2, ecBlocks3, ecBlocks4) {
		this.versionNumber = versionNumber;
		this.alignmentPatternCenters = alignmentPatternCenters;
		this.ecBlocks = [
			ecBlocks1,
			ecBlocks2,
			ecBlocks3,
			ecBlocks4
		];
		let total = 0;
		const ecCodewords = ecBlocks1.ecCodewordsPerBlock;
		const ecbArray = ecBlocks1.getECBlocks();
		for (let i = 0; i < ecbArray.length; i++) {
			const ecBlock = ecbArray[i];
			total += ecBlock.count * (ecBlock.dataCodewords + ecCodewords);
		}
		this.totalCodewords = total;
	}
	get DimensionForVersion() {
		return 17 + 4 * this.versionNumber;
	}
	buildFunctionPattern() {
		const dimension = this.DimensionForVersion;
		const bitMatrix = new BitMatrix(dimension);
		bitMatrix.setRegion(0, 0, 9, 9);
		bitMatrix.setRegion(dimension - 8, 0, 8, 9);
		bitMatrix.setRegion(0, dimension - 8, 9, 8);
		const max = this.alignmentPatternCenters.length;
		for (let x = 0; x < max; x++) {
			const i = this.alignmentPatternCenters[x] - 2;
			for (let y = 0; y < max; y++) {
				if (x === 0 && (y === 0 || y === max - 1) || x === max - 1 && y === 0) continue;
				bitMatrix.setRegion(this.alignmentPatternCenters[y] - 2, i, 5, 5);
			}
		}
		bitMatrix.setRegion(6, 9, 1, dimension - 17);
		bitMatrix.setRegion(9, 6, dimension - 17, 1);
		if (this.versionNumber > 6) {
			bitMatrix.setRegion(dimension - 11, 0, 3, 6);
			bitMatrix.setRegion(0, dimension - 11, 6, 3);
		}
		return bitMatrix;
	}
	getECBlocksForLevel(ecLevel) {
		return this.ecBlocks[ecLevel.ordinal()];
	}
};
function buildVersions() {
	return [
		new Version(1, [], new ECBlocks(7, new ECB(1, 19)), new ECBlocks(10, new ECB(1, 16)), new ECBlocks(13, new ECB(1, 13)), new ECBlocks(17, new ECB(1, 9))),
		new Version(2, [6, 18], new ECBlocks(10, new ECB(1, 34)), new ECBlocks(16, new ECB(1, 28)), new ECBlocks(22, new ECB(1, 22)), new ECBlocks(28, new ECB(1, 16))),
		new Version(3, [6, 22], new ECBlocks(15, new ECB(1, 55)), new ECBlocks(26, new ECB(1, 44)), new ECBlocks(18, new ECB(2, 17)), new ECBlocks(22, new ECB(2, 13))),
		new Version(4, [6, 26], new ECBlocks(20, new ECB(1, 80)), new ECBlocks(18, new ECB(2, 32)), new ECBlocks(26, new ECB(2, 24)), new ECBlocks(16, new ECB(4, 9))),
		new Version(5, [6, 30], new ECBlocks(26, new ECB(1, 108)), new ECBlocks(24, new ECB(2, 43)), new ECBlocks(18, new ECB(2, 15), new ECB(2, 16)), new ECBlocks(22, new ECB(2, 11), new ECB(2, 12))),
		new Version(6, [6, 34], new ECBlocks(18, new ECB(2, 68)), new ECBlocks(16, new ECB(4, 27)), new ECBlocks(24, new ECB(4, 19)), new ECBlocks(28, new ECB(4, 15))),
		new Version(7, [
			6,
			22,
			38
		], new ECBlocks(20, new ECB(2, 78)), new ECBlocks(18, new ECB(4, 31)), new ECBlocks(18, new ECB(2, 14), new ECB(4, 15)), new ECBlocks(26, new ECB(4, 13), new ECB(1, 14))),
		new Version(8, [
			6,
			24,
			42
		], new ECBlocks(24, new ECB(2, 97)), new ECBlocks(22, new ECB(2, 38), new ECB(2, 39)), new ECBlocks(22, new ECB(4, 18), new ECB(2, 19)), new ECBlocks(26, new ECB(4, 14), new ECB(2, 15))),
		new Version(9, [
			6,
			26,
			46
		], new ECBlocks(30, new ECB(2, 116)), new ECBlocks(22, new ECB(3, 36), new ECB(2, 37)), new ECBlocks(20, new ECB(4, 16), new ECB(4, 17)), new ECBlocks(24, new ECB(4, 12), new ECB(4, 13))),
		new Version(10, [
			6,
			28,
			50
		], new ECBlocks(18, new ECB(2, 68), new ECB(2, 69)), new ECBlocks(26, new ECB(4, 43), new ECB(1, 44)), new ECBlocks(24, new ECB(6, 19), new ECB(2, 20)), new ECBlocks(28, new ECB(6, 15), new ECB(2, 16))),
		new Version(11, [
			6,
			30,
			54
		], new ECBlocks(20, new ECB(4, 81)), new ECBlocks(30, new ECB(1, 50), new ECB(4, 51)), new ECBlocks(28, new ECB(4, 22), new ECB(4, 23)), new ECBlocks(24, new ECB(3, 12), new ECB(8, 13))),
		new Version(12, [
			6,
			32,
			58
		], new ECBlocks(24, new ECB(2, 92), new ECB(2, 93)), new ECBlocks(22, new ECB(6, 36), new ECB(2, 37)), new ECBlocks(26, new ECB(4, 20), new ECB(6, 21)), new ECBlocks(28, new ECB(7, 14), new ECB(4, 15))),
		new Version(13, [
			6,
			34,
			62
		], new ECBlocks(26, new ECB(4, 107)), new ECBlocks(22, new ECB(8, 37), new ECB(1, 38)), new ECBlocks(24, new ECB(8, 20), new ECB(4, 21)), new ECBlocks(22, new ECB(12, 11), new ECB(4, 12))),
		new Version(14, [
			6,
			26,
			46,
			66
		], new ECBlocks(30, new ECB(3, 115), new ECB(1, 116)), new ECBlocks(24, new ECB(4, 40), new ECB(5, 41)), new ECBlocks(20, new ECB(11, 16), new ECB(5, 17)), new ECBlocks(24, new ECB(11, 12), new ECB(5, 13))),
		new Version(15, [
			6,
			26,
			48,
			70
		], new ECBlocks(22, new ECB(5, 87), new ECB(1, 88)), new ECBlocks(24, new ECB(5, 41), new ECB(5, 42)), new ECBlocks(30, new ECB(5, 24), new ECB(7, 25)), new ECBlocks(24, new ECB(11, 12), new ECB(7, 13))),
		new Version(16, [
			6,
			26,
			50,
			74
		], new ECBlocks(24, new ECB(5, 98), new ECB(1, 99)), new ECBlocks(28, new ECB(7, 45), new ECB(3, 46)), new ECBlocks(24, new ECB(15, 19), new ECB(2, 20)), new ECBlocks(30, new ECB(3, 15), new ECB(13, 16))),
		new Version(17, [
			6,
			30,
			54,
			78
		], new ECBlocks(28, new ECB(1, 107), new ECB(5, 108)), new ECBlocks(28, new ECB(10, 46), new ECB(1, 47)), new ECBlocks(28, new ECB(1, 22), new ECB(15, 23)), new ECBlocks(28, new ECB(2, 14), new ECB(17, 15))),
		new Version(18, [
			6,
			30,
			56,
			82
		], new ECBlocks(30, new ECB(5, 120), new ECB(1, 121)), new ECBlocks(26, new ECB(9, 43), new ECB(4, 44)), new ECBlocks(28, new ECB(17, 22), new ECB(1, 23)), new ECBlocks(28, new ECB(2, 14), new ECB(19, 15))),
		new Version(19, [
			6,
			30,
			58,
			86
		], new ECBlocks(28, new ECB(3, 113), new ECB(4, 114)), new ECBlocks(26, new ECB(3, 44), new ECB(11, 45)), new ECBlocks(26, new ECB(17, 21), new ECB(4, 22)), new ECBlocks(26, new ECB(9, 13), new ECB(16, 14))),
		new Version(20, [
			6,
			34,
			62,
			90
		], new ECBlocks(28, new ECB(3, 107), new ECB(5, 108)), new ECBlocks(26, new ECB(3, 41), new ECB(13, 42)), new ECBlocks(30, new ECB(15, 24), new ECB(5, 25)), new ECBlocks(28, new ECB(15, 15), new ECB(10, 16))),
		new Version(21, [
			6,
			28,
			50,
			72,
			94
		], new ECBlocks(28, new ECB(4, 116), new ECB(4, 117)), new ECBlocks(26, new ECB(17, 42)), new ECBlocks(28, new ECB(17, 22), new ECB(6, 23)), new ECBlocks(30, new ECB(19, 16), new ECB(6, 17))),
		new Version(22, [
			6,
			26,
			50,
			74,
			98
		], new ECBlocks(28, new ECB(2, 111), new ECB(7, 112)), new ECBlocks(28, new ECB(17, 46)), new ECBlocks(30, new ECB(7, 24), new ECB(16, 25)), new ECBlocks(24, new ECB(34, 13))),
		new Version(23, [
			6,
			30,
			54,
			74,
			102
		], new ECBlocks(30, new ECB(4, 121), new ECB(5, 122)), new ECBlocks(28, new ECB(4, 47), new ECB(14, 48)), new ECBlocks(30, new ECB(11, 24), new ECB(14, 25)), new ECBlocks(30, new ECB(16, 15), new ECB(14, 16))),
		new Version(24, [
			6,
			28,
			54,
			80,
			106
		], new ECBlocks(30, new ECB(6, 117), new ECB(4, 118)), new ECBlocks(28, new ECB(6, 45), new ECB(14, 46)), new ECBlocks(30, new ECB(11, 24), new ECB(16, 25)), new ECBlocks(30, new ECB(30, 16), new ECB(2, 17))),
		new Version(25, [
			6,
			32,
			58,
			84,
			110
		], new ECBlocks(26, new ECB(8, 106), new ECB(4, 107)), new ECBlocks(28, new ECB(8, 47), new ECB(13, 48)), new ECBlocks(30, new ECB(7, 24), new ECB(22, 25)), new ECBlocks(30, new ECB(22, 15), new ECB(13, 16))),
		new Version(26, [
			6,
			30,
			58,
			86,
			114
		], new ECBlocks(28, new ECB(10, 114), new ECB(2, 115)), new ECBlocks(28, new ECB(19, 46), new ECB(4, 47)), new ECBlocks(28, new ECB(28, 22), new ECB(6, 23)), new ECBlocks(30, new ECB(33, 16), new ECB(4, 17))),
		new Version(27, [
			6,
			34,
			62,
			90,
			118
		], new ECBlocks(30, new ECB(8, 122), new ECB(4, 123)), new ECBlocks(28, new ECB(22, 45), new ECB(3, 46)), new ECBlocks(30, new ECB(8, 23), new ECB(26, 24)), new ECBlocks(30, new ECB(12, 15), new ECB(28, 16))),
		new Version(28, [
			6,
			26,
			50,
			74,
			98,
			122
		], new ECBlocks(30, new ECB(3, 117), new ECB(10, 118)), new ECBlocks(28, new ECB(3, 45), new ECB(23, 46)), new ECBlocks(30, new ECB(4, 24), new ECB(31, 25)), new ECBlocks(30, new ECB(11, 15), new ECB(31, 16))),
		new Version(29, [
			6,
			30,
			54,
			78,
			102,
			126
		], new ECBlocks(30, new ECB(7, 116), new ECB(7, 117)), new ECBlocks(28, new ECB(21, 45), new ECB(7, 46)), new ECBlocks(30, new ECB(1, 23), new ECB(37, 24)), new ECBlocks(30, new ECB(19, 15), new ECB(26, 16))),
		new Version(30, [
			6,
			26,
			52,
			78,
			104,
			130
		], new ECBlocks(30, new ECB(5, 115), new ECB(10, 116)), new ECBlocks(28, new ECB(19, 47), new ECB(10, 48)), new ECBlocks(30, new ECB(15, 24), new ECB(25, 25)), new ECBlocks(30, new ECB(23, 15), new ECB(25, 16))),
		new Version(31, [
			6,
			30,
			56,
			82,
			108,
			134
		], new ECBlocks(30, new ECB(13, 115), new ECB(3, 116)), new ECBlocks(28, new ECB(2, 46), new ECB(29, 47)), new ECBlocks(30, new ECB(42, 24), new ECB(1, 25)), new ECBlocks(30, new ECB(23, 15), new ECB(28, 16))),
		new Version(32, [
			6,
			34,
			60,
			86,
			112,
			138
		], new ECBlocks(30, new ECB(17, 115)), new ECBlocks(28, new ECB(10, 46), new ECB(23, 47)), new ECBlocks(30, new ECB(10, 24), new ECB(35, 25)), new ECBlocks(30, new ECB(19, 15), new ECB(35, 16))),
		new Version(33, [
			6,
			30,
			58,
			86,
			114,
			142
		], new ECBlocks(30, new ECB(17, 115), new ECB(1, 116)), new ECBlocks(28, new ECB(14, 46), new ECB(21, 47)), new ECBlocks(30, new ECB(29, 24), new ECB(19, 25)), new ECBlocks(30, new ECB(11, 15), new ECB(46, 16))),
		new Version(34, [
			6,
			34,
			62,
			90,
			118,
			146
		], new ECBlocks(30, new ECB(13, 115), new ECB(6, 116)), new ECBlocks(28, new ECB(14, 46), new ECB(23, 47)), new ECBlocks(30, new ECB(44, 24), new ECB(7, 25)), new ECBlocks(30, new ECB(59, 16), new ECB(1, 17))),
		new Version(35, [
			6,
			30,
			54,
			78,
			102,
			126,
			150
		], new ECBlocks(30, new ECB(12, 121), new ECB(7, 122)), new ECBlocks(28, new ECB(12, 47), new ECB(26, 48)), new ECBlocks(30, new ECB(39, 24), new ECB(14, 25)), new ECBlocks(30, new ECB(22, 15), new ECB(41, 16))),
		new Version(36, [
			6,
			24,
			50,
			76,
			102,
			128,
			154
		], new ECBlocks(30, new ECB(6, 121), new ECB(14, 122)), new ECBlocks(28, new ECB(6, 47), new ECB(34, 48)), new ECBlocks(30, new ECB(46, 24), new ECB(10, 25)), new ECBlocks(30, new ECB(2, 15), new ECB(64, 16))),
		new Version(37, [
			6,
			28,
			54,
			80,
			106,
			132,
			158
		], new ECBlocks(30, new ECB(17, 122), new ECB(4, 123)), new ECBlocks(28, new ECB(29, 46), new ECB(14, 47)), new ECBlocks(30, new ECB(49, 24), new ECB(10, 25)), new ECBlocks(30, new ECB(24, 15), new ECB(46, 16))),
		new Version(38, [
			6,
			32,
			58,
			84,
			110,
			136,
			162
		], new ECBlocks(30, new ECB(4, 122), new ECB(18, 123)), new ECBlocks(28, new ECB(13, 46), new ECB(32, 47)), new ECBlocks(30, new ECB(48, 24), new ECB(14, 25)), new ECBlocks(30, new ECB(42, 15), new ECB(32, 16))),
		new Version(39, [
			6,
			26,
			54,
			82,
			110,
			138,
			166
		], new ECBlocks(30, new ECB(20, 117), new ECB(4, 118)), new ECBlocks(28, new ECB(40, 47), new ECB(7, 48)), new ECBlocks(30, new ECB(43, 24), new ECB(22, 25)), new ECBlocks(30, new ECB(10, 15), new ECB(67, 16))),
		new Version(40, [
			6,
			30,
			58,
			86,
			114,
			142,
			170
		], new ECBlocks(30, new ECB(19, 118), new ECB(6, 119)), new ECBlocks(28, new ECB(18, 47), new ECB(31, 48)), new ECBlocks(30, new ECB(34, 24), new ECB(34, 25)), new ECBlocks(30, new ECB(20, 15), new ECB(61, 16)))
	];
}
Version.VERSIONS = buildVersions();
//#endregion
//#region src/reader/detect/alignpat.ts
var AlignmentPattern = class {
	x;
	y;
	count;
	estimatedModuleSize;
	constructor(posX, posY, estimatedModuleSize) {
		this.x = posX;
		this.y = posY;
		this.count = 1;
		this.estimatedModuleSize = estimatedModuleSize;
	}
	get X() {
		return Math.floor(this.x);
	}
	get Y() {
		return Math.floor(this.y);
	}
	incrementCount() {
		this.count++;
	}
	aboutEquals(moduleSize, i, j) {
		if (Math.abs(i - this.y) <= moduleSize && Math.abs(j - this.x) <= moduleSize) {
			const moduleSizeDiff = Math.abs(moduleSize - this.estimatedModuleSize);
			return moduleSizeDiff <= 1 || moduleSizeDiff / this.estimatedModuleSize <= 1;
		}
		return false;
	}
};
var AlignmentPatternFinder = class {
	image;
	possibleCenters;
	startX;
	startY;
	width;
	height;
	moduleSize;
	crossCheckStateCount;
	resultPointCallback;
	constructor(image, startX, startY, width, height, moduleSize, resultPointCallback) {
		this.image = image;
		this.possibleCenters = [];
		this.startX = startX;
		this.startY = startY;
		this.width = width;
		this.height = height;
		this.moduleSize = moduleSize;
		this.crossCheckStateCount = [
			0,
			0,
			0
		];
		this.resultPointCallback = resultPointCallback;
	}
	centerFromEnd(stateCount, end) {
		return end - stateCount[2] - stateCount[1] / 2;
	}
	foundPatternCross(stateCount) {
		const maxVariance = this.moduleSize / 2;
		for (let i = 0; i < 3; i++) if (Math.abs(this.moduleSize - stateCount[i]) >= maxVariance) return false;
		return true;
	}
	crossCheckVertical(startI, centerJ, maxCount, originalStateCountTotal) {
		const maxI = this.image.height;
		const stateCount = this.crossCheckStateCount;
		stateCount[0] = 0;
		stateCount[1] = 0;
		stateCount[2] = 0;
		let i = startI;
		while (i >= 0 && this.image.data[centerJ + i * this.image.width] && stateCount[1] <= maxCount) {
			stateCount[1]++;
			i--;
		}
		if (i < 0 || stateCount[1] > maxCount) return NaN;
		while (i >= 0 && !this.image.data[centerJ + i * this.image.width] && stateCount[0] <= maxCount) {
			stateCount[0]++;
			i--;
		}
		if (stateCount[0] > maxCount) return NaN;
		i = startI + 1;
		while (i < maxI && this.image.data[centerJ + i * this.image.width] && stateCount[1] <= maxCount) {
			stateCount[1]++;
			i++;
		}
		if (i === maxI || stateCount[1] > maxCount) return NaN;
		while (i < maxI && !this.image.data[centerJ + i * this.image.width] && stateCount[2] <= maxCount) {
			stateCount[2]++;
			i++;
		}
		if (stateCount[2] > maxCount) return NaN;
		const stateCountTotal = stateCount[0] + stateCount[1] + stateCount[2];
		if (5 * Math.abs(stateCountTotal - originalStateCountTotal) >= 2 * originalStateCountTotal) return NaN;
		return this.foundPatternCross(stateCount) ? this.centerFromEnd(stateCount, i) : NaN;
	}
	handlePossibleCenter(stateCount, i, j) {
		const stateCountTotal = stateCount[0] + stateCount[1] + stateCount[2];
		const centerJ = this.centerFromEnd(stateCount, j);
		const centerI = this.crossCheckVertical(i, Math.floor(centerJ), 2 * stateCount[1], stateCountTotal);
		if (!Number.isNaN(centerI)) {
			const estimatedModuleSize = stateCountTotal / 3;
			for (let index = 0; index < this.possibleCenters.length; index++) if (this.possibleCenters[index].aboutEquals(estimatedModuleSize, centerI, centerJ)) return new AlignmentPattern(centerJ, centerI, estimatedModuleSize);
			const point = new AlignmentPattern(centerJ, centerI, estimatedModuleSize);
			this.possibleCenters.push(point);
			this.resultPointCallback?.foundPossibleResultPoint?.(point);
			return point;
		}
		return null;
	}
	find() {
		const startX = this.startX;
		const maxJ = startX + this.width;
		const middleI = this.startY + (this.height >> 1);
		const stateCount = [
			0,
			0,
			0
		];
		for (let iGen = 0; iGen < this.height; iGen++) {
			const i = middleI + ((iGen & 1) === 0 ? iGen + 1 >> 1 : -(iGen + 1 >> 1));
			stateCount[0] = 0;
			stateCount[1] = 0;
			stateCount[2] = 0;
			let j = startX;
			while (j < maxJ && !this.image.data[j + this.image.width * i]) j++;
			let currentState = 0;
			while (j < maxJ) {
				if (this.image.data[j + i * this.image.width]) if (currentState === 1) stateCount[currentState]++;
				else if (currentState === 2) {
					if (this.foundPatternCross(stateCount)) {
						const confirmed = this.handlePossibleCenter(stateCount, i, j);
						if (confirmed != null) return confirmed;
					}
					stateCount[0] = stateCount[2];
					stateCount[1] = 1;
					stateCount[2] = 0;
					currentState = 1;
				} else stateCount[++currentState]++;
				else {
					if (currentState === 1) currentState++;
					stateCount[currentState]++;
				}
				j++;
			}
			if (this.foundPatternCross(stateCount)) {
				const confirmed = this.handlePossibleCenter(stateCount, i, maxJ);
				if (confirmed != null) return confirmed;
			}
		}
		if (this.possibleCenters.length > 0) return this.possibleCenters[0];
		throw new Error("Couldn't find enough alignment patterns");
	}
};
//#endregion
//#region src/reader/detect/findpat.ts
/**
* @file 二维码解析内部模块：findpat。
*/
const MIN_SKIP = 3;
const MAX_MODULES = 57;
const INTEGER_MATH_SHIFT = 8;
const CENTER_QUORUM = 2;
function orderBestPatterns(patterns) {
	function distance(pattern1, pattern2) {
		const xDiff = pattern1.X - pattern2.X;
		const yDiff = pattern1.Y - pattern2.Y;
		return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
	}
	function crossProductZ(pointA, pointB, pointC) {
		const bX = pointB.x;
		const bY = pointB.y;
		return (pointC.x - bX) * (pointA.y - bY) - (pointC.y - bY) * (pointA.x - bX);
	}
	const zeroOneDistance = distance(patterns[0], patterns[1]);
	const oneTwoDistance = distance(patterns[1], patterns[2]);
	const zeroTwoDistance = distance(patterns[0], patterns[2]);
	let pointA;
	let pointB;
	let pointC;
	if (oneTwoDistance >= zeroOneDistance && oneTwoDistance >= zeroTwoDistance) {
		pointB = patterns[0];
		pointA = patterns[1];
		pointC = patterns[2];
	} else if (zeroTwoDistance >= oneTwoDistance && zeroTwoDistance >= zeroOneDistance) {
		pointB = patterns[1];
		pointA = patterns[0];
		pointC = patterns[2];
	} else {
		pointB = patterns[2];
		pointA = patterns[0];
		pointC = patterns[1];
	}
	if (crossProductZ(pointA, pointB, pointC) < 0) {
		const temp = pointA;
		pointA = pointC;
		pointC = temp;
	}
	patterns[0] = pointA;
	patterns[1] = pointB;
	patterns[2] = pointC;
}
var FinderPattern = class {
	x;
	y;
	count;
	estimatedModuleSize;
	constructor(posX, posY, estimatedModuleSize) {
		this.x = posX;
		this.y = posY;
		this.count = 1;
		this.estimatedModuleSize = estimatedModuleSize;
	}
	get X() {
		return this.x;
	}
	get Y() {
		return this.y;
	}
	incrementCount() {
		this.count++;
	}
	aboutEquals(moduleSize, i, j) {
		if (Math.abs(i - this.y) <= moduleSize && Math.abs(j - this.x) <= moduleSize) {
			const moduleSizeDiff = Math.abs(moduleSize - this.estimatedModuleSize);
			return moduleSizeDiff <= 1 || moduleSizeDiff / this.estimatedModuleSize <= 1;
		}
		return false;
	}
};
var FinderPatternInfo = class {
	bottomLeft;
	topLeft;
	topRight;
	constructor(patternCenters) {
		this.bottomLeft = patternCenters[0];
		this.topLeft = patternCenters[1];
		this.topRight = patternCenters[2];
	}
};
var FinderPatternFinder = class {
	image;
	possibleCenters;
	hasSkipped;
	crossCheckStateCount;
	resultPointCallback;
	constructor() {
		this.image = null;
		this.possibleCenters = [];
		this.hasSkipped = false;
		this.crossCheckStateCount = [
			0,
			0,
			0,
			0,
			0
		];
		this.resultPointCallback = null;
	}
	get CrossCheckStateCount() {
		this.crossCheckStateCount[0] = 0;
		this.crossCheckStateCount[1] = 0;
		this.crossCheckStateCount[2] = 0;
		this.crossCheckStateCount[3] = 0;
		this.crossCheckStateCount[4] = 0;
		return this.crossCheckStateCount;
	}
	foundPatternCross(stateCount) {
		let totalModuleSize = 0;
		for (let i = 0; i < 5; i++) {
			const count = stateCount[i];
			if (count === 0) return false;
			totalModuleSize += count;
		}
		if (totalModuleSize < 7) return false;
		const moduleSize = Math.floor((totalModuleSize << INTEGER_MATH_SHIFT) / 7);
		const maxVariance = Math.floor(moduleSize / 2);
		return Math.abs(moduleSize - (stateCount[0] << INTEGER_MATH_SHIFT)) < maxVariance && Math.abs(moduleSize - (stateCount[1] << INTEGER_MATH_SHIFT)) < maxVariance && Math.abs(3 * moduleSize - (stateCount[2] << INTEGER_MATH_SHIFT)) < 3 * maxVariance && Math.abs(moduleSize - (stateCount[3] << INTEGER_MATH_SHIFT)) < maxVariance && Math.abs(moduleSize - (stateCount[4] << INTEGER_MATH_SHIFT)) < maxVariance;
	}
	centerFromEnd(stateCount, end) {
		return end - stateCount[4] - stateCount[3] - stateCount[2] / 2;
	}
	crossCheckVertical(startI, centerJ, maxCount, originalStateCountTotal) {
		if (this.image == null) throw new Error("Image is not initialized");
		const maxI = this.image.height;
		const stateCount = this.CrossCheckStateCount;
		let i = startI;
		while (i >= 0 && this.image.data[centerJ + i * this.image.width]) {
			stateCount[2]++;
			i--;
		}
		if (i < 0) return NaN;
		while (i >= 0 && !this.image.data[centerJ + i * this.image.width] && stateCount[1] <= maxCount) {
			stateCount[1]++;
			i--;
		}
		if (i < 0 || stateCount[1] > maxCount) return NaN;
		while (i >= 0 && this.image.data[centerJ + i * this.image.width] && stateCount[0] <= maxCount) {
			stateCount[0]++;
			i--;
		}
		if (stateCount[0] > maxCount) return NaN;
		i = startI + 1;
		while (i < maxI && this.image.data[centerJ + i * this.image.width]) {
			stateCount[2]++;
			i++;
		}
		if (i === maxI) return NaN;
		while (i < maxI && !this.image.data[centerJ + i * this.image.width] && stateCount[3] < maxCount) {
			stateCount[3]++;
			i++;
		}
		if (i === maxI || stateCount[3] >= maxCount) return NaN;
		while (i < maxI && this.image.data[centerJ + i * this.image.width] && stateCount[4] < maxCount) {
			stateCount[4]++;
			i++;
		}
		if (stateCount[4] >= maxCount) return NaN;
		const stateCountTotal = stateCount[0] + stateCount[1] + stateCount[2] + stateCount[3] + stateCount[4];
		if (5 * Math.abs(stateCountTotal - originalStateCountTotal) >= 2 * originalStateCountTotal) return NaN;
		return this.foundPatternCross(stateCount) ? this.centerFromEnd(stateCount, i) : NaN;
	}
	crossCheckHorizontal(startJ, centerI, maxCount, originalStateCountTotal) {
		if (this.image == null) throw new Error("Image is not initialized");
		const maxJ = this.image.width;
		const stateCount = this.CrossCheckStateCount;
		let j = startJ;
		while (j >= 0 && this.image.data[j + centerI * this.image.width]) {
			stateCount[2]++;
			j--;
		}
		if (j < 0) return NaN;
		while (j >= 0 && !this.image.data[j + centerI * this.image.width] && stateCount[1] <= maxCount) {
			stateCount[1]++;
			j--;
		}
		if (j < 0 || stateCount[1] > maxCount) return NaN;
		while (j >= 0 && this.image.data[j + centerI * this.image.width] && stateCount[0] <= maxCount) {
			stateCount[0]++;
			j--;
		}
		if (stateCount[0] > maxCount) return NaN;
		j = startJ + 1;
		while (j < maxJ && this.image.data[j + centerI * this.image.width]) {
			stateCount[2]++;
			j++;
		}
		if (j === maxJ) return NaN;
		while (j < maxJ && !this.image.data[j + centerI * this.image.width] && stateCount[3] < maxCount) {
			stateCount[3]++;
			j++;
		}
		if (j === maxJ || stateCount[3] >= maxCount) return NaN;
		while (j < maxJ && this.image.data[j + centerI * this.image.width] && stateCount[4] < maxCount) {
			stateCount[4]++;
			j++;
		}
		if (stateCount[4] >= maxCount) return NaN;
		const stateCountTotal = stateCount[0] + stateCount[1] + stateCount[2] + stateCount[3] + stateCount[4];
		if (5 * Math.abs(stateCountTotal - originalStateCountTotal) >= originalStateCountTotal) return NaN;
		return this.foundPatternCross(stateCount) ? this.centerFromEnd(stateCount, j) : NaN;
	}
	handlePossibleCenter(stateCount, i, j) {
		const stateCountTotal = stateCount[0] + stateCount[1] + stateCount[2] + stateCount[3] + stateCount[4];
		let centerJ = this.centerFromEnd(stateCount, j);
		const centerI = this.crossCheckVertical(i, Math.floor(centerJ), stateCount[2], stateCountTotal);
		if (!Number.isNaN(centerI)) {
			centerJ = this.crossCheckHorizontal(Math.floor(centerJ), Math.floor(centerI), stateCount[2], stateCountTotal);
			if (!Number.isNaN(centerJ)) {
				const estimatedModuleSize = stateCountTotal / 7;
				let found = false;
				for (let index = 0; index < this.possibleCenters.length; index++) {
					const center = this.possibleCenters[index];
					if (center.aboutEquals(estimatedModuleSize, centerI, centerJ)) {
						center.incrementCount();
						found = true;
						break;
					}
				}
				if (!found) {
					const point = new FinderPattern(centerJ, centerI, estimatedModuleSize);
					this.possibleCenters.push(point);
					this.resultPointCallback?.foundPossibleResultPoint?.(point);
				}
				return true;
			}
		}
		return false;
	}
	selectBestPatterns() {
		const startSize = this.possibleCenters.length;
		if (startSize < 3) throw new Error(`Couldn't find enough finder patterns:${startSize} patterns found`);
		if (startSize > 3) {
			let totalModuleSize = 0;
			let square = 0;
			for (let i = 0; i < startSize; i++) {
				const centerValue = this.possibleCenters[i].estimatedModuleSize;
				totalModuleSize += centerValue;
				square += centerValue * centerValue;
			}
			const average = totalModuleSize / startSize;
			this.possibleCenters.sort((center1, center2) => {
				const dA = Math.abs(center2.estimatedModuleSize - average);
				const dB = Math.abs(center1.estimatedModuleSize - average);
				if (dA < dB) return -1;
				if (dA === dB) return 0;
				return 1;
			});
			const stdDev = Math.sqrt(square / startSize - average * average);
			const limit = Math.max(.2 * average, stdDev);
			for (let i = this.possibleCenters.length - 1; i >= 0; i--) {
				const pattern = this.possibleCenters[i];
				if (Math.abs(pattern.estimatedModuleSize - average) > limit) this.possibleCenters.splice(i, 1);
			}
		}
		if (this.possibleCenters.length > 3) this.possibleCenters.sort((a, b) => {
			if (a.count > b.count) return -1;
			if (a.count < b.count) return 1;
			return 0;
		});
		return [
			this.possibleCenters[0],
			this.possibleCenters[1],
			this.possibleCenters[2]
		];
	}
	findRowSkip() {
		if (this.possibleCenters.length <= 1) return 0;
		let firstConfirmedCenter = null;
		for (let i = 0; i < this.possibleCenters.length; i++) {
			const center = this.possibleCenters[i];
			if (center.count >= CENTER_QUORUM) if (firstConfirmedCenter == null) firstConfirmedCenter = center;
			else {
				this.hasSkipped = true;
				return Math.floor((Math.abs(firstConfirmedCenter.X - center.X) - Math.abs(firstConfirmedCenter.Y - center.Y)) / 2);
			}
		}
		return 0;
	}
	haveMultiplyConfirmedCenters() {
		let confirmedCount = 0;
		let totalModuleSize = 0;
		for (let i = 0; i < this.possibleCenters.length; i++) {
			const pattern = this.possibleCenters[i];
			if (pattern.count >= CENTER_QUORUM) {
				confirmedCount++;
				totalModuleSize += pattern.estimatedModuleSize;
			}
		}
		if (confirmedCount < 3) return false;
		const average = totalModuleSize / this.possibleCenters.length;
		let totalDeviation = 0;
		for (let i = 0; i < this.possibleCenters.length; i++) {
			const pattern = this.possibleCenters[i];
			totalDeviation += Math.abs(pattern.estimatedModuleSize - average);
		}
		return totalDeviation <= .05 * totalModuleSize;
	}
	findFinderPattern(image) {
		const tryHarder = false;
		this.image = image;
		const maxI = image.height;
		const maxJ = image.width;
		let iSkip = Math.floor(3 * maxI / (4 * MAX_MODULES));
		if (iSkip < MIN_SKIP || tryHarder) iSkip = MIN_SKIP;
		let done = false;
		const stateCount = [
			0,
			0,
			0,
			0,
			0
		];
		for (let i = iSkip - 1; i < maxI && !done; i += iSkip) {
			stateCount[0] = 0;
			stateCount[1] = 0;
			stateCount[2] = 0;
			stateCount[3] = 0;
			stateCount[4] = 0;
			let currentState = 0;
			for (let j = 0; j < maxJ; j++) if (image.data[j + i * image.width]) {
				if ((currentState & 1) === 1) currentState++;
				stateCount[currentState]++;
			} else if ((currentState & 1) === 0) if (currentState === 4) if (this.foundPatternCross(stateCount)) {
				if (this.handlePossibleCenter(stateCount, i, j)) {
					iSkip = 2;
					if (this.hasSkipped) done = this.haveMultiplyConfirmedCenters();
					else {
						const rowSkip = this.findRowSkip();
						if (rowSkip > stateCount[2]) {
							i += rowSkip - stateCount[2] - iSkip;
							j = maxJ - 1;
						}
					}
				} else {
					do
						j++;
					while (j < maxJ && !image.data[j + i * image.width]);
					j--;
				}
				currentState = 0;
				stateCount[0] = 0;
				stateCount[1] = 0;
				stateCount[2] = 0;
				stateCount[3] = 0;
				stateCount[4] = 0;
			} else {
				stateCount[0] = stateCount[2];
				stateCount[1] = stateCount[3];
				stateCount[2] = stateCount[4];
				stateCount[3] = 1;
				stateCount[4] = 0;
				currentState = 3;
			}
			else stateCount[++currentState]++;
			else stateCount[currentState]++;
			if (this.foundPatternCross(stateCount)) {
				if (this.handlePossibleCenter(stateCount, i, maxJ)) {
					iSkip = stateCount[0];
					if (this.hasSkipped) done = this.haveMultiplyConfirmedCenters();
				}
			}
		}
		const patternInfo = this.selectBestPatterns();
		orderBestPatterns(patternInfo);
		return new FinderPatternInfo(patternInfo);
	}
};
//#endregion
//#region src/reader/detect/grid.ts
/**
* @file 二维码解析内部模块：grid。
*/
const GridSampler = {
	checkAndNudgePoints(image, points) {
		const width = image.width;
		const height = image.height;
		let nudged = true;
		for (let offset = 0; offset < points.length && nudged; offset += 2) {
			const x = Math.floor(points[offset]);
			const y = Math.floor(points[offset + 1]);
			if (x < -1 || x > width || y < -1 || y > height) throw new Error("Error.checkAndNudgePoints");
			nudged = false;
			if (x === -1) {
				points[offset] = 0;
				nudged = true;
			} else if (x === width) {
				points[offset] = width - 1;
				nudged = true;
			}
			if (y === -1) {
				points[offset + 1] = 0;
				nudged = true;
			} else if (y === height) {
				points[offset + 1] = height - 1;
				nudged = true;
			}
		}
		nudged = true;
		for (let offset = points.length - 2; offset >= 0 && nudged; offset -= 2) {
			const x = Math.floor(points[offset]);
			const y = Math.floor(points[offset + 1]);
			if (x < -1 || x > width || y < -1 || y > height) throw new Error("Error.checkAndNudgePoints");
			nudged = false;
			if (x === -1) {
				points[offset] = 0;
				nudged = true;
			} else if (x === width) {
				points[offset] = width - 1;
				nudged = true;
			}
			if (y === -1) {
				points[offset + 1] = 0;
				nudged = true;
			} else if (y === height) {
				points[offset + 1] = height - 1;
				nudged = true;
			}
		}
	},
	sampleGrid3(image, dimension, transform) {
		const bits = new BitMatrix(dimension);
		const points = Array.from({ length: dimension << 1 }, () => 0);
		for (let y = 0; y < dimension; y++) {
			const max = points.length;
			const iValue = y + .5;
			for (let x = 0; x < max; x += 2) {
				points[x] = (x >> 1) + .5;
				points[x + 1] = iValue;
			}
			transform.transformPoints1(points);
			GridSampler.checkAndNudgePoints(image, points);
			try {
				for (let x = 0; x < max; x += 2) if (image.data[Math.floor(points[x]) + image.width * Math.floor(points[x + 1])]) bits.set_Renamed(x >> 1, y);
			} catch {
				throw new Error("Error.checkAndNudgePoints");
			}
		}
		return bits;
	}
};
//#endregion
//#region src/reader/detect/detector.ts
var PerspectiveTransform = class PerspectiveTransform {
	static quadrilateralToQuadrilateral(x0, y0, x1, y1, x2, y2, x3, y3, x0p, y0p, x1p, y1p, x2p, y2p, x3p, y3p) {
		const qToS = PerspectiveTransform.quadrilateralToSquare(x0, y0, x1, y1, x2, y2, x3, y3);
		return PerspectiveTransform.squareToQuadrilateral(x0p, y0p, x1p, y1p, x2p, y2p, x3p, y3p).times(qToS);
	}
	static squareToQuadrilateral(x0, y0, x1, y1, x2, y2, x3, y3) {
		const dy2 = y3 - y2;
		const dy3 = y0 - y1 + y2 - y3;
		if (dy2 === 0 && dy3 === 0) return new PerspectiveTransform(x1 - x0, x2 - x1, x0, y1 - y0, y2 - y1, y0, 0, 0, 1);
		const dx1 = x1 - x2;
		const dx2 = x3 - x2;
		const dx3 = x0 - x1 + x2 - x3;
		const dy1 = y1 - y2;
		const denominator = dx1 * dy2 - dx2 * dy1;
		const a13 = (dx3 * dy2 - dx2 * dy3) / denominator;
		const a23 = (dx1 * dy3 - dx3 * dy1) / denominator;
		return new PerspectiveTransform(x1 - x0 + a13 * x1, x3 - x0 + a23 * x3, x0, y1 - y0 + a13 * y1, y3 - y0 + a23 * y3, y0, a13, a23, 1);
	}
	static quadrilateralToSquare(x0, y0, x1, y1, x2, y2, x3, y3) {
		return PerspectiveTransform.squareToQuadrilateral(x0, y0, x1, y1, x2, y2, x3, y3).buildAdjoint();
	}
	a11;
	a12;
	a13;
	a21;
	a22;
	a23;
	a31;
	a32;
	a33;
	constructor(a11, a21, a31, a12, a22, a32, a13, a23, a33) {
		this.a11 = a11;
		this.a12 = a12;
		this.a13 = a13;
		this.a21 = a21;
		this.a22 = a22;
		this.a23 = a23;
		this.a31 = a31;
		this.a32 = a32;
		this.a33 = a33;
	}
	transformPoints1(points) {
		for (let i = 0; i < points.length; i += 2) {
			const x = points[i];
			const y = points[i + 1];
			const denominator = this.a13 * x + this.a23 * y + this.a33;
			points[i] = (this.a11 * x + this.a21 * y + this.a31) / denominator;
			points[i + 1] = (this.a12 * x + this.a22 * y + this.a32) / denominator;
		}
	}
	transformPoints2(xValues, yValues) {
		for (let i = 0; i < xValues.length; i++) {
			const x = xValues[i];
			const y = yValues[i];
			const denominator = this.a13 * x + this.a23 * y + this.a33;
			xValues[i] = (this.a11 * x + this.a21 * y + this.a31) / denominator;
			yValues[i] = (this.a12 * x + this.a22 * y + this.a32) / denominator;
		}
	}
	buildAdjoint() {
		return new PerspectiveTransform(this.a22 * this.a33 - this.a23 * this.a32, this.a23 * this.a31 - this.a21 * this.a33, this.a21 * this.a32 - this.a22 * this.a31, this.a13 * this.a32 - this.a12 * this.a33, this.a11 * this.a33 - this.a13 * this.a31, this.a12 * this.a31 - this.a11 * this.a32, this.a12 * this.a23 - this.a13 * this.a22, this.a13 * this.a21 - this.a11 * this.a23, this.a11 * this.a22 - this.a12 * this.a21);
	}
	times(other) {
		return new PerspectiveTransform(this.a11 * other.a11 + this.a21 * other.a12 + this.a31 * other.a13, this.a11 * other.a21 + this.a21 * other.a22 + this.a31 * other.a23, this.a11 * other.a31 + this.a21 * other.a32 + this.a31 * other.a33, this.a12 * other.a11 + this.a22 * other.a12 + this.a32 * other.a13, this.a12 * other.a21 + this.a22 * other.a22 + this.a32 * other.a23, this.a12 * other.a31 + this.a22 * other.a32 + this.a32 * other.a33, this.a13 * other.a11 + this.a23 * other.a12 + this.a33 * other.a13, this.a13 * other.a21 + this.a23 * other.a22 + this.a33 * other.a23, this.a13 * other.a31 + this.a23 * other.a32 + this.a33 * other.a33);
	}
};
var DetectorResult = class {
	bits;
	points;
	constructor(bits, points) {
		this.bits = bits;
		this.points = points;
	}
};
var Detector = class {
	image;
	resultPointCallback;
	constructor(image) {
		this.image = image;
		this.resultPointCallback = null;
	}
	sizeOfBlackWhiteBlackRun(fromX, fromY, toX, toY) {
		const steep = Math.abs(toY - fromY) > Math.abs(toX - fromX);
		if (steep) {
			[fromX, fromY] = [fromY, fromX];
			[toX, toY] = [toY, toX];
		}
		const dx = Math.abs(toX - fromX);
		const dy = Math.abs(toY - fromY);
		let error = -dx >> 1;
		const ystep = fromY < toY ? 1 : -1;
		const xstep = fromX < toX ? 1 : -1;
		let state = 0;
		for (let x = fromX, y = fromY; x !== toX; x += xstep) {
			const realX = steep ? y : x;
			const realY = steep ? x : y;
			if (state === 1) {
				if (this.image.data[realX + realY * this.image.width]) state++;
			} else if (!this.image.data[realX + realY * this.image.width]) state++;
			if (state === 3) {
				const diffX = x - fromX;
				const diffY = y - fromY;
				return Math.sqrt(diffX * diffX + diffY * diffY);
			}
			error += dy;
			if (error > 0) {
				if (y === toY) break;
				y += ystep;
				error -= dx;
			}
		}
		const diffX2 = toX - fromX;
		const diffY2 = toY - fromY;
		return Math.sqrt(diffX2 * diffX2 + diffY2 * diffY2);
	}
	sizeOfBlackWhiteBlackRunBothWays(fromX, fromY, toX, toY) {
		let result = this.sizeOfBlackWhiteBlackRun(fromX, fromY, toX, toY);
		let scale = 1;
		let otherToX = fromX - (toX - fromX);
		if (otherToX < 0) {
			scale = fromX / (fromX - otherToX);
			otherToX = 0;
		} else if (otherToX >= this.image.width) {
			scale = (this.image.width - 1 - fromX) / (otherToX - fromX);
			otherToX = this.image.width - 1;
		}
		let otherToY = Math.floor(fromY - (toY - fromY) * scale);
		scale = 1;
		if (otherToY < 0) {
			scale = fromY / (fromY - otherToY);
			otherToY = 0;
		} else if (otherToY >= this.image.height) {
			scale = (this.image.height - 1 - fromY) / (otherToY - fromY);
			otherToY = this.image.height - 1;
		}
		otherToX = Math.floor(fromX + (otherToX - fromX) * scale);
		result += this.sizeOfBlackWhiteBlackRun(fromX, fromY, otherToX, otherToY);
		return result - 1;
	}
	calculateModuleSizeOneWay(pattern, otherPattern) {
		const moduleSizeEst1 = this.sizeOfBlackWhiteBlackRunBothWays(Math.floor(pattern.X), Math.floor(pattern.Y), Math.floor(otherPattern.X), Math.floor(otherPattern.Y));
		const moduleSizeEst2 = this.sizeOfBlackWhiteBlackRunBothWays(Math.floor(otherPattern.X), Math.floor(otherPattern.Y), Math.floor(pattern.X), Math.floor(pattern.Y));
		if (Number.isNaN(moduleSizeEst1)) return moduleSizeEst2 / 7;
		if (Number.isNaN(moduleSizeEst2)) return moduleSizeEst1 / 7;
		return (moduleSizeEst1 + moduleSizeEst2) / 14;
	}
	calculateModuleSize(topLeft, topRight, bottomLeft) {
		return (this.calculateModuleSizeOneWay(topLeft, topRight) + this.calculateModuleSizeOneWay(topLeft, bottomLeft)) / 2;
	}
	distance(pattern1, pattern2) {
		const xDiff = pattern1.X - pattern2.X;
		const yDiff = pattern1.Y - pattern2.Y;
		return Math.sqrt(xDiff * xDiff + yDiff * yDiff);
	}
	computeDimension(topLeft, topRight, bottomLeft, moduleSize) {
		let dimension = (Math.round(this.distance(topLeft, topRight) / moduleSize) + Math.round(this.distance(topLeft, bottomLeft) / moduleSize) >> 1) + 7;
		switch (dimension & 3) {
			case 0:
				dimension++;
				break;
			case 2:
				dimension--;
				break;
			case 3: throw new Error("Error");
		}
		return dimension;
	}
	findAlignmentInRegion(overallEstModuleSize, estAlignmentX, estAlignmentY, allowanceFactor) {
		const allowance = Math.floor(allowanceFactor * overallEstModuleSize);
		const alignmentAreaLeftX = Math.max(0, estAlignmentX - allowance);
		const alignmentAreaRightX = Math.min(this.image.width - 1, estAlignmentX + allowance);
		if (alignmentAreaRightX - alignmentAreaLeftX < overallEstModuleSize * 3) throw new Error("Error");
		const alignmentAreaTopY = Math.max(0, estAlignmentY - allowance);
		const alignmentAreaBottomY = Math.min(this.image.height - 1, estAlignmentY + allowance);
		return new AlignmentPatternFinder(this.image, alignmentAreaLeftX, alignmentAreaTopY, alignmentAreaRightX - alignmentAreaLeftX, alignmentAreaBottomY - alignmentAreaTopY, overallEstModuleSize, this.resultPointCallback).find();
	}
	createTransform(topLeft, topRight, bottomLeft, alignmentPattern, dimension) {
		const dimMinusThree = dimension - 3.5;
		let bottomRightX;
		let bottomRightY;
		let sourceBottomRightX;
		let sourceBottomRightY;
		if (alignmentPattern != null) {
			bottomRightX = alignmentPattern.X;
			bottomRightY = alignmentPattern.Y;
			sourceBottomRightX = dimMinusThree - 3;
			sourceBottomRightY = dimMinusThree - 3;
		} else {
			bottomRightX = topRight.X - topLeft.X + bottomLeft.X;
			bottomRightY = topRight.Y - topLeft.Y + bottomLeft.Y;
			sourceBottomRightX = dimMinusThree;
			sourceBottomRightY = dimMinusThree;
		}
		return PerspectiveTransform.quadrilateralToQuadrilateral(3.5, 3.5, dimMinusThree, 3.5, sourceBottomRightX, sourceBottomRightY, 3.5, dimMinusThree, topLeft.X, topLeft.Y, topRight.X, topRight.Y, bottomRightX, bottomRightY, bottomLeft.X, bottomLeft.Y);
	}
	sampleGrid(image, transform, dimension) {
		return GridSampler.sampleGrid3(image, dimension, transform);
	}
	processFinderPatternInfo(info) {
		const topLeft = info.topLeft;
		const topRight = info.topRight;
		const bottomLeft = info.bottomLeft;
		const moduleSize = this.calculateModuleSize(topLeft, topRight, bottomLeft);
		if (moduleSize < 1) throw new Error("Error");
		const dimension = this.computeDimension(topLeft, topRight, bottomLeft, moduleSize);
		const provisionalVersion = Version.getProvisionalVersionForDimension(dimension);
		const modulesBetweenFPCenters = provisionalVersion.DimensionForVersion - 7;
		let alignmentPattern = null;
		if (provisionalVersion.alignmentPatternCenters.length > 0) {
			const bottomRightX = topRight.X - topLeft.X + bottomLeft.X;
			const bottomRightY = topRight.Y - topLeft.Y + bottomLeft.Y;
			const correctionToTopLeft = 1 - 3 / modulesBetweenFPCenters;
			const estAlignmentX = Math.floor(topLeft.X + correctionToTopLeft * (bottomRightX - topLeft.X));
			const estAlignmentY = Math.floor(topLeft.Y + correctionToTopLeft * (bottomRightY - topLeft.Y));
			alignmentPattern = this.findAlignmentInRegion(moduleSize, estAlignmentX, estAlignmentY, 4);
		}
		const transform = this.createTransform(topLeft, topRight, bottomLeft, alignmentPattern, dimension);
		return new DetectorResult(this.sampleGrid(this.image, transform, dimension), alignmentPattern == null ? [
			bottomLeft,
			topLeft,
			topRight
		] : [
			bottomLeft,
			topLeft,
			topRight,
			alignmentPattern
		]);
	}
	detect() {
		const info = new FinderPatternFinder().findFinderPattern(this.image);
		return this.processFinderPatternInfo(info);
	}
};
//#endregion
//#region src/reader/error-correction/gf256poly.ts
var GF256Poly = class GF256Poly {
	field;
	coefficients;
	constructor(field, coefficients) {
		if (coefficients == null || coefficients.length === 0) throw new Error("System.ArgumentException");
		this.field = field;
		const coefficientsLength = coefficients.length;
		if (coefficientsLength > 1 && coefficients[0] === 0) {
			let firstNonZero = 1;
			while (firstNonZero < coefficientsLength && coefficients[firstNonZero] === 0) firstNonZero++;
			if (firstNonZero === coefficientsLength) this.coefficients = field.Zero.coefficients;
			else {
				this.coefficients = Array.from({ length: coefficientsLength - firstNonZero }, () => 0);
				for (let ci = 0; ci < this.coefficients.length; ci++) this.coefficients[ci] = coefficients[firstNonZero + ci];
			}
		} else this.coefficients = coefficients;
	}
	get Zero() {
		return this.coefficients[0] === 0;
	}
	get Degree() {
		return this.coefficients.length - 1;
	}
	getCoefficient(degree) {
		return this.coefficients[this.coefficients.length - 1 - degree];
	}
	evaluateAt(a) {
		if (a === 0) return this.getCoefficient(0);
		const size = this.coefficients.length;
		if (a === 1) {
			let result = 0;
			for (let i = 0; i < size; i++) result = this.field.addOrSubtract(result, this.coefficients[i]);
			return result;
		}
		let result = this.coefficients[0];
		for (let i = 1; i < size; i++) result = this.field.addOrSubtract(this.field.multiply(a, result), this.coefficients[i]);
		return result;
	}
	addOrSubtract(other) {
		if (this.field !== other.field) throw new Error("GF256Polys do not have same GF256 field");
		if (this.Zero) return other;
		if (other.Zero) return this;
		let smallerCoefficients = this.coefficients;
		let largerCoefficients = other.coefficients;
		if (smallerCoefficients.length > largerCoefficients.length) {
			const temp = smallerCoefficients;
			smallerCoefficients = largerCoefficients;
			largerCoefficients = temp;
		}
		const sumDiff = Array.from({ length: largerCoefficients.length }, () => 0);
		const lengthDiff = largerCoefficients.length - smallerCoefficients.length;
		for (let ci = 0; ci < lengthDiff; ci++) sumDiff[ci] = largerCoefficients[ci];
		for (let i = lengthDiff; i < largerCoefficients.length; i++) sumDiff[i] = this.field.addOrSubtract(smallerCoefficients[i - lengthDiff], largerCoefficients[i]);
		return new GF256Poly(this.field, sumDiff);
	}
	multiply1(other) {
		if (this.field !== other.field) throw new Error("GF256Polys do not have same GF256 field");
		if (this.Zero || other.Zero) return this.field.Zero;
		const aCoefficients = this.coefficients;
		const aLength = aCoefficients.length;
		const bCoefficients = other.coefficients;
		const bLength = bCoefficients.length;
		const product = Array.from({ length: aLength + bLength - 1 }, () => 0);
		for (let i = 0; i < aLength; i++) {
			const aCoeff = aCoefficients[i];
			for (let j = 0; j < bLength; j++) product[i + j] = this.field.addOrSubtract(product[i + j], this.field.multiply(aCoeff, bCoefficients[j]));
		}
		return new GF256Poly(this.field, product);
	}
	multiply2(scalar) {
		if (scalar === 0) return this.field.Zero;
		if (scalar === 1) return this;
		const size = this.coefficients.length;
		const product = Array.from({ length: size }, () => 0);
		for (let i = 0; i < size; i++) product[i] = this.field.multiply(this.coefficients[i], scalar);
		return new GF256Poly(this.field, product);
	}
	multiplyByMonomial(degree, coefficient) {
		if (degree < 0) throw new Error("System.ArgumentException");
		if (coefficient === 0) return this.field.Zero;
		const size = this.coefficients.length;
		const product = Array.from({ length: size + degree }, () => 0);
		for (let i = 0; i < size; i++) product[i] = this.field.multiply(this.coefficients[i], coefficient);
		return new GF256Poly(this.field, product);
	}
	divide(other) {
		if (this.field !== other.field) throw new Error("GF256Polys do not have same GF256 field");
		if (other.Zero) throw new Error("Divide by 0");
		let quotient = this.field.Zero;
		let remainder = new GF256Poly(this.field, this.coefficients);
		const denominatorLeadingTerm = other.getCoefficient(other.Degree);
		const inverseDenominatorLeadingTerm = this.field.inverse(denominatorLeadingTerm);
		while (remainder.Degree >= other.Degree && !remainder.Zero) {
			const degreeDifference = remainder.Degree - other.Degree;
			const scale = this.field.multiply(remainder.getCoefficient(remainder.Degree), inverseDenominatorLeadingTerm);
			const term = other.multiplyByMonomial(degreeDifference, scale);
			const iterationQuotient = this.field.buildMonomial(degreeDifference, scale);
			quotient = quotient.addOrSubtract(iterationQuotient);
			remainder = remainder.addOrSubtract(term);
		}
		return [quotient, remainder];
	}
};
//#endregion
//#region src/reader/error-correction/gf256.ts
/**
* @file 二维码解析内部模块：gf256。
*/
var GF256 = class {
	static QR_CODE_FIELD;
	static DATA_MATRIX_FIELD;
	expTable;
	logTable;
	zero;
	one;
	constructor(primitive) {
		this.expTable = Array.from({ length: 256 }, () => 0);
		this.logTable = Array.from({ length: 256 }, () => 0);
		let x = 1;
		for (let i = 0; i < 256; i++) {
			this.expTable[i] = x;
			x <<= 1;
			if (x >= 256) x ^= primitive;
		}
		for (let i = 0; i < 255; i++) this.logTable[this.expTable[i]] = i;
		this.zero = new GF256Poly(this, [0]);
		this.one = new GF256Poly(this, [1]);
	}
	get Zero() {
		return this.zero;
	}
	get One() {
		return this.one;
	}
	buildMonomial(degree, coefficient) {
		if (degree < 0) throw new Error("System.ArgumentException");
		if (coefficient === 0) return this.zero;
		const coefficients = Array.from({ length: degree + 1 }, () => 0);
		coefficients[0] = coefficient;
		return new GF256Poly(this, coefficients);
	}
	exp(a) {
		return this.expTable[a];
	}
	log(a) {
		if (a === 0) throw new Error("System.ArgumentException");
		return this.logTable[a];
	}
	inverse(a) {
		if (a === 0) throw new Error("System.ArithmeticException");
		return this.expTable[255 - this.logTable[a]];
	}
	addOrSubtract(a, b) {
		return a ^ b;
	}
	multiply(a, b) {
		if (a === 0 || b === 0) return 0;
		if (a === 1) return b;
		if (b === 1) return a;
		return this.expTable[(this.logTable[a] + this.logTable[b]) % 255];
	}
};
GF256.QR_CODE_FIELD = new GF256(285);
GF256.DATA_MATRIX_FIELD = new GF256(301);
//#endregion
//#region src/reader/error-correction/rsdecoder.ts
var ReedSolomonDecoder = class {
	field;
	constructor(field) {
		this.field = field;
	}
	decode(received, twoS) {
		const poly = new GF256Poly(this.field, received);
		const syndromeCoefficients = Array.from({ length: twoS }, () => 0);
		const dataMatrix = false;
		let noError = true;
		for (let i = 0; i < twoS; i++) {
			const evaluation = poly.evaluateAt(this.field.exp(i));
			syndromeCoefficients[syndromeCoefficients.length - 1 - i] = evaluation;
			if (evaluation !== 0) noError = false;
		}
		if (noError) return;
		const syndrome = new GF256Poly(this.field, syndromeCoefficients);
		const sigmaOmega = this.runEuclideanAlgorithm(this.field.buildMonomial(twoS, 1), syndrome, twoS);
		const sigma = sigmaOmega[0];
		const omega = sigmaOmega[1];
		const errorLocations = this.findErrorLocations(sigma);
		const errorMagnitudes = this.findErrorMagnitudes(omega, errorLocations, dataMatrix);
		for (let i = 0; i < errorLocations.length; i++) {
			const position = received.length - 1 - this.field.log(errorLocations[i]);
			if (position < 0) throw new Error("ReedSolomonException Bad error location");
			received[position] = this.field.addOrSubtract(received[position], errorMagnitudes[i]);
		}
	}
	runEuclideanAlgorithm(a, b, rValue) {
		if (a.Degree < b.Degree) {
			const temp = a;
			a = b;
			b = temp;
		}
		let rLast = a;
		let r = b;
		let sLast = this.field.One;
		let s = this.field.Zero;
		let tLast = this.field.Zero;
		let t = this.field.One;
		while (r.Degree >= Math.floor(rValue / 2)) {
			const rLastLast = rLast;
			const sLastLast = sLast;
			const tLastLast = tLast;
			rLast = r;
			sLast = s;
			tLast = t;
			if (rLast.Zero) throw new Error("r_{i-1} was zero");
			r = rLastLast;
			let q = this.field.Zero;
			const denominatorLeadingTerm = rLast.getCoefficient(rLast.Degree);
			const dltInverse = this.field.inverse(denominatorLeadingTerm);
			while (r.Degree >= rLast.Degree && !r.Zero) {
				const degreeDiff = r.Degree - rLast.Degree;
				const scale = this.field.multiply(r.getCoefficient(r.Degree), dltInverse);
				q = q.addOrSubtract(this.field.buildMonomial(degreeDiff, scale));
				r = r.addOrSubtract(rLast.multiplyByMonomial(degreeDiff, scale));
			}
			s = q.multiply1(sLast).addOrSubtract(sLastLast);
			t = q.multiply1(tLast).addOrSubtract(tLastLast);
		}
		const sigmaTildeAtZero = t.getCoefficient(0);
		if (sigmaTildeAtZero === 0) throw new Error("ReedSolomonException sigmaTilde(0) was zero");
		const inverse = this.field.inverse(sigmaTildeAtZero);
		return [t.multiply2(inverse), r.multiply2(inverse)];
	}
	findErrorLocations(errorLocator) {
		const numErrors = errorLocator.Degree;
		if (numErrors === 1) return [errorLocator.getCoefficient(1)];
		const result = Array.from({ length: numErrors }, () => 0);
		let errorIndex = 0;
		for (let i = 1; i < 256 && errorIndex < numErrors; i++) if (errorLocator.evaluateAt(i) === 0) {
			result[errorIndex] = this.field.inverse(i);
			errorIndex++;
		}
		if (errorIndex !== numErrors) throw new Error("Error locator degree does not match number of roots");
		return result;
	}
	findErrorMagnitudes(errorEvaluator, errorLocations, dataMatrix) {
		const size = errorLocations.length;
		const result = Array.from({ length: size }, () => 0);
		for (let i = 0; i < size; i++) {
			const xiInverse = this.field.inverse(errorLocations[i]);
			let denominator = 1;
			for (let j = 0; j < size; j++) if (i !== j) denominator = this.field.multiply(denominator, this.field.addOrSubtract(1, this.field.multiply(errorLocations[j], xiInverse)));
			result[i] = this.field.multiply(errorEvaluator.evaluateAt(xiInverse), this.field.inverse(denominator));
			if (dataMatrix) result[i] = this.field.multiply(result[i], xiInverse);
		}
		return result;
	}
};
//#endregion
//#region src/reader/parse/datamask.ts
/**
* @file 二维码解析内部模块：datamask。
*/
function createDataMaskStrategy(isMasked) {
	return {
		unmaskBitMatrix(bits, dimension) {
			for (let i = 0; i < dimension; i++) for (let j = 0; j < dimension; j++) if (isMasked(i, j)) bits.flip(j, i);
		},
		isMasked
	};
}
const DATA_MASKS = [
	createDataMaskStrategy((i, j) => (i + j & 1) === 0),
	createDataMaskStrategy((i) => (i & 1) === 0),
	createDataMaskStrategy((_i, j) => j % 3 === 0),
	createDataMaskStrategy((i, j) => (i + j) % 3 === 0),
	createDataMaskStrategy((i, j) => (URShift(i, 1) + Math.floor(j / 3) & 1) === 0),
	createDataMaskStrategy((i, j) => {
		const temp = i * j;
		return (temp & 1) + temp % 3 === 0;
	}),
	createDataMaskStrategy((i, j) => {
		const temp = i * j;
		return ((temp & 1) + temp % 3 & 1) === 0;
	}),
	createDataMaskStrategy((i, j) => ((i + j & 1) + i * j % 3 & 1) === 0)
];
const DataMask = {
	DATA_MASKS,
	forReference(reference) {
		if (reference < 0 || reference > 7) throw new Error("System.ArgumentException");
		return DATA_MASKS[reference];
	}
};
//#endregion
//#region src/reader/parse/bmparser.ts
var BitMatrixParser = class {
	bitMatrix;
	parsedVersion;
	parsedFormatInfo;
	constructor(bitMatrix) {
		const dimension = bitMatrix.Dimension;
		if (dimension < 21 || (dimension & 3) !== 1) throw new Error("Error BitMatrixParser");
		this.bitMatrix = bitMatrix;
		this.parsedVersion = null;
		this.parsedFormatInfo = null;
	}
	copyBit(i, j, versionBits) {
		return this.bitMatrix.get_Renamed(i, j) ? versionBits << 1 | 1 : versionBits << 1;
	}
	readFormatInformation() {
		if (this.parsedFormatInfo != null) return this.parsedFormatInfo;
		let formatInfoBits = 0;
		for (let i = 0; i < 6; i++) formatInfoBits = this.copyBit(i, 8, formatInfoBits);
		formatInfoBits = this.copyBit(7, 8, formatInfoBits);
		formatInfoBits = this.copyBit(8, 8, formatInfoBits);
		formatInfoBits = this.copyBit(8, 7, formatInfoBits);
		for (let j = 5; j >= 0; j--) formatInfoBits = this.copyBit(8, j, formatInfoBits);
		this.parsedFormatInfo = FormatInformation.decodeFormatInformation(formatInfoBits);
		if (this.parsedFormatInfo != null) return this.parsedFormatInfo;
		const dimension = this.bitMatrix.Dimension;
		formatInfoBits = 0;
		const iMin = dimension - 8;
		for (let i = dimension - 1; i >= iMin; i--) formatInfoBits = this.copyBit(i, 8, formatInfoBits);
		for (let j = dimension - 7; j < dimension; j++) formatInfoBits = this.copyBit(8, j, formatInfoBits);
		this.parsedFormatInfo = FormatInformation.decodeFormatInformation(formatInfoBits);
		if (this.parsedFormatInfo != null) return this.parsedFormatInfo;
		throw new Error("Error readFormatInformation");
	}
	readVersion() {
		if (this.parsedVersion != null) return this.parsedVersion;
		const dimension = this.bitMatrix.Dimension;
		const provisionalVersion = dimension - 17 >> 2;
		if (provisionalVersion <= 6) return Version.getVersionForNumber(provisionalVersion);
		let versionBits = 0;
		const ijMin = dimension - 11;
		for (let j = 5; j >= 0; j--) for (let i = dimension - 9; i >= ijMin; i--) versionBits = this.copyBit(i, j, versionBits);
		this.parsedVersion = Version.decodeVersionInformation(versionBits);
		if (this.parsedVersion != null && this.parsedVersion.DimensionForVersion === dimension) return this.parsedVersion;
		versionBits = 0;
		for (let i = 5; i >= 0; i--) for (let j = dimension - 9; j >= ijMin; j--) versionBits = this.copyBit(i, j, versionBits);
		this.parsedVersion = Version.decodeVersionInformation(versionBits);
		if (this.parsedVersion != null && this.parsedVersion.DimensionForVersion === dimension) return this.parsedVersion;
		throw new Error("Error readVersion");
	}
	readCodewords() {
		const formatInfo = this.readFormatInformation();
		const version = this.readVersion();
		const dataMask = DataMask.forReference(formatInfo.dataMask);
		const dimension = this.bitMatrix.Dimension;
		dataMask.unmaskBitMatrix(this.bitMatrix, dimension);
		const functionPattern = version.buildFunctionPattern();
		let readingUp = true;
		const result = Array.from({ length: version.totalCodewords }, () => 0);
		let resultOffset = 0;
		let currentByte = 0;
		let bitsRead = 0;
		for (let j = dimension - 1; j > 0; j -= 2) {
			if (j === 6) j--;
			for (let count = 0; count < dimension; count++) {
				const i = readingUp ? dimension - 1 - count : count;
				for (let col = 0; col < 2; col++) if (!functionPattern.get_Renamed(j - col, i)) {
					bitsRead++;
					currentByte <<= 1;
					if (this.bitMatrix.get_Renamed(j - col, i)) currentByte |= 1;
					if (bitsRead === 8) {
						result[resultOffset++] = currentByte;
						bitsRead = 0;
						currentByte = 0;
					}
				}
			}
			readingUp = !readingUp;
		}
		if (resultOffset !== version.totalCodewords) throw new Error("Error readCodewords");
		return result;
	}
};
//#endregion
//#region src/reader/parse/datablock.ts
var DataBlock = class DataBlock {
	static getDataBlocks(rawCodewords, version, ecLevel) {
		if (rawCodewords.length !== version.totalCodewords) throw new Error("ArgumentException");
		const ecBlocks = version.getECBlocksForLevel(ecLevel);
		let totalBlocks = 0;
		const ecBlockArray = ecBlocks.getECBlocks();
		for (let i = 0; i < ecBlockArray.length; i++) totalBlocks += ecBlockArray[i].count;
		const result = Array.from({ length: totalBlocks }, () => null);
		let numResultBlocks = 0;
		for (let j = 0; j < ecBlockArray.length; j++) {
			const ecBlock = ecBlockArray[j];
			for (let i = 0; i < ecBlock.count; i++) {
				const numDataCodewords = ecBlock.dataCodewords;
				const numBlockCodewords = ecBlocks.ecCodewordsPerBlock + numDataCodewords;
				result[numResultBlocks++] = new DataBlock(numDataCodewords, Array.from({ length: numBlockCodewords }, () => 0));
			}
		}
		const shorterBlocksTotalCodewords = result[0].codewords.length;
		let longerBlocksStartAt = result.length - 1;
		while (longerBlocksStartAt >= 0) {
			if (result[longerBlocksStartAt].codewords.length === shorterBlocksTotalCodewords) break;
			longerBlocksStartAt--;
		}
		longerBlocksStartAt++;
		const shorterBlocksNumDataCodewords = shorterBlocksTotalCodewords - ecBlocks.ecCodewordsPerBlock;
		let rawCodewordsOffset = 0;
		for (let i = 0; i < shorterBlocksNumDataCodewords; i++) for (let j = 0; j < numResultBlocks; j++) result[j].codewords[i] = rawCodewords[rawCodewordsOffset++];
		for (let j = longerBlocksStartAt; j < numResultBlocks; j++) result[j].codewords[shorterBlocksNumDataCodewords] = rawCodewords[rawCodewordsOffset++];
		const max = result[0].codewords.length;
		for (let i = shorterBlocksNumDataCodewords; i < max; i++) for (let j = 0; j < numResultBlocks; j++) {
			const iOffset = j < longerBlocksStartAt ? i : i + 1;
			result[j].codewords[iOffset] = rawCodewords[rawCodewordsOffset++];
		}
		return result;
	}
	numDataCodewords;
	codewords;
	constructor(numDataCodewords, codewords) {
		this.numDataCodewords = numDataCodewords;
		this.codewords = codewords;
	}
};
//#endregion
//#region src/reader/parse/databr.ts
/**
* @file 二维码解析内部模块：databr。
*/
const MODE_NUMBER = 1;
const MODE_ROMAN_AND_NUMBER = 2;
const MODE_8BIT_BYTE = 4;
const MODE_KANJI = 8;
const TABLE_ROMAN_AND_FIGURE = [
	"0",
	"1",
	"2",
	"3",
	"4",
	"5",
	"6",
	"7",
	"8",
	"9",
	"A",
	"B",
	"C",
	"D",
	"E",
	"F",
	"G",
	"H",
	"I",
	"J",
	"K",
	"L",
	"M",
	"N",
	"O",
	"P",
	"Q",
	"R",
	"S",
	"T",
	"U",
	"V",
	"W",
	"X",
	"Y",
	"Z",
	" ",
	"$",
	"%",
	"*",
	"+",
	"-",
	".",
	"/",
	":"
];
var QRCodeDataBlockReader = class {
	blockPointer;
	bitPointer;
	dataLength;
	blocks;
	numErrorCorrectionCode;
	dataLengthMode;
	constructor(blocks, version, numErrorCorrectionCode) {
		this.blockPointer = 0;
		this.bitPointer = 7;
		this.dataLength = 0;
		this.blocks = blocks;
		this.numErrorCorrectionCode = numErrorCorrectionCode;
		if (version <= 9) this.dataLengthMode = 0;
		else if (version <= 26) this.dataLengthMode = 1;
		else this.dataLengthMode = 2;
	}
	getNextBits(numBits) {
		let bits = 0;
		if (numBits < this.bitPointer + 1) {
			let mask = 0;
			for (let i = 0; i < numBits; i++) mask += 1 << i;
			mask <<= this.bitPointer - numBits + 1;
			bits = (this.blocks[this.blockPointer] & mask) >> this.bitPointer - numBits + 1;
			this.bitPointer -= numBits;
			return bits;
		}
		if (numBits < this.bitPointer + 1 + 8) {
			let mask1 = 0;
			for (let i = 0; i < this.bitPointer + 1; i++) mask1 += 1 << i;
			bits = (this.blocks[this.blockPointer] & mask1) << numBits - (this.bitPointer + 1);
			this.blockPointer++;
			bits += this.blocks[this.blockPointer] >> 8 - (numBits - (this.bitPointer + 1));
			this.bitPointer -= numBits % 8;
			if (this.bitPointer < 0) this.bitPointer = 8 + this.bitPointer;
			return bits;
		}
		if (numBits < this.bitPointer + 1 + 16) {
			let mask1 = 0;
			let mask3 = 0;
			for (let i = 0; i < this.bitPointer + 1; i++) mask1 += 1 << i;
			const bitsFirstBlock = (this.blocks[this.blockPointer] & mask1) << numBits - (this.bitPointer + 1);
			this.blockPointer++;
			const bitsSecondBlock = this.blocks[this.blockPointer] << numBits - (this.bitPointer + 1 + 8);
			this.blockPointer++;
			for (let i = 0; i < numBits - (this.bitPointer + 1 + 8); i++) mask3 += 1 << i;
			mask3 <<= 8 - (numBits - (this.bitPointer + 1 + 8));
			const bitsThirdBlock = (this.blocks[this.blockPointer] & mask3) >> 8 - (numBits - (this.bitPointer + 1 + 8));
			bits = bitsFirstBlock + bitsSecondBlock + bitsThirdBlock;
			this.bitPointer -= (numBits - 8) % 8;
			if (this.bitPointer < 0) this.bitPointer = 8 + this.bitPointer;
			return bits;
		}
		return 0;
	}
	NextMode() {
		if (this.blockPointer > this.blocks.length - this.numErrorCorrectionCode - 2) return 0;
		return this.getNextBits(4);
	}
	getDataLength(modeIndicator) {
		let index = 0;
		while (true) {
			if (modeIndicator >> index === 1) break;
			index++;
		}
		return this.getNextBits(qrcode.sizeOfDataLengthInfo[this.dataLengthMode][index]);
	}
	getRomanAndFigureString(dataLength) {
		let length = dataLength;
		let intData = 0;
		let strData = "";
		do
			if (length > 1) {
				intData = this.getNextBits(11);
				const firstLetter = Math.floor(intData / 45);
				const secondLetter = intData % 45;
				strData += TABLE_ROMAN_AND_FIGURE[firstLetter];
				strData += TABLE_ROMAN_AND_FIGURE[secondLetter];
				length -= 2;
			} else if (length === 1) {
				intData = this.getNextBits(6);
				strData += TABLE_ROMAN_AND_FIGURE[intData];
				length -= 1;
			}
		while (length > 0);
		return strData;
	}
	getFigureString(dataLength) {
		let length = dataLength;
		let intData = 0;
		let strData = "";
		do {
			if (length >= 3) {
				intData = this.getNextBits(10);
				if (intData < 100) strData += "0";
				if (intData < 10) strData += "0";
				length -= 3;
			} else if (length === 2) {
				intData = this.getNextBits(7);
				if (intData < 10) strData += "0";
				length -= 2;
			} else if (length === 1) {
				intData = this.getNextBits(4);
				length -= 1;
			}
			strData += intData;
		} while (length > 0);
		return strData;
	}
	get8bitByteArray(dataLength) {
		let length = dataLength;
		let intData = 0;
		const output = [];
		do {
			intData = this.getNextBits(8);
			output.push(intData);
			length--;
		} while (length > 0);
		return output;
	}
	getKanjiString(dataLength) {
		let length = dataLength;
		let intData = 0;
		let unicodeString = "";
		do {
			intData = this.getNextBits(13);
			const lowerByte = intData % 192;
			const tempWord = (Math.floor(intData / 192) << 8) + lowerByte;
			let shiftjisWord = 0;
			if (tempWord + 33088 <= 40956) shiftjisWord = tempWord + 33088;
			else shiftjisWord = tempWord + 49472;
			unicodeString += String.fromCharCode(shiftjisWord);
			length--;
		} while (length > 0);
		return unicodeString;
	}
	get DataByte() {
		const output = [];
		do {
			const mode = this.NextMode();
			if (mode === 0) {
				if (output.length > 0) break;
				throw new Error("Empty data block");
			}
			if (mode !== MODE_NUMBER && mode !== MODE_ROMAN_AND_NUMBER && mode !== MODE_8BIT_BYTE && mode !== MODE_KANJI && mode !== 7) throw new Error(`Invalid mode: ${mode} in (block:${this.blockPointer} bit:${this.bitPointer})`);
			const dataLength = this.getDataLength(mode);
			if (dataLength < 1) throw new Error(`Invalid data length: ${dataLength}`);
			switch (mode) {
				case MODE_NUMBER: {
					const temp = this.getFigureString(dataLength);
					output.push(Array.from(temp, (char) => char.charCodeAt(0)));
					break;
				}
				case MODE_ROMAN_AND_NUMBER: {
					const temp = this.getRomanAndFigureString(dataLength);
					output.push(Array.from(temp, (char) => char.charCodeAt(0)));
					break;
				}
				case MODE_8BIT_BYTE:
					output.push(this.get8bitByteArray(dataLength));
					break;
				case MODE_KANJI:
					output.push(this.getKanjiString(dataLength));
					break;
			}
		} while (true);
		return output;
	}
};
//#endregion
//#region src/reader/parse/decoder.ts
/**
* @file 二维码解析内部模块：decoder。
*/
const rsDecoder = new ReedSolomonDecoder(GF256.QR_CODE_FIELD);
const Decoder = {
	rsDecoder,
	correctErrors(codewordBytes, numDataCodewords) {
		const numCodewords = codewordBytes.length;
		const codewordsInts = Array.from({ length: numCodewords }, () => 0);
		for (let i = 0; i < numCodewords; i++) codewordsInts[i] = codewordBytes[i] & 255;
		const numECCodewords = codewordBytes.length - numDataCodewords;
		rsDecoder.decode(codewordsInts, numECCodewords);
		for (let i = 0; i < numDataCodewords; i++) codewordBytes[i] = codewordsInts[i];
	},
	decode(bits) {
		const parser = new BitMatrixParser(bits);
		const version = parser.readVersion();
		const ecLevel = parser.readFormatInformation().errorCorrectionLevel;
		const codewords = parser.readCodewords();
		const dataBlocks = DataBlock.getDataBlocks(codewords, version, ecLevel);
		let totalBytes = 0;
		for (let i = 0; i < dataBlocks.length; i++) totalBytes += dataBlocks[i].numDataCodewords;
		const resultBytes = Array.from({ length: totalBytes }, () => 0);
		let resultOffset = 0;
		for (let j = 0; j < dataBlocks.length; j++) {
			const dataBlock = dataBlocks[j];
			const codewordBytes = dataBlock.codewords;
			const numDataCodewords = dataBlock.numDataCodewords;
			Decoder.correctErrors(codewordBytes, numDataCodewords);
			for (let i = 0; i < numDataCodewords; i++) resultBytes[resultOffset++] = codewordBytes[i];
		}
		return new QRCodeDataBlockReader(resultBytes, version.versionNumber, ecLevel.bits);
	}
};
//#endregion
//#region src/reader/core/qrcode.ts
/**
* @file 二维码解析内部模块：qrcode。
*/
const qrcode = { sizeOfDataLengthInfo: [
	[
		10,
		9,
		8,
		8
	],
	[
		12,
		11,
		16,
		10
	],
	[
		14,
		13,
		16,
		12
	]
] };
var QrCode = class {
	imagedata;
	width;
	height;
	qrCodeSymbol;
	debug;
	error;
	result;
	constructor() {
		this.imagedata = null;
		this.width = 0;
		this.height = 0;
		this.qrCodeSymbol = null;
		this.debug = false;
		this.error = void 0;
		this.result = void 0;
	}
	runDecode() {
		this.error = void 0;
		if (this.imagedata == null) throw new Error("Image data is not initialized");
		this.result = this.process(this.imagedata);
		return this.result;
	}
	async decode(src, data) {
		if (typeof src !== "string" && src.width !== void 0) {
			this.width = src.width;
			this.height = src.height;
			this.imagedata = {
				data: Array.from(data ?? src.data),
				width: src.width,
				height: src.height
			};
			return this.runDecode();
		}
		const imageSrc = src;
		const ImageCtor = globalThis.Image;
		if (ImageCtor == null) throw new TypeError("This source format is not supported in your environment, you need to pass an image buffer with width and height (see https://github.com/edi9999/jsqrcode/blob/master/test/qrcode.js)");
		const image = new ImageCtor();
		image.crossOrigin = "Anonymous";
		return await new Promise((resolve, reject) => {
			image.onload = () => {
				const documentRef = globalThis.document;
				if (documentRef == null) {
					reject(/* @__PURE__ */ new Error("Document API is not available in this environment"));
					return;
				}
				const canvasQr = documentRef.createElement("canvas");
				const context = canvasQr.getContext("2d");
				if (context == null) {
					reject(/* @__PURE__ */ new Error("Canvas 2D context is not available"));
					return;
				}
				const canvasOut = documentRef.getElementById("out-canvas");
				if (canvasOut != null) {
					const outctx = canvasOut.getContext("2d");
					outctx?.clearRect(0, 0, 320, 240);
					outctx?.drawImage(image, 0, 0, 320, 240);
				}
				canvasQr.width = image.width;
				canvasQr.height = image.height;
				context.drawImage(image, 0, 0);
				this.width = image.width;
				this.height = image.height;
				try {
					this.imagedata = context.getImageData(0, 0, image.width, image.height);
				} catch {
					resolve({
						result: "Cross domain image reading not supported in your browser! Save it to your computer then drag and drop the file!",
						points: []
					});
					return;
				}
				try {
					resolve(this.runDecode());
				} catch (error) {
					reject(error instanceof Error ? error : new Error(String(error)));
				}
			};
			image.src = imageSrc;
		});
	}
	decode_utf8(s) {
		return decodeURIComponent(escape(s));
	}
	process(imageData) {
		const start = Date.now();
		const qrCodeMatrix = new Detector(this.grayScaleToBitmap(this.grayscale(imageData))).detect();
		const data = Decoder.decode(qrCodeMatrix.bits).DataByte;
		let str = "";
		for (let i = 0; i < data.length; i++) {
			const chunk = data[i];
			if (typeof chunk === "string") {
				str += chunk;
				continue;
			}
			for (let j = 0; j < chunk.length; j++) str += String.fromCharCode(chunk[j]);
		}
		const time = Date.now() - start;
		if (this.debug) console.warn(`QR Code processing time (ms): ${time}`);
		return {
			result: this.decode_utf8(str),
			points: qrCodeMatrix.points
		};
	}
	getPixel(imageData, x, y) {
		if (imageData.width < x || imageData.height < y) throw new Error("point error");
		const point = x * 4 + y * imageData.width * 4;
		return (imageData.data[point] * 33 + imageData.data[point + 1] * 34 + imageData.data[point + 2] * 33) / 100;
	}
	binarize(th) {
		if (this.imagedata == null) throw new Error("Image data is not initialized");
		const ret = Array.from({ length: this.width * this.height }, () => false);
		for (let y = 0; y < this.height; y++) for (let x = 0; x < this.width; x++) {
			const gray = this.getPixel(this.imagedata, x, y);
			ret[x + y * this.width] = gray <= th;
		}
		return ret;
	}
	getMiddleBrightnessPerArea(imageData) {
		const numSqrtArea = 4;
		const areaWidth = Math.floor(imageData.width / numSqrtArea);
		const areaHeight = Math.floor(imageData.height / numSqrtArea);
		const minmax = Array.from({ length: numSqrtArea }, () => Array.from({ length: numSqrtArea }, () => [0, 0]));
		for (let ay = 0; ay < numSqrtArea; ay++) for (let ax = 0; ax < numSqrtArea; ax++) {
			minmax[ax][ay][0] = 255;
			for (let dy = 0; dy < areaHeight; dy++) for (let dx = 0; dx < areaWidth; dx++) {
				const target = imageData.data[areaWidth * ax + dx + (areaHeight * ay + dy) * imageData.width];
				if (target < minmax[ax][ay][0]) minmax[ax][ay][0] = target;
				if (target > minmax[ax][ay][1]) minmax[ax][ay][1] = target;
			}
		}
		const middle = Array.from({ length: numSqrtArea }, () => Array.from({ length: numSqrtArea }, () => 0));
		for (let ay = 0; ay < numSqrtArea; ay++) for (let ax = 0; ax < numSqrtArea; ax++) middle[ax][ay] = Math.floor((minmax[ax][ay][0] + minmax[ax][ay][1]) / 2);
		return middle;
	}
	grayScaleToBitmap(grayScaleImageData) {
		const middle = this.getMiddleBrightnessPerArea(grayScaleImageData);
		const sqrtNumArea = middle.length;
		const areaWidth = Math.floor(grayScaleImageData.width / sqrtNumArea);
		const areaHeight = Math.floor(grayScaleImageData.height / sqrtNumArea);
		const data = [...grayScaleImageData.data];
		for (let ay = 0; ay < sqrtNumArea; ay++) for (let ax = 0; ax < sqrtNumArea; ax++) for (let dy = 0; dy < areaHeight; dy++) for (let dx = 0; dx < areaWidth; dx++) {
			const index = areaWidth * ax + dx + (areaHeight * ay + dy) * grayScaleImageData.width;
			data[index] = Number(data[index] < middle[ax][ay]);
		}
		return {
			height: grayScaleImageData.height,
			width: grayScaleImageData.width,
			data: data.map((value) => Boolean(value))
		};
	}
	grayscale(imageData) {
		const ret = Array.from({ length: imageData.width * imageData.height }, () => 0);
		for (let y = 0; y < imageData.height; y++) for (let x = 0; x < imageData.width; x++) {
			const gray = this.getPixel(imageData, x, y);
			ret[x + y * imageData.width] = gray;
		}
		return {
			height: imageData.height,
			width: imageData.width,
			data: ret
		};
	}
};
function URShift(number, bits) {
	if (number >= 0) return number >> bits;
	return (number >> bits) + (2 << ~bits);
}
//#endregion
//#region src/reader/core/runtime.ts
/**
* @file 二维码核心 reader 运行时适配。
*/
/** createLegacyQrCodeReader 的方法封装。 */
function createLegacyQrCodeReader() {
	return new QrCode();
}
//#endregion
//#region src/reader/core/decode.ts
/** decodeWithLegacyQrCodeCore 的方法封装。 */
async function decodeWithLegacyQrCodeCore(input) {
	return normalizeQrCodeReaderResult(await createLegacyQrCodeReader().decode(input));
}
//#endregion
//#region src/reader/decode.ts
/**
* @file 二维码 reader 纯函数解码入口。
*/
/** decodeWithQrReader 的方法封装。 */
async function decodeWithQrReader(input) {
	return await decodeWithLegacyQrCodeCore(input);
}
//#endregion
//#region src/decode.ts
/**
* @file 基于 sharp 的二维码解码封装。
*/
async function decodeQrCodeBuffer(buffer) {
	const { data, info } = await sharp(buffer).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
	return (await decodeWithQrReader({
		width: info.width,
		height: info.height,
		data
	})).result;
}
/** decodeQrCodeFromBuffer 的方法封装。 */
async function decodeQrCodeFromBuffer(buffer) {
	return await decodeQrCodeBuffer(buffer);
}
/** decodeQrCodeFromBase64 的方法封装。 */
async function decodeQrCodeFromBase64(content) {
	return await decodeQrCodeFromBuffer(Buffer.from(content, "base64"));
}
/** decodeQrCodeFromFile 的方法封装。 */
async function decodeQrCodeFromFile(filePath) {
	return await decodeQrCodeFromBuffer(await readFile(filePath));
}
//#endregion
//#region src/miniProgramCode.ts
/**
* @file 微信小程序码结构识别封装。
*/
function isForeground(r, g, b) {
	return 255 - r + (255 - g) + (255 - b) > 80;
}
function getQuadrant(point, width, height) {
	const normalizedX = point.x / width;
	const normalizedY = point.y / height;
	if (normalizedX < .35 && normalizedY < .35) return "top-left";
	if (normalizedX > .65 && normalizedY < .35) return "top-right";
	if (normalizedX < .35 && normalizedY > .65) return "bottom-left";
	if (normalizedX > .65 && normalizedY > .65) return "bottom-right";
	return "other";
}
function isSquareLike(component) {
	const ratio = component.width / component.height;
	return ratio >= .7 && ratio <= 1.3;
}
function findConnectedComponents(data, width, height) {
	const visited = new Uint8Array(width * height);
	const components = [];
	function isForegroundIndex(index) {
		const offset = index * 4;
		return isForeground(data[offset], data[offset + 1], data[offset + 2]);
	}
	for (let index = 0; index < width * height; index += 1) {
		if (visited[index] || !isForegroundIndex(index)) continue;
		const queue = [index];
		visited[index] = 1;
		let cursor = 0;
		let count = 0;
		let minX = width;
		let minY = height;
		let maxX = 0;
		let maxY = 0;
		let sumX = 0;
		let sumY = 0;
		let sumR = 0;
		let sumG = 0;
		let sumB = 0;
		while (cursor < queue.length) {
			const current = queue[cursor];
			cursor += 1;
			const x = current % width;
			const y = Math.floor(current / width);
			const offset = current * 4;
			count += 1;
			sumX += x;
			sumY += y;
			sumR += data[offset];
			sumG += data[offset + 1];
			sumB += data[offset + 2];
			minX = Math.min(minX, x);
			minY = Math.min(minY, y);
			maxX = Math.max(maxX, x);
			maxY = Math.max(maxY, y);
			const neighbors = [
				current - 1,
				current + 1,
				current - width,
				current + width
			];
			for (const neighbor of neighbors) {
				if (neighbor < 0 || neighbor >= width * height || visited[neighbor]) continue;
				const neighborX = neighbor % width;
				const neighborY = Math.floor(neighbor / width);
				if (Math.abs(neighborX - x) + Math.abs(neighborY - y) !== 1) continue;
				if (!isForegroundIndex(neighbor)) continue;
				visited[neighbor] = 1;
				queue.push(neighbor);
			}
		}
		components.push({
			left: minX,
			top: minY,
			width: maxX - minX + 1,
			height: maxY - minY + 1,
			count,
			center: {
				x: sumX / count,
				y: sumY / count
			},
			averageColor: [
				sumR / count,
				sumG / count,
				sumB / count
			]
		});
	}
	return components.sort((a, b) => b.count - a.count);
}
function findLocatorPoints(components, width, height) {
	const minSide = Math.min(width, height);
	const candidates = components.filter((component) => {
		const normalizedSize = Math.max(component.width, component.height) / minSide;
		return normalizedSize >= .05 && normalizedSize <= .16 && component.count >= minSide * .6 && isSquareLike(component);
	});
	const topLeft = candidates.find((component) => getQuadrant(component.center, width, height) === "top-left");
	const topRight = candidates.find((component) => getQuadrant(component.center, width, height) === "top-right");
	const bottomLeft = candidates.find((component) => getQuadrant(component.center, width, height) === "bottom-left");
	if (!topLeft || !topRight || !bottomLeft) return;
	return [
		topLeft.center,
		topRight.center,
		bottomLeft.center
	];
}
function findBadgeBounds(components, width, height) {
	const minSide = Math.min(width, height);
	return components.find((component) => {
		const normalizedSize = Math.max(component.width, component.height) / minSide;
		const [r, g, b] = component.averageColor;
		return getQuadrant(component.center, width, height) === "bottom-right" && normalizedSize >= .08 && normalizedSize <= .24 && isSquareLike(component) && g > r + 18 && g > b + 18;
	});
}
function findLogoBounds(components, width, height) {
	const centerX = width / 2;
	const centerY = height / 2;
	const limit = Math.min(width, height) * .22;
	const nearCenter = components.filter((component) => {
		return Math.hypot(component.center.x - centerX, component.center.y - centerY) <= limit && component.width <= width * .5 && component.height <= height * .5 && component.count >= Math.min(width, height) * .3;
	});
	if (nearCenter.length === 0) return;
	const left = Math.min(...nearCenter.map((component) => component.left));
	const top = Math.min(...nearCenter.map((component) => component.top));
	const right = Math.max(...nearCenter.map((component) => component.left + component.width));
	const bottom = Math.max(...nearCenter.map((component) => component.top + component.height));
	return {
		left,
		top,
		width: right - left,
		height: bottom - top
	};
}
function calculateAnnulusCoverage(data, width, height) {
	const centerX = (width - 1) / 2;
	const centerY = (height - 1) / 2;
	const minSide = Math.min(width, height);
	const innerRadius = minSide * .24;
	const outerRadius = minSide * .49;
	const bins = Array.from({ length: 36 }).fill(0);
	for (let y = 0; y < height; y += 1) for (let x = 0; x < width; x += 1) {
		const radius = Math.hypot(x - centerX, y - centerY);
		if (radius < innerRadius || radius > outerRadius) continue;
		const offset = (y * width + x) * 4;
		if (!isForeground(data[offset], data[offset + 1], data[offset + 2])) continue;
		const normalizedAngle = (Math.atan2(y - centerY, x - centerX) + Math.PI) / (Math.PI * 2);
		const index = Math.min(bins.length - 1, Math.floor(normalizedAngle * bins.length));
		bins[index] += 1;
	}
	return bins.filter((value) => value > minSide * .02).length / bins.length;
}
/** detectMiniProgramCodeFromBuffer 的方法封装。 */
async function detectMiniProgramCodeFromBuffer(buffer) {
	const { data, info } = await sharp(buffer).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
	const components = findConnectedComponents(data, info.width, info.height);
	const locatorPoints = findLocatorPoints(components, info.width, info.height);
	const badge = findBadgeBounds(components, info.width, info.height);
	const logoBounds = findLogoBounds(components, info.width, info.height);
	const annulusCoverage = calculateAnnulusCoverage(data, info.width, info.height);
	if (!locatorPoints || !badge || annulusCoverage < .55) return null;
	const aspectRatio = info.width / info.height;
	const confidence = Math.min(.99, .45 + (Math.abs(aspectRatio - 1) <= .05 ? .1 : 0) + .15 + .15 + .15 + .05 + Math.min(.09, annulusCoverage * .09));
	return {
		kind: "wechat-mini-program-code",
		width: info.width,
		height: info.height,
		confidence,
		locatorPoints,
		badgeBounds: badge && {
			left: badge.left,
			top: badge.top,
			width: badge.width,
			height: badge.height
		},
		logoBounds
	};
}
/** detectMiniProgramCodeFromBase64 的方法封装。 */
async function detectMiniProgramCodeFromBase64(content) {
	return await detectMiniProgramCodeFromBuffer(Buffer.from(content, "base64"));
}
/** detectMiniProgramCodeFromFile 的方法封装。 */
async function detectMiniProgramCodeFromFile(filePath) {
	return await detectMiniProgramCodeFromBuffer(await readFile(filePath));
}
//#endregion
//#region src/codeType.ts
/**
* @file 二维码类型检测封装。
*/
async function detectCodeTypeBuffer(buffer) {
	try {
		await decodeQrCodeFromBuffer(buffer);
		return "qr";
	} catch {
		try {
			return await detectMiniProgramCodeFromBuffer(buffer) ? "mini-program-code" : "unknown";
		} catch {
			return "unknown";
		}
	}
}
/** detectCodeTypeFromBuffer 的方法封装。 */
async function detectCodeTypeFromBuffer(buffer) {
	return await detectCodeTypeBuffer(buffer);
}
/** detectCodeTypeFromBase64 的方法封装。 */
async function detectCodeTypeFromBase64(content) {
	return await detectCodeTypeFromBuffer(Buffer.from(content, "base64"));
}
/** detectCodeTypeFromFile 的方法封装。 */
async function detectCodeTypeFromFile(filePath) {
	return await detectCodeTypeFromBuffer(await readFile(filePath));
}
//#endregion
//#region src/terminalEncoder/errorCorrectLevel.ts
/**
* @file 终端二维码编码内部模块：errorCorrectLevel。
*/
const QRErrorCorrectLevel = {
	L: 1,
	M: 0,
	Q: 3,
	H: 2
};
//#endregion
//#region src/terminalEncoder/bitBuffer.ts
/**
* @file 终端二维码编码内部模块：bitBuffer。
*/
var QRBitBuffer = class {
	buffer;
	length;
	constructor() {
		this.buffer = [];
		this.length = 0;
	}
	get(index) {
		const bufIndex = Math.floor(index / 8);
		return (this.buffer[bufIndex] >>> 7 - index % 8 & 1) === 1;
	}
	put(num, length) {
		for (let i = 0; i < length; i++) this.putBit((num >>> length - i - 1 & 1) === 1);
	}
	getLengthInBits() {
		return this.length;
	}
	putBit(bit) {
		const bufIndex = Math.floor(this.length / 8);
		if (this.buffer.length <= bufIndex) this.buffer.push(0);
		if (bit) this.buffer[bufIndex] |= 128 >>> this.length % 8;
		this.length++;
	}
};
//#endregion
//#region src/terminalEncoder/mode.ts
/**
* @file 终端二维码编码内部模块：mode。
*/
const QRMode = {
	MODE_NUMBER: 1,
	MODE_ALPHA_NUM: 2,
	MODE_8BIT_BYTE: 4,
	MODE_KANJI: 8
};
//#endregion
//#region src/terminalEncoder/byte.ts
/**
* @file 终端二维码编码内部模块：byte。
*/
var QR8bitByte = class {
	mode;
	data;
	constructor(data) {
		this.mode = QRMode.MODE_8BIT_BYTE;
		this.data = data;
	}
	getLength() {
		return this.data.length;
	}
	write(buffer) {
		for (let i = 0; i < this.data.length; i++) buffer.put(this.data.charCodeAt(i), 8);
	}
};
//#endregion
//#region src/terminalEncoder/math.ts
/**
* @file 终端二维码编码内部模块：math。
*/
const EXP_TABLE = Array.from({ length: 256 }, () => 0);
const LOG_TABLE = Array.from({ length: 256 }, () => 0);
const QRMath = {
	glog(n) {
		if (n < 1) throw new Error(`glog(${n})`);
		return LOG_TABLE[n];
	},
	gexp(n) {
		while (n < 0) n += 255;
		while (n >= 256) n -= 255;
		return EXP_TABLE[n];
	}
};
for (let i = 0; i < 8; i++) EXP_TABLE[i] = 1 << i;
for (let i = 8; i < 256; i++) EXP_TABLE[i] = EXP_TABLE[i - 4] ^ EXP_TABLE[i - 5] ^ EXP_TABLE[i - 6] ^ EXP_TABLE[i - 8];
for (let i = 0; i < 255; i++) LOG_TABLE[EXP_TABLE[i]] = i;
//#endregion
//#region src/terminalEncoder/polynomial.ts
/**
* @file 终端二维码编码内部模块：polynomial。
*/
var QRPolynomial = class QRPolynomial {
	num;
	constructor(num, shift) {
		if (num.length === void 0) throw new Error(`${num.length}/${shift}`);
		let offset = 0;
		while (offset < num.length && num[offset] === 0) offset++;
		this.num = Array.from({ length: num.length - offset + shift }, () => 0);
		for (let i = 0; i < num.length - offset; i++) this.num[i] = num[i + offset];
	}
	get(index) {
		return this.num[index];
	}
	getLength() {
		return this.num.length;
	}
	multiply(e) {
		const num = Array.from({ length: this.getLength() + e.getLength() - 1 }, () => 0);
		for (let i = 0; i < this.getLength(); i++) for (let j = 0; j < e.getLength(); j++) num[i + j] ^= QRMath.gexp(QRMath.glog(this.get(i)) + QRMath.glog(e.get(j)));
		return new QRPolynomial(num, 0);
	}
	mod(e) {
		if (this.getLength() - e.getLength() < 0) return this;
		const ratio = QRMath.glog(this.get(0)) - QRMath.glog(e.get(0));
		const num = Array.from({ length: this.getLength() }, () => 0);
		for (let i = 0; i < this.getLength(); i++) num[i] = this.get(i);
		for (let x = 0; x < e.getLength(); x++) num[x] ^= QRMath.gexp(QRMath.glog(e.get(x)) + ratio);
		return new QRPolynomial(num, 0).mod(e);
	}
};
//#endregion
//#region src/terminalEncoder/rsBlock.ts
/**
* @file 终端二维码编码内部模块：rsBlock。
*/
var QRRSBlock = class QRRSBlock {
	totalCount;
	dataCount;
	static RS_BLOCK_TABLE = [
		[
			1,
			26,
			19
		],
		[
			1,
			26,
			16
		],
		[
			1,
			26,
			13
		],
		[
			1,
			26,
			9
		],
		[
			1,
			44,
			34
		],
		[
			1,
			44,
			28
		],
		[
			1,
			44,
			22
		],
		[
			1,
			44,
			16
		],
		[
			1,
			70,
			55
		],
		[
			1,
			70,
			44
		],
		[
			2,
			35,
			17
		],
		[
			2,
			35,
			13
		],
		[
			1,
			100,
			80
		],
		[
			2,
			50,
			32
		],
		[
			2,
			50,
			24
		],
		[
			4,
			25,
			9
		],
		[
			1,
			134,
			108
		],
		[
			2,
			67,
			43
		],
		[
			2,
			33,
			15,
			2,
			34,
			16
		],
		[
			2,
			33,
			11,
			2,
			34,
			12
		],
		[
			2,
			86,
			68
		],
		[
			4,
			43,
			27
		],
		[
			4,
			43,
			19
		],
		[
			4,
			43,
			15
		],
		[
			2,
			98,
			78
		],
		[
			4,
			49,
			31
		],
		[
			2,
			32,
			14,
			4,
			33,
			15
		],
		[
			4,
			39,
			13,
			1,
			40,
			14
		],
		[
			2,
			121,
			97
		],
		[
			2,
			60,
			38,
			2,
			61,
			39
		],
		[
			4,
			40,
			18,
			2,
			41,
			19
		],
		[
			4,
			40,
			14,
			2,
			41,
			15
		],
		[
			2,
			146,
			116
		],
		[
			3,
			58,
			36,
			2,
			59,
			37
		],
		[
			4,
			36,
			16,
			4,
			37,
			17
		],
		[
			4,
			36,
			12,
			4,
			37,
			13
		],
		[
			2,
			86,
			68,
			2,
			87,
			69
		],
		[
			4,
			69,
			43,
			1,
			70,
			44
		],
		[
			6,
			43,
			19,
			2,
			44,
			20
		],
		[
			6,
			43,
			15,
			2,
			44,
			16
		],
		[
			4,
			101,
			81
		],
		[
			1,
			80,
			50,
			4,
			81,
			51
		],
		[
			4,
			50,
			22,
			4,
			51,
			23
		],
		[
			3,
			36,
			12,
			8,
			37,
			13
		],
		[
			2,
			116,
			92,
			2,
			117,
			93
		],
		[
			6,
			58,
			36,
			2,
			59,
			37
		],
		[
			4,
			46,
			20,
			6,
			47,
			21
		],
		[
			7,
			42,
			14,
			4,
			43,
			15
		],
		[
			4,
			133,
			107
		],
		[
			8,
			59,
			37,
			1,
			60,
			38
		],
		[
			8,
			44,
			20,
			4,
			45,
			21
		],
		[
			12,
			33,
			11,
			4,
			34,
			12
		],
		[
			3,
			145,
			115,
			1,
			146,
			116
		],
		[
			4,
			64,
			40,
			5,
			65,
			41
		],
		[
			11,
			36,
			16,
			5,
			37,
			17
		],
		[
			11,
			36,
			12,
			5,
			37,
			13
		],
		[
			5,
			109,
			87,
			1,
			110,
			88
		],
		[
			5,
			65,
			41,
			5,
			66,
			42
		],
		[
			5,
			54,
			24,
			7,
			55,
			25
		],
		[
			11,
			36,
			12
		],
		[
			5,
			122,
			98,
			1,
			123,
			99
		],
		[
			7,
			73,
			45,
			3,
			74,
			46
		],
		[
			15,
			43,
			19,
			2,
			44,
			20
		],
		[
			3,
			45,
			15,
			13,
			46,
			16
		],
		[
			1,
			135,
			107,
			5,
			136,
			108
		],
		[
			10,
			74,
			46,
			1,
			75,
			47
		],
		[
			1,
			50,
			22,
			15,
			51,
			23
		],
		[
			2,
			42,
			14,
			17,
			43,
			15
		],
		[
			5,
			150,
			120,
			1,
			151,
			121
		],
		[
			9,
			69,
			43,
			4,
			70,
			44
		],
		[
			17,
			50,
			22,
			1,
			51,
			23
		],
		[
			2,
			42,
			14,
			19,
			43,
			15
		],
		[
			3,
			141,
			113,
			4,
			142,
			114
		],
		[
			3,
			70,
			44,
			11,
			71,
			45
		],
		[
			17,
			47,
			21,
			4,
			48,
			22
		],
		[
			9,
			39,
			13,
			16,
			40,
			14
		],
		[
			3,
			135,
			107,
			5,
			136,
			108
		],
		[
			3,
			67,
			41,
			13,
			68,
			42
		],
		[
			15,
			54,
			24,
			5,
			55,
			25
		],
		[
			15,
			43,
			15,
			10,
			44,
			16
		],
		[
			4,
			144,
			116,
			4,
			145,
			117
		],
		[
			17,
			68,
			42
		],
		[
			17,
			50,
			22,
			6,
			51,
			23
		],
		[
			19,
			46,
			16,
			6,
			47,
			17
		],
		[
			2,
			139,
			111,
			7,
			140,
			112
		],
		[
			17,
			74,
			46
		],
		[
			7,
			54,
			24,
			16,
			55,
			25
		],
		[
			34,
			37,
			13
		],
		[
			4,
			151,
			121,
			5,
			152,
			122
		],
		[
			4,
			75,
			47,
			14,
			76,
			48
		],
		[
			11,
			54,
			24,
			14,
			55,
			25
		],
		[
			16,
			45,
			15,
			14,
			46,
			16
		],
		[
			6,
			147,
			117,
			4,
			148,
			118
		],
		[
			6,
			73,
			45,
			14,
			74,
			46
		],
		[
			11,
			54,
			24,
			16,
			55,
			25
		],
		[
			30,
			46,
			16,
			2,
			47,
			17
		],
		[
			8,
			132,
			106,
			4,
			133,
			107
		],
		[
			8,
			75,
			47,
			13,
			76,
			48
		],
		[
			7,
			54,
			24,
			22,
			55,
			25
		],
		[
			22,
			45,
			15,
			13,
			46,
			16
		],
		[
			10,
			142,
			114,
			2,
			143,
			115
		],
		[
			19,
			74,
			46,
			4,
			75,
			47
		],
		[
			28,
			50,
			22,
			6,
			51,
			23
		],
		[
			33,
			46,
			16,
			4,
			47,
			17
		],
		[
			8,
			152,
			122,
			4,
			153,
			123
		],
		[
			22,
			73,
			45,
			3,
			74,
			46
		],
		[
			8,
			53,
			23,
			26,
			54,
			24
		],
		[
			12,
			45,
			15,
			28,
			46,
			16
		],
		[
			3,
			147,
			117,
			10,
			148,
			118
		],
		[
			3,
			73,
			45,
			23,
			74,
			46
		],
		[
			4,
			54,
			24,
			31,
			55,
			25
		],
		[
			11,
			45,
			15,
			31,
			46,
			16
		],
		[
			7,
			146,
			116,
			7,
			147,
			117
		],
		[
			21,
			73,
			45,
			7,
			74,
			46
		],
		[
			1,
			53,
			23,
			37,
			54,
			24
		],
		[
			19,
			45,
			15,
			26,
			46,
			16
		],
		[
			5,
			145,
			115,
			10,
			146,
			116
		],
		[
			19,
			75,
			47,
			10,
			76,
			48
		],
		[
			15,
			54,
			24,
			25,
			55,
			25
		],
		[
			23,
			45,
			15,
			25,
			46,
			16
		],
		[
			13,
			145,
			115,
			3,
			146,
			116
		],
		[
			2,
			74,
			46,
			29,
			75,
			47
		],
		[
			42,
			54,
			24,
			1,
			55,
			25
		],
		[
			23,
			45,
			15,
			28,
			46,
			16
		],
		[
			17,
			145,
			115
		],
		[
			10,
			74,
			46,
			23,
			75,
			47
		],
		[
			10,
			54,
			24,
			35,
			55,
			25
		],
		[
			19,
			45,
			15,
			35,
			46,
			16
		],
		[
			17,
			145,
			115,
			1,
			146,
			116
		],
		[
			14,
			74,
			46,
			21,
			75,
			47
		],
		[
			29,
			54,
			24,
			19,
			55,
			25
		],
		[
			11,
			45,
			15,
			46,
			46,
			16
		],
		[
			13,
			145,
			115,
			6,
			146,
			116
		],
		[
			14,
			74,
			46,
			23,
			75,
			47
		],
		[
			44,
			54,
			24,
			7,
			55,
			25
		],
		[
			59,
			46,
			16,
			1,
			47,
			17
		],
		[
			12,
			151,
			121,
			7,
			152,
			122
		],
		[
			12,
			75,
			47,
			26,
			76,
			48
		],
		[
			39,
			54,
			24,
			14,
			55,
			25
		],
		[
			22,
			45,
			15,
			41,
			46,
			16
		],
		[
			6,
			151,
			121,
			14,
			152,
			122
		],
		[
			6,
			75,
			47,
			34,
			76,
			48
		],
		[
			46,
			54,
			24,
			10,
			55,
			25
		],
		[
			2,
			45,
			15,
			64,
			46,
			16
		],
		[
			17,
			152,
			122,
			4,
			153,
			123
		],
		[
			29,
			74,
			46,
			14,
			75,
			47
		],
		[
			49,
			54,
			24,
			10,
			55,
			25
		],
		[
			24,
			45,
			15,
			46,
			46,
			16
		],
		[
			4,
			152,
			122,
			18,
			153,
			123
		],
		[
			13,
			74,
			46,
			32,
			75,
			47
		],
		[
			48,
			54,
			24,
			14,
			55,
			25
		],
		[
			42,
			45,
			15,
			32,
			46,
			16
		],
		[
			20,
			147,
			117,
			4,
			148,
			118
		],
		[
			40,
			75,
			47,
			7,
			76,
			48
		],
		[
			43,
			54,
			24,
			22,
			55,
			25
		],
		[
			10,
			45,
			15,
			67,
			46,
			16
		],
		[
			19,
			148,
			118,
			6,
			149,
			119
		],
		[
			18,
			75,
			47,
			31,
			76,
			48
		],
		[
			34,
			54,
			24,
			34,
			55,
			25
		],
		[
			20,
			45,
			15,
			61,
			46,
			16
		]
	];
	constructor(totalCount, dataCount) {
		this.totalCount = totalCount;
		this.dataCount = dataCount;
	}
	static getRSBlocks(typeNumber, errorCorrectLevel) {
		const rsBlock = QRRSBlock.getRsBlockTable(typeNumber, errorCorrectLevel);
		if (rsBlock === void 0) throw new Error(`bad rs block @ typeNumber:${typeNumber}/errorCorrectLevel:${errorCorrectLevel}`);
		const length = rsBlock.length / 3;
		const list = [];
		for (let i = 0; i < length; i++) {
			const count = rsBlock[i * 3 + 0];
			const totalCount = rsBlock[i * 3 + 1];
			const dataCount = rsBlock[i * 3 + 2];
			for (let j = 0; j < count; j++) list.push(new QRRSBlock(totalCount, dataCount));
		}
		return list;
	}
	static getRsBlockTable(typeNumber, errorCorrectLevel) {
		switch (errorCorrectLevel) {
			case QRErrorCorrectLevel.L: return QRRSBlock.RS_BLOCK_TABLE[(typeNumber - 1) * 4 + 0];
			case QRErrorCorrectLevel.M: return QRRSBlock.RS_BLOCK_TABLE[(typeNumber - 1) * 4 + 1];
			case QRErrorCorrectLevel.Q: return QRRSBlock.RS_BLOCK_TABLE[(typeNumber - 1) * 4 + 2];
			case QRErrorCorrectLevel.H: return QRRSBlock.RS_BLOCK_TABLE[(typeNumber - 1) * 4 + 3];
			default: return;
		}
	}
};
//#endregion
//#region src/terminalEncoder/maskPattern.ts
/**
* @file 终端二维码编码内部模块：maskPattern。
*/
const QRMaskPattern = {
	PATTERN000: 0,
	PATTERN001: 1,
	PATTERN010: 2,
	PATTERN011: 3,
	PATTERN100: 4,
	PATTERN101: 5,
	PATTERN110: 6,
	PATTERN111: 7
};
//#endregion
//#region src/terminalEncoder/util.ts
/**
* @file 终端二维码编码内部模块：util。
*/
const QRUtil = {
	PATTERN_POSITION_TABLE: [
		[],
		[6, 18],
		[6, 22],
		[6, 26],
		[6, 30],
		[6, 34],
		[
			6,
			22,
			38
		],
		[
			6,
			24,
			42
		],
		[
			6,
			26,
			46
		],
		[
			6,
			28,
			50
		],
		[
			6,
			30,
			54
		],
		[
			6,
			32,
			58
		],
		[
			6,
			34,
			62
		],
		[
			6,
			26,
			46,
			66
		],
		[
			6,
			26,
			48,
			70
		],
		[
			6,
			26,
			50,
			74
		],
		[
			6,
			30,
			54,
			78
		],
		[
			6,
			30,
			56,
			82
		],
		[
			6,
			30,
			58,
			86
		],
		[
			6,
			34,
			62,
			90
		],
		[
			6,
			28,
			50,
			72,
			94
		],
		[
			6,
			26,
			50,
			74,
			98
		],
		[
			6,
			30,
			54,
			78,
			102
		],
		[
			6,
			28,
			54,
			80,
			106
		],
		[
			6,
			32,
			58,
			84,
			110
		],
		[
			6,
			30,
			58,
			86,
			114
		],
		[
			6,
			34,
			62,
			90,
			118
		],
		[
			6,
			26,
			50,
			74,
			98,
			122
		],
		[
			6,
			30,
			54,
			78,
			102,
			126
		],
		[
			6,
			26,
			52,
			78,
			104,
			130
		],
		[
			6,
			30,
			56,
			82,
			108,
			134
		],
		[
			6,
			34,
			60,
			86,
			112,
			138
		],
		[
			6,
			30,
			58,
			86,
			114,
			142
		],
		[
			6,
			34,
			62,
			90,
			118,
			146
		],
		[
			6,
			30,
			54,
			78,
			102,
			126,
			150
		],
		[
			6,
			24,
			50,
			76,
			102,
			128,
			154
		],
		[
			6,
			28,
			54,
			80,
			106,
			132,
			158
		],
		[
			6,
			32,
			58,
			84,
			110,
			136,
			162
		],
		[
			6,
			26,
			54,
			82,
			110,
			138,
			166
		],
		[
			6,
			30,
			58,
			86,
			114,
			142,
			170
		]
	],
	G15: 1335,
	G18: 7973,
	G15_MASK: 21522,
	getBCHTypeInfo(data) {
		let d = data << 10;
		while (QRUtil.getBCHDigit(d) - QRUtil.getBCHDigit(QRUtil.G15) >= 0) d ^= QRUtil.G15 << QRUtil.getBCHDigit(d) - QRUtil.getBCHDigit(QRUtil.G15);
		return (data << 10 | d) ^ QRUtil.G15_MASK;
	},
	getBCHTypeNumber(data) {
		let d = data << 12;
		while (QRUtil.getBCHDigit(d) - QRUtil.getBCHDigit(QRUtil.G18) >= 0) d ^= QRUtil.G18 << QRUtil.getBCHDigit(d) - QRUtil.getBCHDigit(QRUtil.G18);
		return data << 12 | d;
	},
	getBCHDigit(data) {
		let digit = 0;
		while (data !== 0) {
			digit++;
			data >>>= 1;
		}
		return digit;
	},
	getPatternPosition(typeNumber) {
		return QRUtil.PATTERN_POSITION_TABLE[typeNumber - 1];
	},
	getMask(maskPattern, i, j) {
		switch (maskPattern) {
			case QRMaskPattern.PATTERN000: return (i + j) % 2 === 0;
			case QRMaskPattern.PATTERN001: return i % 2 === 0;
			case QRMaskPattern.PATTERN010: return j % 3 === 0;
			case QRMaskPattern.PATTERN011: return (i + j) % 3 === 0;
			case QRMaskPattern.PATTERN100: return (Math.floor(i / 2) + Math.floor(j / 3)) % 2 === 0;
			case QRMaskPattern.PATTERN101: return i * j % 2 + i * j % 3 === 0;
			case QRMaskPattern.PATTERN110: return (i * j % 2 + i * j % 3) % 2 === 0;
			case QRMaskPattern.PATTERN111: return (i * j % 3 + (i + j) % 2) % 2 === 0;
			default: throw new Error(`bad maskPattern:${maskPattern}`);
		}
	},
	getErrorCorrectPolynomial(errorCorrectLength) {
		let a = new QRPolynomial([1], 0);
		for (let i = 0; i < errorCorrectLength; i++) a = a.multiply(new QRPolynomial([1, QRMath.gexp(i)], 0));
		return a;
	},
	getLengthInBits(mode, type) {
		if (type >= 1 && type < 10) switch (mode) {
			case QRMode.MODE_NUMBER: return 10;
			case QRMode.MODE_ALPHA_NUM: return 9;
			case QRMode.MODE_8BIT_BYTE: return 8;
			case QRMode.MODE_KANJI: return 8;
			default: throw new Error(`mode:${mode}`);
		}
		else if (type < 27) switch (mode) {
			case QRMode.MODE_NUMBER: return 12;
			case QRMode.MODE_ALPHA_NUM: return 11;
			case QRMode.MODE_8BIT_BYTE: return 16;
			case QRMode.MODE_KANJI: return 10;
			default: throw new Error(`mode:${mode}`);
		}
		else if (type < 41) switch (mode) {
			case QRMode.MODE_NUMBER: return 14;
			case QRMode.MODE_ALPHA_NUM: return 13;
			case QRMode.MODE_8BIT_BYTE: return 16;
			case QRMode.MODE_KANJI: return 12;
			default: throw new Error(`mode:${mode}`);
		}
		else throw new Error(`type:${type}`);
	},
	getLostPoint(qrCode) {
		const moduleCount = qrCode.getModuleCount();
		let lostPoint = 0;
		let row = 0;
		let col = 0;
		for (row = 0; row < moduleCount; row++) for (col = 0; col < moduleCount; col++) {
			let sameCount = 0;
			const dark = qrCode.isDark(row, col);
			for (let r = -1; r <= 1; r++) {
				if (row + r < 0 || moduleCount <= row + r) continue;
				for (let c = -1; c <= 1; c++) {
					if (col + c < 0 || moduleCount <= col + c) continue;
					if (r === 0 && c === 0) continue;
					if (dark === qrCode.isDark(row + r, col + c)) sameCount++;
				}
			}
			if (sameCount > 5) lostPoint += 3 + sameCount - 5;
		}
		for (row = 0; row < moduleCount - 1; row++) for (col = 0; col < moduleCount - 1; col++) {
			let count = 0;
			if (qrCode.isDark(row, col)) count++;
			if (qrCode.isDark(row + 1, col)) count++;
			if (qrCode.isDark(row, col + 1)) count++;
			if (qrCode.isDark(row + 1, col + 1)) count++;
			if (count === 0 || count === 4) lostPoint += 3;
		}
		for (row = 0; row < moduleCount; row++) for (col = 0; col < moduleCount - 6; col++) if (qrCode.isDark(row, col) && !qrCode.isDark(row, col + 1) && qrCode.isDark(row, col + 2) && qrCode.isDark(row, col + 3) && qrCode.isDark(row, col + 4) && !qrCode.isDark(row, col + 5) && qrCode.isDark(row, col + 6)) lostPoint += 40;
		for (col = 0; col < moduleCount; col++) for (row = 0; row < moduleCount - 6; row++) if (qrCode.isDark(row, col) && !qrCode.isDark(row + 1, col) && qrCode.isDark(row + 2, col) && qrCode.isDark(row + 3, col) && qrCode.isDark(row + 4, col) && !qrCode.isDark(row + 5, col) && qrCode.isDark(row + 6, col)) lostPoint += 40;
		let darkCount = 0;
		for (col = 0; col < moduleCount; col++) for (row = 0; row < moduleCount; row++) if (qrCode.isDark(row, col)) darkCount++;
		const ratio = Math.abs(100 * darkCount / moduleCount / moduleCount - 50) / 5;
		lostPoint += ratio * 10;
		return lostPoint;
	}
};
//#endregion
//#region src/terminalEncoder/qrCode.ts
/**
* @file 终端二维码编码内部模块：qrCode。
*/
var QRCode = class QRCode {
	static PAD0 = 236;
	static PAD1 = 17;
	typeNumber;
	errorCorrectLevel;
	modules;
	moduleCount;
	dataCache;
	dataList;
	constructor(typeNumber, errorCorrectLevel) {
		this.typeNumber = typeNumber;
		this.errorCorrectLevel = errorCorrectLevel;
		this.modules = [];
		this.moduleCount = 0;
		this.dataCache = null;
		this.dataList = [];
	}
	addData(data) {
		const newData = new QR8bitByte(data);
		this.dataList.push(newData);
		this.dataCache = null;
	}
	isDark(row, col) {
		if (row < 0 || this.moduleCount <= row || col < 0 || this.moduleCount <= col) throw new Error(`${row},${col}`);
		return this.modules[row][col] === true;
	}
	getModuleCount() {
		return this.moduleCount;
	}
	make() {
		if (this.typeNumber < 1) {
			let typeNumber = 1;
			for (typeNumber = 1; typeNumber < 40; typeNumber++) {
				const rsBlocks = QRRSBlock.getRSBlocks(typeNumber, this.errorCorrectLevel);
				const buffer = new QRBitBuffer();
				let totalDataCount = 0;
				for (let i = 0; i < rsBlocks.length; i++) totalDataCount += rsBlocks[i].dataCount;
				for (let x = 0; x < this.dataList.length; x++) {
					const data = this.dataList[x];
					buffer.put(data.mode, 4);
					buffer.put(data.getLength(), QRUtil.getLengthInBits(data.mode, typeNumber));
					data.write(buffer);
				}
				if (buffer.getLengthInBits() <= totalDataCount * 8) break;
			}
			this.typeNumber = typeNumber;
		}
		this.makeImpl(false, this.getBestMaskPattern());
	}
	makeImpl(test, maskPattern) {
		this.moduleCount = this.typeNumber * 4 + 17;
		this.modules = Array.from({ length: this.moduleCount }, () => []);
		for (let row = 0; row < this.moduleCount; row++) {
			this.modules[row] = Array.from({ length: this.moduleCount }, () => null);
			for (let col = 0; col < this.moduleCount; col++) this.modules[row][col] = null;
		}
		this.setupPositionProbePattern(0, 0);
		this.setupPositionProbePattern(this.moduleCount - 7, 0);
		this.setupPositionProbePattern(0, this.moduleCount - 7);
		this.setupPositionAdjustPattern();
		this.setupTimingPattern();
		this.setupTypeInfo(test, maskPattern);
		if (this.typeNumber >= 7) this.setupTypeNumber(test);
		if (this.dataCache === null) this.dataCache = QRCode.createData(this.typeNumber, this.errorCorrectLevel, this.dataList);
		this.mapData(this.dataCache, maskPattern);
	}
	setupPositionProbePattern(row, col) {
		for (let r = -1; r <= 7; r++) {
			if (row + r <= -1 || this.moduleCount <= row + r) continue;
			for (let c = -1; c <= 7; c++) {
				if (col + c <= -1 || this.moduleCount <= col + c) continue;
				if (r >= 0 && r <= 6 && (c === 0 || c === 6) || c >= 0 && c <= 6 && (r === 0 || r === 6) || r >= 2 && r <= 4 && c >= 2 && c <= 4) this.modules[row + r][col + c] = true;
				else this.modules[row + r][col + c] = false;
			}
		}
	}
	getBestMaskPattern() {
		let minLostPoint = 0;
		let pattern = 0;
		for (let i = 0; i < 8; i++) {
			this.makeImpl(true, i);
			const lostPoint = QRUtil.getLostPoint(this);
			if (i === 0 || minLostPoint > lostPoint) {
				minLostPoint = lostPoint;
				pattern = i;
			}
		}
		return pattern;
	}
	createMovieClip(target_mc, instance_name, depth) {
		const qr_mc = target_mc.createEmptyMovieClip(instance_name, depth);
		const cs = 1;
		this.make();
		for (let row = 0; row < this.modules.length; row++) {
			const y = row * cs;
			for (let col = 0; col < this.modules[row].length; col++) {
				const x = col * cs;
				if (this.modules[row][col]) {
					qr_mc.beginFill(0, 100);
					qr_mc.moveTo(x, y);
					qr_mc.lineTo(x + cs, y);
					qr_mc.lineTo(x + cs, y + cs);
					qr_mc.lineTo(x, y + cs);
					qr_mc.endFill();
				}
			}
		}
		return qr_mc;
	}
	setupTimingPattern() {
		for (let r = 8; r < this.moduleCount - 8; r++) {
			if (this.modules[r][6] !== null) continue;
			this.modules[r][6] = r % 2 === 0;
		}
		for (let c = 8; c < this.moduleCount - 8; c++) {
			if (this.modules[6][c] !== null) continue;
			this.modules[6][c] = c % 2 === 0;
		}
	}
	setupPositionAdjustPattern() {
		const pos = QRUtil.getPatternPosition(this.typeNumber);
		for (let i = 0; i < pos.length; i++) for (let j = 0; j < pos.length; j++) {
			const row = pos[i];
			const col = pos[j];
			if (this.modules[row][col] !== null) continue;
			for (let r = -2; r <= 2; r++) for (let c = -2; c <= 2; c++) if (Math.abs(r) === 2 || Math.abs(c) === 2 || r === 0 && c === 0) this.modules[row + r][col + c] = true;
			else this.modules[row + r][col + c] = false;
		}
	}
	setupTypeNumber(test) {
		const bits = QRUtil.getBCHTypeNumber(this.typeNumber);
		let mod = false;
		for (let i = 0; i < 18; i++) {
			mod = !test && (bits >> i & 1) === 1;
			this.modules[Math.floor(i / 3)][i % 3 + this.moduleCount - 8 - 3] = mod;
		}
		for (let x = 0; x < 18; x++) {
			mod = !test && (bits >> x & 1) === 1;
			this.modules[x % 3 + this.moduleCount - 8 - 3][Math.floor(x / 3)] = mod;
		}
	}
	setupTypeInfo(test, maskPattern) {
		const data = this.errorCorrectLevel << 3 | maskPattern;
		const bits = QRUtil.getBCHTypeInfo(data);
		let mod = false;
		for (let v = 0; v < 15; v++) {
			mod = !test && (bits >> v & 1) === 1;
			if (v < 6) this.modules[v][8] = mod;
			else if (v < 8) this.modules[v + 1][8] = mod;
			else this.modules[this.moduleCount - 15 + v][8] = mod;
		}
		for (let h = 0; h < 15; h++) {
			mod = !test && (bits >> h & 1) === 1;
			if (h < 8) this.modules[8][this.moduleCount - h - 1] = mod;
			else if (h < 9) this.modules[8][15 - h - 1 + 1] = mod;
			else this.modules[8][15 - h - 1] = mod;
		}
		this.modules[this.moduleCount - 8][8] = !test;
	}
	mapData(data, maskPattern) {
		let inc = -1;
		let row = this.moduleCount - 1;
		let bitIndex = 7;
		let byteIndex = 0;
		for (let col = this.moduleCount - 1; col > 0; col -= 2) {
			if (col === 6) col--;
			while (true) {
				for (let c = 0; c < 2; c++) if (this.modules[row][col - c] === null) {
					let dark = false;
					if (byteIndex < data.length) dark = (data[byteIndex] >>> bitIndex & 1) === 1;
					if (QRUtil.getMask(maskPattern, row, col - c)) dark = !dark;
					this.modules[row][col - c] = dark;
					bitIndex--;
					if (bitIndex === -1) {
						byteIndex++;
						bitIndex = 7;
					}
				}
				row += inc;
				if (row < 0 || this.moduleCount <= row) {
					row -= inc;
					inc = -inc;
					break;
				}
			}
		}
	}
	static createData(typeNumber, errorCorrectLevel, dataList) {
		const rsBlocks = QRRSBlock.getRSBlocks(typeNumber, errorCorrectLevel);
		const buffer = new QRBitBuffer();
		for (let i = 0; i < dataList.length; i++) {
			const data = dataList[i];
			buffer.put(data.mode, 4);
			buffer.put(data.getLength(), QRUtil.getLengthInBits(data.mode, typeNumber));
			data.write(buffer);
		}
		let totalDataCount = 0;
		for (let x = 0; x < rsBlocks.length; x++) totalDataCount += rsBlocks[x].dataCount;
		if (buffer.getLengthInBits() > totalDataCount * 8) throw new Error(`code length overflow. (${buffer.getLengthInBits()}>${totalDataCount * 8})`);
		if (buffer.getLengthInBits() + 4 <= totalDataCount * 8) buffer.put(0, 4);
		while (buffer.getLengthInBits() % 8 !== 0) buffer.putBit(false);
		while (true) {
			if (buffer.getLengthInBits() >= totalDataCount * 8) break;
			buffer.put(QRCode.PAD0, 8);
			if (buffer.getLengthInBits() >= totalDataCount * 8) break;
			buffer.put(QRCode.PAD1, 8);
		}
		return QRCode.createBytes(buffer, rsBlocks);
	}
	static createBytes(buffer, rsBlocks) {
		let offset = 0;
		let maxDcCount = 0;
		let maxEcCount = 0;
		const dcdata = Array.from({ length: rsBlocks.length }, () => []);
		const ecdata = Array.from({ length: rsBlocks.length }, () => []);
		for (let r = 0; r < rsBlocks.length; r++) {
			const dcCount = rsBlocks[r].dataCount;
			const ecCount = rsBlocks[r].totalCount - dcCount;
			maxDcCount = Math.max(maxDcCount, dcCount);
			maxEcCount = Math.max(maxEcCount, ecCount);
			dcdata[r] = Array.from({ length: dcCount }, () => 0);
			for (let i = 0; i < dcdata[r].length; i++) dcdata[r][i] = 255 & buffer.buffer[i + offset];
			offset += dcCount;
			const rsPoly = QRUtil.getErrorCorrectPolynomial(ecCount);
			const modPoly = new QRPolynomial(dcdata[r], rsPoly.getLength() - 1).mod(rsPoly);
			ecdata[r] = Array.from({ length: rsPoly.getLength() - 1 }, () => 0);
			for (let x = 0; x < ecdata[r].length; x++) {
				const modIndex = x + modPoly.getLength() - ecdata[r].length;
				ecdata[r][x] = modIndex >= 0 ? modPoly.get(modIndex) : 0;
			}
		}
		let totalCodeCount = 0;
		for (let y = 0; y < rsBlocks.length; y++) totalCodeCount += rsBlocks[y].totalCount;
		const data = Array.from({ length: totalCodeCount }, () => 0);
		let index = 0;
		for (let z = 0; z < maxDcCount; z++) for (let s = 0; s < rsBlocks.length; s++) if (z < dcdata[s].length) data[index++] = dcdata[s][z];
		for (let xx = 0; xx < maxEcCount; xx++) for (let t = 0; t < rsBlocks.length; t++) if (xx < ecdata[t].length) data[index++] = ecdata[t][xx];
		return data;
	}
};
//#endregion
//#region src/encode.ts
const LOW_ERROR_CORRECTION_LEVEL = QRErrorCorrectLevel.L;
/** createQrCodeMatrix 的方法封装。 */
function createQrCodeMatrix(input) {
	const qrcode = new QRCode(-1, LOW_ERROR_CORRECTION_LEVEL);
	qrcode.addData(input);
	qrcode.make();
	return qrcode.modules;
}
//#endregion
//#region src/render.ts
const WHITE_ALL = "█";
const WHITE_BLACK = "▀";
const BLACK_WHITE = "▄";
const BLACK_ALL = " ";
function fill(length, value) {
	return Array.from({ length }).fill(value);
}
function renderCompactQrCode(matrix) {
	const moduleCount = matrix.length;
	const moduleData = [...matrix];
	const oddRow = moduleCount % 2 === 1;
	if (oddRow) moduleData.push(fill(moduleCount, false));
	const borderTop = Array.from({ length: moduleCount + 2 }).fill(BLACK_WHITE).join("");
	const borderBottom = Array.from({ length: moduleCount + 2 }).fill(WHITE_BLACK).join("");
	let output = `${borderTop}\n`;
	for (let row = 0; row < moduleCount; row += 2) {
		output += WHITE_ALL;
		for (let col = 0; col < moduleCount; col += 1) {
			const top = moduleData[row][col];
			const bottom = moduleData[row + 1][col];
			if (!top && !bottom) output += WHITE_ALL;
			else if (!top && bottom) output += WHITE_BLACK;
			else if (top && !bottom) output += BLACK_WHITE;
			else output += BLACK_ALL;
		}
		output += `${WHITE_ALL}\n`;
	}
	if (!oddRow) output += borderBottom;
	return output;
}
function renderFullQrCode(matrix) {
	const black = "\x1B[40m  \x1B[0m";
	const white = "\x1B[47m  \x1B[0m";
	const border = Array.from({ length: matrix.length + 2 }).fill(white).join("");
	let output = `${border}\n`;
	matrix.forEach((row) => {
		output += white;
		output += row.map((cell) => cell ? black : white).join("");
		output += `${white}\n`;
	});
	output += border;
	return output;
}
/** renderTerminalQrCodeFromMatrix 的方法封装。 */
function renderTerminalQrCodeFromMatrix(matrix, options = {}) {
	return options.small ? renderCompactQrCode(matrix) : renderFullQrCode(matrix);
}
/** renderTerminalQrCode 的方法封装。 */
function renderTerminalQrCode(input, options = {}) {
	return renderTerminalQrCodeFromMatrix(createQrCodeMatrix(input), options);
}
//#endregion
export { createQrCodeMatrix, decodeQrCodeFromBase64, decodeQrCodeFromBuffer, decodeQrCodeFromFile, decodeWithQrReader, detectCodeTypeFromBase64, detectCodeTypeFromBuffer, detectCodeTypeFromFile, detectMiniProgramCodeFromBase64, detectMiniProgramCodeFromBuffer, detectMiniProgramCodeFromFile, renderTerminalQrCode, renderTerminalQrCodeFromMatrix };
